// @azure/monitor-opentelemetry-exporter/dist/commonjs/generated/models/index.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var models_exports = {};
__export(models_exports, {
  AvailabilityData: () => import_models.AvailabilityData,
  ContextTagKeys: () => import_models.ContextTagKeys,
  DataPointType: () => import_models.DataPointType,
  KnownContextTagKeys: () => import_models.KnownContextTagKeys,
  KnownDataPointType: () => import_models.KnownDataPointType,
  KnownMonitorDomainKind: () => import_models.KnownMonitorDomainKind,
  KnownSeverityLevel: () => import_models.KnownSeverityLevel,
  MessageData: () => import_models.MessageData,
  MetricDataPoint: () => import_models.MetricDataPoint,
  MetricsData: () => import_models.MetricsData,
  MonitorBase: () => import_models.MonitorBase,
  MonitorDomain: () => import_models.MonitorDomain,
  MonitorDomainKind: () => import_models.MonitorDomainKind,
  MonitorDomainUnion: () => import_models.MonitorDomainUnion,
  PageViewData: () => import_models.PageViewData,
  PageViewPerfData: () => import_models.PageViewPerfData,
  RemoteDependencyData: () => import_models.RemoteDependencyData,
  RequestData: () => import_models.RequestData,
  SeverityLevel: () => import_models.SeverityLevel,
  StackFrame: () => import_models.StackFrame,
  TelemetryErrorDetails: () => import_models.TelemetryErrorDetails,
  TelemetryEventData: () => import_models.TelemetryEventData,
  TelemetryExceptionData: () => import_models.TelemetryExceptionData,
  TelemetryExceptionDetails: () => import_models.TelemetryExceptionDetails,
  TelemetryItem: () => import_models.TelemetryItem,
  TrackResponse: () => import_models.TrackResponse,
  Versions: () => import_models.Versions
});
module.exports = __toCommonJS(models_exports);
var import_models = require("./models.js");
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  AvailabilityData,
  ContextTagKeys,
  DataPointType,
  KnownContextTagKeys,
  KnownDataPointType,
  KnownMonitorDomainKind,
  KnownSeverityLevel,
  MessageData,
  MetricDataPoint,
  MetricsData,
  MonitorBase,
  MonitorDomain,
  MonitorDomainKind,
  MonitorDomainUnion,
  PageViewData,
  PageViewPerfData,
  RemoteDependencyData,
  RequestData,
  SeverityLevel,
  StackFrame,
  TelemetryErrorDetails,
  TelemetryEventData,
  TelemetryExceptionData,
  TelemetryExceptionDetails,
  TelemetryItem,
  TrackResponse,
  Versions
});
