// @azure/monitor-opentelemetry-exporter/dist/commonjs/generated/models/models.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var models_exports = {};
__export(models_exports, {
  KnownContextTagKeys: () => KnownContextTagKeys,
  KnownDataPointType: () => KnownDataPointType,
  KnownMonitorDomainKind: () => KnownMonitorDomainKind,
  KnownSeverityLevel: () => KnownSeverityLevel,
  availabilityDataDeserializer: () => availabilityDataDeserializer,
  availabilityDataSerializer: () => availabilityDataSerializer,
  messageDataDeserializer: () => messageDataDeserializer,
  messageDataSerializer: () => messageDataSerializer,
  metricDataPointArrayDeserializer: () => metricDataPointArrayDeserializer,
  metricDataPointArraySerializer: () => metricDataPointArraySerializer,
  metricDataPointDeserializer: () => metricDataPointDeserializer,
  metricDataPointSerializer: () => metricDataPointSerializer,
  metricsDataDeserializer: () => metricsDataDeserializer,
  metricsDataSerializer: () => metricsDataSerializer,
  monitorBaseSerializer: () => monitorBaseSerializer,
  monitorDomainDeserializer: () => monitorDomainDeserializer,
  monitorDomainSerializer: () => monitorDomainSerializer,
  monitorDomainUnionDeserializer: () => monitorDomainUnionDeserializer,
  monitorDomainUnionSerializer: () => monitorDomainUnionSerializer,
  pageViewDataDeserializer: () => pageViewDataDeserializer,
  pageViewDataSerializer: () => pageViewDataSerializer,
  pageViewPerfDataDeserializer: () => pageViewPerfDataDeserializer,
  pageViewPerfDataSerializer: () => pageViewPerfDataSerializer,
  remoteDependencyDataDeserializer: () => remoteDependencyDataDeserializer,
  remoteDependencyDataSerializer: () => remoteDependencyDataSerializer,
  requestDataDeserializer: () => requestDataDeserializer,
  requestDataSerializer: () => requestDataSerializer,
  stackFrameArrayDeserializer: () => stackFrameArrayDeserializer,
  stackFrameArraySerializer: () => stackFrameArraySerializer,
  stackFrameDeserializer: () => stackFrameDeserializer,
  stackFrameSerializer: () => stackFrameSerializer,
  telemetryErrorDetailsArrayDeserializer: () => telemetryErrorDetailsArrayDeserializer,
  telemetryErrorDetailsDeserializer: () => telemetryErrorDetailsDeserializer,
  telemetryEventDataDeserializer: () => telemetryEventDataDeserializer,
  telemetryEventDataSerializer: () => telemetryEventDataSerializer,
  telemetryExceptionDataDeserializer: () => telemetryExceptionDataDeserializer,
  telemetryExceptionDataSerializer: () => telemetryExceptionDataSerializer,
  telemetryExceptionDetailsArrayDeserializer: () => telemetryExceptionDetailsArrayDeserializer,
  telemetryExceptionDetailsArraySerializer: () => telemetryExceptionDetailsArraySerializer,
  telemetryExceptionDetailsDeserializer: () => telemetryExceptionDetailsDeserializer,
  telemetryExceptionDetailsSerializer: () => telemetryExceptionDetailsSerializer,
  telemetryItemArraySerializer: () => telemetryItemArraySerializer,
  telemetryItemSerializer: () => telemetryItemSerializer,
  trackResponseDeserializer: () => trackResponseDeserializer
});
module.exports = __toCommonJS(models_exports);
var import_serialize_record = require("../static-helpers/serialization/serialize-record.js");
function telemetryItemSerializer(item) {
  return {
    ver: item["version"],
    name: item["name"],
    time: item["time"].toISOString(),
    sampleRate: item["sampleRate"],
    seq: item["sequence"],
    iKey: item["instrumentationKey"],
    tags: item["tags"],
    data: !item["data"] ? item["data"] : monitorBaseSerializer(item["data"])
  };
}
function monitorBaseSerializer(item) {
  return {
    baseType: item["baseType"],
    baseData: !item["baseData"] ? item["baseData"] : monitorDomainUnionSerializer(item["baseData"])
  };
}
function monitorDomainSerializer(item) {
  return { ...(0, import_serialize_record.serializeRecord)(item.additionalProperties ?? {}), ver: item["version"] };
}
function monitorDomainDeserializer(item) {
  return {
    additionalProperties: (0, import_serialize_record.serializeRecord)(item, ["version", "kind"]),
    version: item["ver"],
    kind: item["kind"]
  };
}
function monitorDomainUnionSerializer(item) {
  switch (item.kind) {
    case "AvailabilityData":
      return availabilityDataSerializer(item);
    case "EventData":
      return telemetryEventDataSerializer(item);
    case "ExceptionData":
      return telemetryExceptionDataSerializer(item);
    case "MessageData":
      return messageDataSerializer(item);
    case "MetricsData":
      return metricsDataSerializer(item);
    case "PageViewData":
      return pageViewDataSerializer(item);
    case "PageViewPerfData":
      return pageViewPerfDataSerializer(item);
    case "RemoteDependencyData":
      return remoteDependencyDataSerializer(item);
    case "RequestData":
      return requestDataSerializer(item);
    default:
      return monitorDomainSerializer(item);
  }
}
function monitorDomainUnionDeserializer(item) {
  switch (item.kind) {
    case "AvailabilityData":
      return availabilityDataDeserializer(item);
    case "EventData":
      return telemetryEventDataDeserializer(item);
    case "ExceptionData":
      return telemetryExceptionDataDeserializer(item);
    case "MessageData":
      return messageDataDeserializer(item);
    case "MetricsData":
      return metricsDataDeserializer(item);
    case "PageViewData":
      return pageViewDataDeserializer(item);
    case "PageViewPerfData":
      return pageViewPerfDataDeserializer(item);
    case "RemoteDependencyData":
      return remoteDependencyDataDeserializer(item);
    case "RequestData":
      return requestDataDeserializer(item);
    default:
      return monitorDomainDeserializer(item);
  }
}
var KnownMonitorDomainKind = /* @__PURE__ */ ((KnownMonitorDomainKind2) => {
  KnownMonitorDomainKind2["AvailabilityData"] = "AvailabilityData";
  KnownMonitorDomainKind2["EventData"] = "EventData";
  KnownMonitorDomainKind2["ExceptionData"] = "ExceptionData";
  KnownMonitorDomainKind2["MessageData"] = "MessageData";
  KnownMonitorDomainKind2["MetricsData"] = "MetricsData";
  KnownMonitorDomainKind2["PageViewData"] = "PageViewData";
  KnownMonitorDomainKind2["PageViewPerfData"] = "PageViewPerfData";
  KnownMonitorDomainKind2["RemoteDependencyData"] = "RemoteDependencyData";
  KnownMonitorDomainKind2["RequestData"] = "RequestData";
  return KnownMonitorDomainKind2;
})(KnownMonitorDomainKind || {});
function availabilityDataSerializer(item) {
  return {
    ...(0, import_serialize_record.serializeRecord)(item.additionalProperties ?? {}),
    ver: item["version"],
    id: item["id"],
    name: item["name"],
    duration: item["duration"],
    success: item["success"],
    runLocation: item["runLocation"],
    message: item["message"],
    properties: item["properties"],
    measurements: item["measurements"]
  };
}
function availabilityDataDeserializer(item) {
  return {
    additionalProperties: (0, import_serialize_record.serializeRecord)(item, [
      "version",
      "kind",
      "id",
      "name",
      "duration",
      "success",
      "runLocation",
      "message",
      "properties",
      "measurements"
    ]),
    version: item["ver"],
    kind: item["kind"],
    id: item["id"],
    name: item["name"],
    duration: item["duration"],
    success: item["success"],
    runLocation: item["runLocation"],
    message: item["message"],
    properties: !item["properties"] ? item["properties"] : Object.fromEntries(
      Object.entries(item["properties"]).map(([k, p]) => [k, p])
    ),
    measurements: !item["measurements"] ? item["measurements"] : Object.fromEntries(
      Object.entries(item["measurements"]).map(([k, p]) => [k, p])
    )
  };
}
function telemetryEventDataSerializer(item) {
  return {
    ...(0, import_serialize_record.serializeRecord)(item.additionalProperties ?? {}),
    ver: item["version"],
    name: item["name"],
    properties: item["properties"],
    measurements: item["measurements"]
  };
}
function telemetryEventDataDeserializer(item) {
  return {
    additionalProperties: (0, import_serialize_record.serializeRecord)(item, [
      "version",
      "kind",
      "name",
      "properties",
      "measurements"
    ]),
    version: item["ver"],
    kind: item["kind"],
    name: item["name"],
    properties: !item["properties"] ? item["properties"] : Object.fromEntries(
      Object.entries(item["properties"]).map(([k, p]) => [k, p])
    ),
    measurements: !item["measurements"] ? item["measurements"] : Object.fromEntries(
      Object.entries(item["measurements"]).map(([k, p]) => [k, p])
    )
  };
}
function telemetryExceptionDataSerializer(item) {
  return {
    ...(0, import_serialize_record.serializeRecord)(item.additionalProperties ?? {}),
    ver: item["version"],
    exceptions: telemetryExceptionDetailsArraySerializer(item["exceptions"]),
    severityLevel: item["severityLevel"],
    problemId: item["problemId"],
    properties: item["properties"],
    measurements: item["measurements"]
  };
}
function telemetryExceptionDataDeserializer(item) {
  return {
    additionalProperties: (0, import_serialize_record.serializeRecord)(item, [
      "version",
      "kind",
      "exceptions",
      "severityLevel",
      "problemId",
      "properties",
      "measurements"
    ]),
    version: item["ver"],
    kind: item["kind"],
    exceptions: telemetryExceptionDetailsArrayDeserializer(item["exceptions"]),
    severityLevel: item["severityLevel"],
    problemId: item["problemId"],
    properties: !item["properties"] ? item["properties"] : Object.fromEntries(
      Object.entries(item["properties"]).map(([k, p]) => [k, p])
    ),
    measurements: !item["measurements"] ? item["measurements"] : Object.fromEntries(
      Object.entries(item["measurements"]).map(([k, p]) => [k, p])
    )
  };
}
function telemetryExceptionDetailsArraySerializer(result) {
  return result.map((item) => {
    return telemetryExceptionDetailsSerializer(item);
  });
}
function telemetryExceptionDetailsArrayDeserializer(result) {
  return result.map((item) => {
    return telemetryExceptionDetailsDeserializer(item);
  });
}
function telemetryExceptionDetailsSerializer(item) {
  return {
    id: item["id"],
    outerId: item["outerId"],
    typeName: item["typeName"],
    message: item["message"],
    hasFullStack: item["hasFullStack"],
    stack: item["stack"],
    parsedStack: !item["parsedStack"] ? item["parsedStack"] : stackFrameArraySerializer(item["parsedStack"])
  };
}
function telemetryExceptionDetailsDeserializer(item) {
  return {
    id: item["id"],
    outerId: item["outerId"],
    typeName: item["typeName"],
    message: item["message"],
    hasFullStack: item["hasFullStack"],
    stack: item["stack"],
    parsedStack: !item["parsedStack"] ? item["parsedStack"] : stackFrameArrayDeserializer(item["parsedStack"])
  };
}
function stackFrameArraySerializer(result) {
  return result.map((item) => {
    return stackFrameSerializer(item);
  });
}
function stackFrameArrayDeserializer(result) {
  return result.map((item) => {
    return stackFrameDeserializer(item);
  });
}
function stackFrameSerializer(item) {
  return {
    level: item["level"],
    method: item["method"],
    assembly: item["assembly"],
    fileName: item["fileName"],
    line: item["line"]
  };
}
function stackFrameDeserializer(item) {
  return {
    level: item["level"],
    method: item["method"],
    assembly: item["assembly"],
    fileName: item["fileName"],
    line: item["line"]
  };
}
var KnownSeverityLevel = /* @__PURE__ */ ((KnownSeverityLevel2) => {
  KnownSeverityLevel2["Verbose"] = "Verbose";
  KnownSeverityLevel2["Information"] = "Information";
  KnownSeverityLevel2["Warning"] = "Warning";
  KnownSeverityLevel2["Error"] = "Error";
  KnownSeverityLevel2["Critical"] = "Critical";
  return KnownSeverityLevel2;
})(KnownSeverityLevel || {});
function messageDataSerializer(item) {
  return {
    ...(0, import_serialize_record.serializeRecord)(item.additionalProperties ?? {}),
    ver: item["version"],
    message: item["message"],
    severityLevel: item["severityLevel"],
    properties: item["properties"],
    measurements: item["measurements"]
  };
}
function messageDataDeserializer(item) {
  return {
    additionalProperties: (0, import_serialize_record.serializeRecord)(item, [
      "version",
      "kind",
      "message",
      "severityLevel",
      "properties",
      "measurements"
    ]),
    version: item["ver"],
    kind: item["kind"],
    message: item["message"],
    severityLevel: item["severityLevel"],
    properties: !item["properties"] ? item["properties"] : Object.fromEntries(
      Object.entries(item["properties"]).map(([k, p]) => [k, p])
    ),
    measurements: !item["measurements"] ? item["measurements"] : Object.fromEntries(
      Object.entries(item["measurements"]).map(([k, p]) => [k, p])
    )
  };
}
function metricsDataSerializer(item) {
  return {
    ...(0, import_serialize_record.serializeRecord)(item.additionalProperties ?? {}),
    ver: item["version"],
    metrics: metricDataPointArraySerializer(item["metrics"]),
    properties: item["properties"]
  };
}
function metricsDataDeserializer(item) {
  return {
    additionalProperties: (0, import_serialize_record.serializeRecord)(item, ["version", "kind", "metrics", "properties"]),
    version: item["ver"],
    kind: item["kind"],
    metrics: metricDataPointArrayDeserializer(item["metrics"]),
    properties: !item["properties"] ? item["properties"] : Object.fromEntries(
      Object.entries(item["properties"]).map(([k, p]) => [k, p])
    )
  };
}
function metricDataPointArraySerializer(result) {
  return result.map((item) => {
    return metricDataPointSerializer(item);
  });
}
function metricDataPointArrayDeserializer(result) {
  return result.map((item) => {
    return metricDataPointDeserializer(item);
  });
}
function metricDataPointSerializer(item) {
  return {
    ns: item["namespace"],
    name: item["name"],
    kind: item["dataPointType"],
    value: item["value"],
    count: item["count"],
    min: item["min"],
    max: item["max"],
    stdDev: item["stdDev"]
  };
}
function metricDataPointDeserializer(item) {
  return {
    namespace: item["ns"],
    name: item["name"],
    dataPointType: item["kind"],
    value: item["value"],
    count: item["count"],
    min: item["min"],
    max: item["max"],
    stdDev: item["stdDev"]
  };
}
var KnownDataPointType = /* @__PURE__ */ ((KnownDataPointType2) => {
  KnownDataPointType2["Measurement"] = "Measurement";
  KnownDataPointType2["Aggregation"] = "Aggregation";
  return KnownDataPointType2;
})(KnownDataPointType || {});
function pageViewDataSerializer(item) {
  return {
    ...(0, import_serialize_record.serializeRecord)(item.additionalProperties ?? {}),
    ver: item["version"],
    id: item["id"],
    name: item["name"],
    url: item["url"],
    duration: item["duration"],
    referredUri: item["referredUri"],
    properties: item["properties"],
    measurements: item["measurements"]
  };
}
function pageViewDataDeserializer(item) {
  return {
    additionalProperties: (0, import_serialize_record.serializeRecord)(item, [
      "version",
      "kind",
      "id",
      "name",
      "url",
      "duration",
      "referredUri",
      "properties",
      "measurements"
    ]),
    version: item["ver"],
    kind: item["kind"],
    id: item["id"],
    name: item["name"],
    url: item["url"],
    duration: item["duration"],
    referredUri: item["referredUri"],
    properties: !item["properties"] ? item["properties"] : Object.fromEntries(
      Object.entries(item["properties"]).map(([k, p]) => [k, p])
    ),
    measurements: !item["measurements"] ? item["measurements"] : Object.fromEntries(
      Object.entries(item["measurements"]).map(([k, p]) => [k, p])
    )
  };
}
function pageViewPerfDataSerializer(item) {
  return {
    ...(0, import_serialize_record.serializeRecord)(item.additionalProperties ?? {}),
    ver: item["version"],
    id: item["id"],
    name: item["name"],
    url: item["url"],
    duration: item["duration"],
    perfTotal: item["perfTotal"],
    networkConnect: item["networkConnect"],
    sentRequest: item["sentRequest"],
    receivedResponse: item["receivedResponse"],
    domProcessing: item["domProcessing"],
    properties: item["properties"],
    measurements: item["measurements"]
  };
}
function pageViewPerfDataDeserializer(item) {
  return {
    additionalProperties: (0, import_serialize_record.serializeRecord)(item, [
      "version",
      "kind",
      "id",
      "name",
      "url",
      "duration",
      "perfTotal",
      "networkConnect",
      "sentRequest",
      "receivedResponse",
      "domProcessing",
      "properties",
      "measurements"
    ]),
    version: item["ver"],
    kind: item["kind"],
    id: item["id"],
    name: item["name"],
    url: item["url"],
    duration: item["duration"],
    perfTotal: item["perfTotal"],
    networkConnect: item["networkConnect"],
    sentRequest: item["sentRequest"],
    receivedResponse: item["receivedResponse"],
    domProcessing: item["domProcessing"],
    properties: !item["properties"] ? item["properties"] : Object.fromEntries(
      Object.entries(item["properties"]).map(([k, p]) => [k, p])
    ),
    measurements: !item["measurements"] ? item["measurements"] : Object.fromEntries(
      Object.entries(item["measurements"]).map(([k, p]) => [k, p])
    )
  };
}
function remoteDependencyDataSerializer(item) {
  return {
    ...(0, import_serialize_record.serializeRecord)(item.additionalProperties ?? {}),
    ver: item["version"],
    id: item["id"],
    name: item["name"],
    resultCode: item["resultCode"],
    data: item["data"],
    type: item["type"],
    target: item["target"],
    duration: item["duration"],
    success: item["success"],
    properties: item["properties"],
    measurements: item["measurements"]
  };
}
function remoteDependencyDataDeserializer(item) {
  return {
    additionalProperties: (0, import_serialize_record.serializeRecord)(item, [
      "version",
      "kind",
      "id",
      "name",
      "resultCode",
      "data",
      "type",
      "target",
      "duration",
      "success",
      "properties",
      "measurements"
    ]),
    version: item["ver"],
    kind: item["kind"],
    id: item["id"],
    name: item["name"],
    resultCode: item["resultCode"],
    data: item["data"],
    type: item["type"],
    target: item["target"],
    duration: item["duration"],
    success: item["success"],
    properties: !item["properties"] ? item["properties"] : Object.fromEntries(
      Object.entries(item["properties"]).map(([k, p]) => [k, p])
    ),
    measurements: !item["measurements"] ? item["measurements"] : Object.fromEntries(
      Object.entries(item["measurements"]).map(([k, p]) => [k, p])
    )
  };
}
function requestDataSerializer(item) {
  return {
    ...(0, import_serialize_record.serializeRecord)(item.additionalProperties ?? {}),
    ver: item["version"],
    id: item["id"],
    name: item["name"],
    duration: item["duration"],
    success: item["success"],
    responseCode: item["responseCode"],
    source: item["source"],
    url: item["url"],
    properties: item["properties"],
    measurements: item["measurements"]
  };
}
function requestDataDeserializer(item) {
  return {
    additionalProperties: (0, import_serialize_record.serializeRecord)(item, [
      "version",
      "kind",
      "id",
      "name",
      "duration",
      "success",
      "responseCode",
      "source",
      "url",
      "properties",
      "measurements"
    ]),
    version: item["ver"],
    kind: item["kind"],
    id: item["id"],
    name: item["name"],
    duration: item["duration"],
    success: item["success"],
    responseCode: item["responseCode"],
    source: item["source"],
    url: item["url"],
    properties: !item["properties"] ? item["properties"] : Object.fromEntries(
      Object.entries(item["properties"]).map(([k, p]) => [k, p])
    ),
    measurements: !item["measurements"] ? item["measurements"] : Object.fromEntries(
      Object.entries(item["measurements"]).map(([k, p]) => [k, p])
    )
  };
}
function trackResponseDeserializer(item) {
  return {
    itemsReceived: item["itemsReceived"],
    itemsAccepted: item["itemsAccepted"],
    errors: !item["errors"] ? item["errors"] : telemetryErrorDetailsArrayDeserializer(item["errors"])
  };
}
function telemetryErrorDetailsArrayDeserializer(result) {
  return result.map((item) => {
    return telemetryErrorDetailsDeserializer(item);
  });
}
function telemetryErrorDetailsDeserializer(item) {
  return {
    index: item["index"],
    statusCode: item["statusCode"],
    message: item["message"]
  };
}
var KnownContextTagKeys = /* @__PURE__ */ ((KnownContextTagKeys2) => {
  KnownContextTagKeys2["AiApplicationVer"] = "ai.application.ver";
  KnownContextTagKeys2["AiDeviceId"] = "ai.device.id";
  KnownContextTagKeys2["AiDeviceLocale"] = "ai.device.locale";
  KnownContextTagKeys2["AiDeviceModel"] = "ai.device.model";
  KnownContextTagKeys2["AiDeviceOemName"] = "ai.device.oemName";
  KnownContextTagKeys2["AiDeviceOSVersion"] = "ai.device.osVersion";
  KnownContextTagKeys2["AiDeviceType"] = "ai.device.type";
  KnownContextTagKeys2["AiLocationIp"] = "ai.location.ip";
  KnownContextTagKeys2["AiLocationCountry"] = "ai.location.country";
  KnownContextTagKeys2["AiLocationProvince"] = "ai.location.province";
  KnownContextTagKeys2["AiLocationCity"] = "ai.location.city";
  KnownContextTagKeys2["AiOperationId"] = "ai.operation.id";
  KnownContextTagKeys2["AiOperationName"] = "ai.operation.name";
  KnownContextTagKeys2["AiOperationParentId"] = "ai.operation.parentId";
  KnownContextTagKeys2["AiOperationSyntheticSource"] = "ai.operation.syntheticSource";
  KnownContextTagKeys2["AiOperationCorrelationVector"] = "ai.operation.correlationVector";
  KnownContextTagKeys2["AiSessionId"] = "ai.session.id";
  KnownContextTagKeys2["AiSessionIsFirst"] = "ai.session.isFirst";
  KnownContextTagKeys2["AiUserAccountId"] = "ai.user.accountId";
  KnownContextTagKeys2["AiUserId"] = "ai.user.id";
  KnownContextTagKeys2["AiUserAuthUserId"] = "ai.user.authUserId";
  KnownContextTagKeys2["AiCloudRole"] = "ai.cloud.role";
  KnownContextTagKeys2["AiCloudRoleVer"] = "ai.cloud.roleVer";
  KnownContextTagKeys2["AiCloudRoleInstance"] = "ai.cloud.roleInstance";
  KnownContextTagKeys2["AiCloudLocation"] = "ai.cloud.location";
  KnownContextTagKeys2["AiInternalSdkVersion"] = "ai.internal.sdkVersion";
  KnownContextTagKeys2["AiInternalAgentVersion"] = "ai.internal.agentVersion";
  KnownContextTagKeys2["AiInternalNodeName"] = "ai.internal.nodeName";
  return KnownContextTagKeys2;
})(KnownContextTagKeys || {});
function telemetryItemArraySerializer(result) {
  return result.map((item) => {
    return telemetryItemSerializer(item);
  });
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  KnownContextTagKeys,
  KnownDataPointType,
  KnownMonitorDomainKind,
  KnownSeverityLevel,
  availabilityDataDeserializer,
  availabilityDataSerializer,
  messageDataDeserializer,
  messageDataSerializer,
  metricDataPointArrayDeserializer,
  metricDataPointArraySerializer,
  metricDataPointDeserializer,
  metricDataPointSerializer,
  metricsDataDeserializer,
  metricsDataSerializer,
  monitorBaseSerializer,
  monitorDomainDeserializer,
  monitorDomainSerializer,
  monitorDomainUnionDeserializer,
  monitorDomainUnionSerializer,
  pageViewDataDeserializer,
  pageViewDataSerializer,
  pageViewPerfDataDeserializer,
  pageViewPerfDataSerializer,
  remoteDependencyDataDeserializer,
  remoteDependencyDataSerializer,
  requestDataDeserializer,
  requestDataSerializer,
  stackFrameArrayDeserializer,
  stackFrameArraySerializer,
  stackFrameDeserializer,
  stackFrameSerializer,
  telemetryErrorDetailsArrayDeserializer,
  telemetryErrorDetailsDeserializer,
  telemetryEventDataDeserializer,
  telemetryEventDataSerializer,
  telemetryExceptionDataDeserializer,
  telemetryExceptionDataSerializer,
  telemetryExceptionDetailsArrayDeserializer,
  telemetryExceptionDetailsArraySerializer,
  telemetryExceptionDetailsDeserializer,
  telemetryExceptionDetailsSerializer,
  telemetryItemArraySerializer,
  telemetryItemSerializer,
  trackResponseDeserializer
});
