// @azure/monitor-opentelemetry-exporter/dist/commonjs/sampling/rateLimitedSampler.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var rateLimitedSampler_exports = {};
__export(rateLimitedSampler_exports, {
  RateLimitedSampler: () => RateLimitedSampler
});
module.exports = __toCommonJS(rateLimitedSampler_exports);
var import_sdk_trace_base = require("@opentelemetry/sdk-trace-base");
var import_samplingUtils = require("./samplingUtils.js");
class RateLimitedSampler {
  nanoTimeSupplier;
  inverseAdaptationTimeNanos;
  targetSpansPerNanosecondLimit;
  state;
  roundToNearest;
  tracesPerSecond;
  /**
   * Initializes a new instance of the RateLimitedSampler class.
   * @param tracesPerSecond - The maximum number of traces to sample per second.
   * @throws Error if tracesPerSecond is negative.
   */
  constructor(tracesPerSecond) {
    this.tracesPerSecond = tracesPerSecond;
    if (this.tracesPerSecond < 0) {
      throw new Error("Limit for sampled traces per second must be nonnegative");
    }
    const adaptationTimeSeconds = 0.1;
    this.nanoTimeSupplier = () => Number(process.hrtime.bigint());
    this.inverseAdaptationTimeNanos = 1e-9 / adaptationTimeSeconds;
    this.targetSpansPerNanosecondLimit = 1e-9 * this.tracesPerSecond;
    const now = this.nanoTimeSupplier();
    this.state = {
      effectiveWindowCount: 0,
      effectiveWindowNanos: 0,
      lastNanoTime: now
    };
    this.roundToNearest = true;
  }
  /**
   * Updates the state of the sampler based on the current time.
   * This method calculates the effective window count and nanos based on the time elapsed since the last sample.
   * @param oldState - The previous state of the sampler.
   * @param currentNanoTime - The current time in nanoseconds.
   * @returns The updated state of the sampler.
   */
  updateState(oldState, currentNanoTime) {
    if (currentNanoTime <= oldState.lastNanoTime) {
      return {
        effectiveWindowCount: oldState.effectiveWindowCount + 1,
        effectiveWindowNanos: oldState.effectiveWindowNanos,
        lastNanoTime: oldState.lastNanoTime
      };
    }
    const nanoTimeDelta = currentNanoTime - oldState.lastNanoTime;
    const decayFactor = Math.exp(-nanoTimeDelta * this.inverseAdaptationTimeNanos);
    const currentEffectiveWindowCount = oldState.effectiveWindowCount * decayFactor + 1;
    const currentEffectiveWindowNanos = oldState.effectiveWindowNanos * decayFactor + nanoTimeDelta;
    return {
      effectiveWindowCount: currentEffectiveWindowCount,
      effectiveWindowNanos: currentEffectiveWindowNanos,
      lastNanoTime: currentNanoTime
    };
  }
  /**
   * Gets the current sample rate based on the effective window count and nanos.
   * This method calculates the sampling probability and returns it as a percentage.
   * If `roundToNearest` is true, it rounds down the sampling percentage to the nearest whole number.
   * @returns The current sample rate as a percentage.
   */
  getSampleRate() {
    const currentNanoTime = this.nanoTimeSupplier();
    this.state = this.updateState(this.state, currentNanoTime);
    const samplingProbability = this.state.effectiveWindowNanos * this.targetSpansPerNanosecondLimit / this.state.effectiveWindowCount;
    let samplingPercentage = 100 * Math.min(samplingProbability, 1);
    if (this.roundToNearest) {
      samplingPercentage = (0, import_samplingUtils.roundDownToNearest)(samplingPercentage);
    }
    return samplingPercentage;
  }
  /**
   * Checks whether span needs to be created and tracked.
   *
   * @param context - Parent Context which may contain a span.
   * @param traceId - traceId of the span to be created. It can be different from the
   *     traceId in the {@link SpanContext}. Typically in situations when the
   *     span to be created starts a new trace.
   * @param spanName - Name of the span to be created.
   * @param spanKind - Kind of the span to be created.
   * @param attributes - Initial set of SpanAttributes for the Span being constructed.
   * @param links - Collection of links that will be associated with the Span to
   *     be created. Typically useful for batch operations.
   * @returns a {@link SamplingResult}.
   */
  shouldSample(context, traceId, spanName, spanKind, attributes, links) {
    const sampleRate = this.getSampleRate();
    return (0, import_samplingUtils.shouldSample)(sampleRate, context, traceId, attributes) ? { decision: import_sdk_trace_base.SamplingDecision.RECORD_AND_SAMPLED, attributes } : { decision: import_sdk_trace_base.SamplingDecision.NOT_RECORD, attributes };
  }
  /**
   * Return Sampler description
   */
  toString() {
    return `RateLimitedSampler{${this.tracesPerSecond}}`;
  }
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  RateLimitedSampler
});
