// @azure/monitor-opentelemetry/dist/esm/generated/models/models.js
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.
export function monitoringDataPointSerializer(item) {
    return {
        Version: item["version"],
        InvariantVersion: item["invariantVersion"],
        Instance: item["instance"],
        RoleName: item["roleName"],
        MachineName: item["machineName"],
        StreamId: item["streamId"],
        Timestamp: !item["timestamp"] ? item["timestamp"] : item["timestamp"].toISOString(),
        TransmissionTime: !item["transmissionTime"]
            ? item["transmissionTime"]
            : item["transmissionTime"].toISOString(),
        IsWebApp: item["isWebApp"],
        PerformanceCollectionSupported: item["performanceCollectionSupported"],
        Metrics: !item["metrics"] ? item["metrics"] : metricPointArraySerializer(item["metrics"]),
        Documents: !item["documents"]
            ? item["documents"]
            : documentIngressUnionArraySerializer(item["documents"]),
        TopCpuProcesses: !item["topCpuProcesses"]
            ? item["topCpuProcesses"]
            : processCpuDataArraySerializer(item["topCpuProcesses"]),
        CollectionConfigurationErrors: !item["collectionConfigurationErrors"]
            ? item["collectionConfigurationErrors"]
            : collectionConfigurationErrorArraySerializer(item["collectionConfigurationErrors"]),
    };
}
export function metricPointArraySerializer(result) {
    return result.map((item) => {
        return metricPointSerializer(item);
    });
}
export function metricPointSerializer(item) {
    return { Name: item["name"], Value: item["value"], Weight: item["weight"] };
}
export function documentIngressUnionArraySerializer(result) {
    return result.map((item) => {
        return documentIngressUnionSerializer(item);
    });
}
export function documentIngressSerializer(item) {
    return {
        DocumentType: item["documentType"],
        DocumentStreamIds: !item["documentStreamIds"]
            ? item["documentStreamIds"]
            : item["documentStreamIds"].map((p) => {
                return p;
            }),
        Properties: !item["properties"]
            ? item["properties"]
            : keyValuePairStringStringArraySerializer(item["properties"]),
    };
}
export function documentIngressUnionSerializer(item) {
    switch (item.documentType) {
        case "Request":
            return requestSerializer(item);
        case "RemoteDependency":
            return remoteDependencySerializer(item);
        case "Exception":
            return exceptionSerializer(item);
        case "Event":
            return eventSerializer(item);
        case "Trace":
            return traceSerializer(item);
        default:
            return documentIngressSerializer(item);
    }
}
export function keyValuePairStringStringArraySerializer(result) {
    return result.map((item) => {
        return keyValuePairStringStringSerializer(item);
    });
}
export function keyValuePairStringStringSerializer(item) {
    return { key: item["key"], value: item["value"] };
}
export function requestSerializer(item) {
    return {
        DocumentType: item["documentType"],
        DocumentStreamIds: !item["documentStreamIds"]
            ? item["documentStreamIds"]
            : item["documentStreamIds"].map((p) => {
                return p;
            }),
        Properties: !item["properties"]
            ? item["properties"]
            : keyValuePairStringStringArraySerializer(item["properties"]),
        Name: item["name"],
        Url: item["url"],
        ResponseCode: item["responseCode"],
        Duration: item["duration"],
    };
}
export function remoteDependencySerializer(item) {
    return {
        DocumentType: item["documentType"],
        DocumentStreamIds: !item["documentStreamIds"]
            ? item["documentStreamIds"]
            : item["documentStreamIds"].map((p) => {
                return p;
            }),
        Properties: !item["properties"]
            ? item["properties"]
            : keyValuePairStringStringArraySerializer(item["properties"]),
        Name: item["name"],
        CommandName: item["commandName"],
        ResultCode: item["resultCode"],
        Duration: item["duration"],
    };
}
export function exceptionSerializer(item) {
    return {
        DocumentType: item["documentType"],
        DocumentStreamIds: !item["documentStreamIds"]
            ? item["documentStreamIds"]
            : item["documentStreamIds"].map((p) => {
                return p;
            }),
        Properties: !item["properties"]
            ? item["properties"]
            : keyValuePairStringStringArraySerializer(item["properties"]),
        ExceptionType: item["exceptionType"],
        ExceptionMessage: item["exceptionMessage"],
    };
}
export function eventSerializer(item) {
    return {
        DocumentType: item["documentType"],
        DocumentStreamIds: !item["documentStreamIds"]
            ? item["documentStreamIds"]
            : item["documentStreamIds"].map((p) => {
                return p;
            }),
        Properties: !item["properties"]
            ? item["properties"]
            : keyValuePairStringStringArraySerializer(item["properties"]),
        Name: item["name"],
    };
}
export function traceSerializer(item) {
    return {
        DocumentType: item["documentType"],
        DocumentStreamIds: !item["documentStreamIds"]
            ? item["documentStreamIds"]
            : item["documentStreamIds"].map((p) => {
                return p;
            }),
        Properties: !item["properties"]
            ? item["properties"]
            : keyValuePairStringStringArraySerializer(item["properties"]),
        Message: item["message"],
    };
}
export function processCpuDataArraySerializer(result) {
    return result.map((item) => {
        return processCpuDataSerializer(item);
    });
}
export function processCpuDataSerializer(item) {
    return { ProcessName: item["processName"], CpuPercentage: item["cpuPercentage"] };
}
export function collectionConfigurationErrorArraySerializer(result) {
    return result.map((item) => {
        return collectionConfigurationErrorSerializer(item);
    });
}
export function collectionConfigurationErrorSerializer(item) {
    return {
        CollectionConfigurationErrorType: item["collectionConfigurationErrorType"],
        Message: item["message"],
        FullException: item["fullException"],
        Data: keyValuePairStringStringArraySerializer(item["data"]),
    };
}
export function collectionConfigurationInfoSerializer(item) {
    return {
        ETag: item["eTag"],
        Metrics: derivedMetricInfoArraySerializer(item["metrics"]),
        DocumentStreams: documentStreamInfoArraySerializer(item["documentStreams"]),
        QuotaInfo: !item["quotaInfo"]
            ? item["quotaInfo"]
            : quotaConfigurationInfoSerializer(item["quotaInfo"]),
    };
}
export function collectionConfigurationInfoDeserializer(item) {
    return {
        eTag: item["ETag"],
        metrics: derivedMetricInfoArrayDeserializer(item["Metrics"]),
        documentStreams: documentStreamInfoArrayDeserializer(item["DocumentStreams"]),
        quotaInfo: !item["QuotaInfo"]
            ? item["QuotaInfo"]
            : quotaConfigurationInfoDeserializer(item["QuotaInfo"]),
    };
}
export function derivedMetricInfoArraySerializer(result) {
    return result.map((item) => {
        return derivedMetricInfoSerializer(item);
    });
}
export function derivedMetricInfoArrayDeserializer(result) {
    return result.map((item) => {
        return derivedMetricInfoDeserializer(item);
    });
}
export function derivedMetricInfoSerializer(item) {
    return {
        Id: item["id"],
        TelemetryType: item["telemetryType"],
        FilterGroups: filterConjunctionGroupInfoArraySerializer(item["filterGroups"]),
        Projection: item["projection"],
        Aggregation: item["aggregation"],
        BackEndAggregation: item["backEndAggregation"],
    };
}
export function derivedMetricInfoDeserializer(item) {
    return {
        id: item["Id"],
        telemetryType: item["TelemetryType"],
        filterGroups: filterConjunctionGroupInfoArrayDeserializer(item["FilterGroups"]),
        projection: item["Projection"],
        aggregation: item["Aggregation"],
        backEndAggregation: item["BackEndAggregation"],
    };
}
export function filterConjunctionGroupInfoArraySerializer(result) {
    return result.map((item) => {
        return filterConjunctionGroupInfoSerializer(item);
    });
}
export function filterConjunctionGroupInfoArrayDeserializer(result) {
    return result.map((item) => {
        return filterConjunctionGroupInfoDeserializer(item);
    });
}
export function filterConjunctionGroupInfoSerializer(item) {
    return { Filters: filterInfoArraySerializer(item["filters"]) };
}
export function filterConjunctionGroupInfoDeserializer(item) {
    return {
        filters: filterInfoArrayDeserializer(item["Filters"]),
    };
}
export function filterInfoArraySerializer(result) {
    return result.map((item) => {
        return filterInfoSerializer(item);
    });
}
export function filterInfoArrayDeserializer(result) {
    return result.map((item) => {
        return filterInfoDeserializer(item);
    });
}
export function filterInfoSerializer(item) {
    return {
        FieldName: item["fieldName"],
        Predicate: item["predicate"],
        Comparand: item["comparand"],
    };
}
export function filterInfoDeserializer(item) {
    return {
        fieldName: item["FieldName"],
        predicate: item["Predicate"],
        comparand: item["Comparand"],
    };
}
export function documentStreamInfoArraySerializer(result) {
    return result.map((item) => {
        return documentStreamInfoSerializer(item);
    });
}
export function documentStreamInfoArrayDeserializer(result) {
    return result.map((item) => {
        return documentStreamInfoDeserializer(item);
    });
}
export function documentStreamInfoSerializer(item) {
    return {
        Id: item["id"],
        DocumentFilterGroups: documentFilterConjunctionGroupInfoArraySerializer(item["documentFilterGroups"]),
    };
}
export function documentStreamInfoDeserializer(item) {
    return {
        id: item["Id"],
        documentFilterGroups: documentFilterConjunctionGroupInfoArrayDeserializer(item["DocumentFilterGroups"]),
    };
}
export function documentFilterConjunctionGroupInfoArraySerializer(result) {
    return result.map((item) => {
        return documentFilterConjunctionGroupInfoSerializer(item);
    });
}
export function documentFilterConjunctionGroupInfoArrayDeserializer(result) {
    return result.map((item) => {
        return documentFilterConjunctionGroupInfoDeserializer(item);
    });
}
export function documentFilterConjunctionGroupInfoSerializer(item) {
    return {
        TelemetryType: item["telemetryType"],
        Filters: filterConjunctionGroupInfoSerializer(item["filters"]),
    };
}
export function documentFilterConjunctionGroupInfoDeserializer(item) {
    return {
        telemetryType: item["TelemetryType"],
        filters: filterConjunctionGroupInfoDeserializer(item["Filters"]),
    };
}
export function quotaConfigurationInfoSerializer(item) {
    return {
        InitialQuota: item["initialQuota"],
        MaxQuota: item["maxQuota"],
        QuotaAccrualRatePerSec: item["quotaAccrualRatePerSec"],
    };
}
export function quotaConfigurationInfoDeserializer(item) {
    return {
        initialQuota: item["InitialQuota"],
        maxQuota: item["MaxQuota"],
        quotaAccrualRatePerSec: item["QuotaAccrualRatePerSec"],
    };
}
export function serviceErrorDeserializer(item) {
    return {
        requestId: item["RequestId"],
        responseDateTime: item["ResponseDateTime"],
        code: item["Code"],
        message: item["Message"],
        exception: item["Exception"],
    };
}
/** Live Metrics service versions. */
export var KnownVersions;
(function (KnownVersions) {
    /** The 2024-04-01-preview version of the Live Metrics service. */
    KnownVersions["V20240401Preview"] = "2024-04-01-preview";
})(KnownVersions || (KnownVersions = {}));
export function monitoringDataPointArraySerializer(result) {
    return result.map((item) => {
        return monitoringDataPointSerializer(item);
    });
}
