// @azure/monitor-opentelemetry-exporter/dist/commonjs/platform/nodejs/baseSender.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var baseSender_exports = {};
__export(baseSender_exports, {
  BaseSender: () => BaseSender
});
module.exports = __toCommonJS(baseSender_exports);
var import_api = require("@opentelemetry/api");
var import_types = require("../../export/statsbeat/types.js");
var import_persist = require("./persist/index.js");
var import_core = require("@opentelemetry/core");
var import_networkStatsbeatMetrics = require("../../export/statsbeat/networkStatsbeatMetrics.js");
var import_longIntervalStatsbeatMetrics = require("../../export/statsbeat/longIntervalStatsbeatMetrics.js");
var import_types2 = require("../../export/statsbeat/types.js");
var import_breezeUtils = require("../../utils/breezeUtils.js");
var import_Constants = require("../../Declarations/Constants.js");
const DEFAULT_BATCH_SEND_RETRY_INTERVAL_MS = 6e4;
class BaseSender {
  persister;
  numConsecutiveRedirects;
  retryTimer;
  retryTimerDeadlineMs = 0;
  networkStatsbeatMetrics;
  customerSDKStatsMetrics;
  longIntervalStatsbeatMetrics;
  statsbeatFailureCount = 0;
  batchSendRetryIntervalMs = DEFAULT_BATCH_SEND_RETRY_INTERVAL_MS;
  isStatsbeatSender;
  disableOfflineStorage;
  constructor(options) {
    this.numConsecutiveRedirects = 0;
    this.disableOfflineStorage = options.exporterOptions.disableOfflineStorage || false;
    if (options.trackStatsbeat) {
      this.networkStatsbeatMetrics = import_networkStatsbeatMetrics.NetworkStatsbeatMetrics.getInstance({
        instrumentationKey: options.instrumentationKey,
        endpointUrl: options.endpointUrl,
        disableOfflineStorage: this.disableOfflineStorage
      });
      this.longIntervalStatsbeatMetrics = import_longIntervalStatsbeatMetrics.LongIntervalStatsbeatMetrics.getInstance({
        instrumentationKey: options.instrumentationKey,
        endpointUrl: options.endpointUrl,
        disableOfflineStorage: this.disableOfflineStorage
      });
      if (!process.env[import_Constants.ENV_DISABLE_SDKSTATS]) {
        let exportInterval;
        if (process.env[import_Constants.ENV_APPLICATIONINSIGHTS_SDKSTATS_EXPORT_INTERVAL]) {
          const envValue = process.env[import_Constants.ENV_APPLICATIONINSIGHTS_SDKSTATS_EXPORT_INTERVAL];
          const exportIntervalSeconds = parseInt(envValue, 10);
          if (!isNaN(exportIntervalSeconds) && exportIntervalSeconds > 0) {
            exportInterval = exportIntervalSeconds * 1e3;
          } else {
            import_api.diag.warn(
              `Invalid value for APPLICATIONINSIGHTS_SDKSTATS_EXPORT_INTERVAL environment variable: '${envValue}'. Expected a positive number (seconds). Using default export interval.`
            );
          }
        }
        if (!this.customerSDKStatsMetrics) {
          import("../../export/statsbeat/customerSDKStats.js").then(
            (module2) => module2.CustomerSDKStatsMetrics.getInstance({
              instrumentationKey: options.instrumentationKey,
              endpointUrl: options.endpointUrl,
              disableOfflineStorage: this.disableOfflineStorage,
              networkCollectionInterval: exportInterval
            })
          ).then((metrics) => {
            this.customerSDKStatsMetrics = metrics;
            return;
          }).catch((error) => {
            import_api.diag.warn("Failed to initialize customer SDK stats metrics:", error);
          });
        }
      }
    }
    this.persister = new import_persist.FileSystemPersist(
      options.instrumentationKey,
      options.exporterOptions,
      this.customerSDKStatsMetrics
    );
    this.retryTimer = null;
    this.isStatsbeatSender = options.isStatsbeatSender || false;
    if (!this.disableOfflineStorage) {
      this.sendAllPersistedFiles();
    }
  }
  /**
   * Export envelopes
   */
  async exportEnvelopes(envelopes) {
    import_api.diag.info(`Exporting ${envelopes.length} envelope(s)`);
    if (envelopes.length < 1) {
      return { code: import_core.ExportResultCode.SUCCESS };
    }
    try {
      const startTime = (/* @__PURE__ */ new Date()).getTime();
      const { result, statusCode, retryAfterMs } = await this.send(envelopes);
      const endTime = (/* @__PURE__ */ new Date()).getTime();
      const duration = endTime - startTime;
      this.numConsecutiveRedirects = 0;
      if (statusCode === 200) {
        this.scheduleRetryTimer(retryAfterMs);
        if (!this.isStatsbeatSender) {
          this.networkStatsbeatMetrics?.countSuccess(duration);
          this.customerSDKStatsMetrics?.countSuccessfulItems(envelopes);
        }
        return { code: import_core.ExportResultCode.SUCCESS };
      } else if (statusCode && (0, import_breezeUtils.isRetriable)(statusCode)) {
        if (statusCode === 429) {
          if (!this.isStatsbeatSender) {
            this.networkStatsbeatMetrics?.countThrottle(statusCode);
            this.customerSDKStatsMetrics?.countRetryItems(envelopes, statusCode);
          }
          this.scheduleRetryTimer(retryAfterMs);
          return await this.persist(envelopes);
        }
        if (result) {
          import_api.diag.info(result);
          const { breezeResponse, retriableEnvelopes: filteredEnvelopes } = this.parseRetriableEnvelopes(envelopes, result);
          const successfulEnvelopes = [...envelopes];
          if (breezeResponse.itemsAccepted > 0 && statusCode === 206 && !this.isStatsbeatSender) {
            this.networkStatsbeatMetrics?.countSuccess(duration);
          }
          if (breezeResponse.errors) {
            breezeResponse.errors.forEach((error) => {
              successfulEnvelopes[error.index] = void 0;
            });
          }
          if (breezeResponse.itemsAccepted > 0) {
            if (!this.isStatsbeatSender) {
              this.networkStatsbeatMetrics?.countSuccess(duration);
              this.customerSDKStatsMetrics?.countSuccessfulItems(envelopes);
            }
          }
          if (filteredEnvelopes.length > 0) {
            if (!this.isStatsbeatSender) {
              this.networkStatsbeatMetrics?.countRetry(statusCode);
              this.customerSDKStatsMetrics?.countRetryItems(envelopes, statusCode);
            }
            this.scheduleRetryTimer(retryAfterMs);
            return await this.persist(filteredEnvelopes);
          }
          if (!this.isStatsbeatSender) {
            this.networkStatsbeatMetrics?.countFailure(duration, statusCode);
            const filteredSuccessfulEnvelopes = successfulEnvelopes.filter(Boolean);
            this.customerSDKStatsMetrics?.countDroppedItems(
              filteredSuccessfulEnvelopes,
              statusCode
            );
          }
          return this.buildExportResult({
            code: import_core.ExportResultCode.FAILED
          });
        } else {
          if (!this.isStatsbeatSender) {
            this.networkStatsbeatMetrics?.countRetry(statusCode);
            this.customerSDKStatsMetrics?.countRetryItems(envelopes, statusCode);
          }
          this.scheduleRetryTimer(retryAfterMs);
          return await this.persist(envelopes);
        }
      } else {
        if (this.networkStatsbeatMetrics && !this.isStatsbeatSender) {
          if (statusCode) {
            this.networkStatsbeatMetrics.countFailure(duration, statusCode);
            this.customerSDKStatsMetrics?.countDroppedItems(envelopes, statusCode);
          }
        } else {
          this.incrementStatsbeatFailure();
          this.customerSDKStatsMetrics?.countDroppedItems(envelopes, import_types2.DropCode.CLIENT_EXCEPTION);
        }
        return this.buildExportResult({
          code: import_core.ExportResultCode.FAILED
        });
      }
    } catch (error) {
      const restError = error;
      if (restError.statusCode && (restError.statusCode === 307 || // Temporary redirect
      restError.statusCode === 308)) {
        this.numConsecutiveRedirects++;
        if (this.numConsecutiveRedirects < 10) {
          const location = this.getLocationFromHeaders(restError.response?.headers);
          if (location) {
            this.handlePermanentRedirect(location);
            return this.exportEnvelopes(envelopes);
          }
        } else {
          const redirectError = new Error("Circular redirect");
          if (!this.isStatsbeatSender) {
            this.networkStatsbeatMetrics?.countException(redirectError);
            this.customerSDKStatsMetrics?.countDroppedItems(
              envelopes,
              import_types2.DropCode.CLIENT_EXCEPTION,
              redirectError.message,
              import_types.ExceptionType.CLIENT_EXCEPTION
            );
          }
          return this.buildExportResult({ code: import_core.ExportResultCode.FAILED, error: redirectError });
        }
      } else if (restError.statusCode && (0, import_breezeUtils.isRetriable)(restError.statusCode) && !this.isStatsbeatSender) {
        this.networkStatsbeatMetrics?.countRetry(restError.statusCode);
        this.customerSDKStatsMetrics?.countRetryItems(envelopes, restError.statusCode);
        return this.persist(envelopes);
      } else if (restError.statusCode === 400 && restError.message.includes("Invalid instrumentation key")) {
        this.shutdownStatsbeat();
        return { code: import_core.ExportResultCode.SUCCESS };
      } else if (restError.statusCode && this.isStatsbeatSender && (0, import_types2.isStatsbeatShutdownStatus)(restError.statusCode)) {
        this.incrementStatsbeatFailure();
        return { code: import_core.ExportResultCode.SUCCESS };
      }
      if (this.isRetriableRestError(restError) && !this.isStatsbeatSender) {
        if (this.customerSDKStatsMetrics?.isTimeoutError(restError) && !this.isStatsbeatSender) {
          this.customerSDKStatsMetrics?.countRetryItems(
            envelopes,
            import_types2.RetryCode.CLIENT_TIMEOUT,
            "timeout_exception",
            import_types.ExceptionType.TIMEOUT_EXCEPTION
          );
          import_api.diag.error("Request timed out. Error message:", restError.message);
        } else if (restError.statusCode) {
          this.networkStatsbeatMetrics?.countRetry(restError.statusCode);
          this.customerSDKStatsMetrics?.countRetryItems(envelopes, restError.statusCode);
        }
        import_api.diag.error(
          "Retrying due to transient client side error. Error message:",
          restError.message
        );
        return this.persist(envelopes);
      }
      if (!this.isStatsbeatSender) {
        this.networkStatsbeatMetrics?.countException(restError);
        this.customerSDKStatsMetrics?.countDroppedItems(
          envelopes,
          import_types2.DropCode.CLIENT_EXCEPTION,
          restError.message
        );
        import_api.diag.error(
          "Envelopes could not be exported and are not retriable. Error message:",
          restError.message
        );
      }
      return this.buildExportResult({ code: import_core.ExportResultCode.FAILED, error: restError });
    }
  }
  /**
   * Persist envelopes to disk
   */
  async persist(envelopes) {
    try {
      const success = await this.persister.push(envelopes);
      return success ? { code: import_core.ExportResultCode.SUCCESS } : this.buildExportResult({
        code: import_core.ExportResultCode.FAILED,
        error: new Error("Failed to persist envelope in disk.")
      });
    } catch (ex) {
      if (!this.isStatsbeatSender) {
        this.networkStatsbeatMetrics?.countWriteFailure();
        if (this.disableOfflineStorage && envelopes) {
          this.customerSDKStatsMetrics?.countDroppedItems(
            envelopes,
            import_types2.DropCode.CLIENT_STORAGE_DISABLED
          );
        }
      }
      return this.buildExportResult({ code: import_core.ExportResultCode.FAILED, error: ex });
    }
  }
  /**
   * Disable collection of statsbeat metrics after max failures
   */
  incrementStatsbeatFailure() {
    this.statsbeatFailureCount++;
    if (this.statsbeatFailureCount > import_types2.MAX_STATSBEAT_FAILURES) {
      this.shutdownStatsbeat();
    }
  }
  /**
   * Shutdown statsbeat metrics
   */
  shutdownStatsbeat() {
    if (this.networkStatsbeatMetrics) {
      this.networkStatsbeatMetrics.shutdown();
    }
    if (this.longIntervalStatsbeatMetrics) {
      this.longIntervalStatsbeatMetrics?.shutdown();
    }
    if (this.customerSDKStatsMetrics) {
      this.customerSDKStatsMetrics.shutdown();
    }
    this.statsbeatFailureCount = 0;
  }
  async sendFirstPersistedFile() {
    const envelopes = await this.persister.shift();
    try {
      if (envelopes) {
        await this.exportEnvelopes(envelopes);
      }
    } catch (err) {
      if (!this.isStatsbeatSender) {
        this.networkStatsbeatMetrics?.countReadFailure();
      }
      import_api.diag.warn(`Failed to fetch persisted file`, err);
    }
  }
  async sendAllPersistedFiles() {
    try {
      await this.persister.cleanExpiredFiles();
      let envelopes = await this.persister.shift();
      while (envelopes) {
        const result = await this.exportEnvelopes(envelopes);
        if (result.code === import_core.ExportResultCode.FAILED) {
          import_api.diag.warn(`Failed to send persisted file during startup, will retry later`);
          break;
        }
        envelopes = await this.persister.shift();
      }
    } catch (err) {
      import_api.diag.warn(`Failed to read persisted files during startup`, err);
    }
  }
  /**
   * Parse a Breeze response and extract envelopes whose errors have retriable status codes.
   */
  parseRetriableEnvelopes(envelopes, responseBody) {
    const breezeResponse = JSON.parse(responseBody);
    const retriableEnvelopes = [];
    if (breezeResponse.errors) {
      for (const error of breezeResponse.errors) {
        if (error.statusCode && (0, import_breezeUtils.isRetriable)(error.statusCode) && !(0, import_breezeUtils.isSamplingRejection)(error)) {
          retriableEnvelopes.push(envelopes[error.index]);
        }
      }
    }
    return { breezeResponse, retriableEnvelopes };
  }
  scheduleRetryTimer(retryAfterMs) {
    const delay = retryAfterMs ?? this.batchSendRetryIntervalMs;
    const newDeadline = Date.now() + delay;
    if (this.retryTimer && retryAfterMs !== void 0 && newDeadline > this.retryTimerDeadlineMs) {
      clearTimeout(this.retryTimer);
      this.retryTimer = null;
    }
    if (!this.retryTimer) {
      const adjustedDelay = Math.max(newDeadline - Date.now(), 0);
      this.retryTimerDeadlineMs = newDeadline;
      this.retryTimer = setTimeout(() => {
        this.retryTimer = null;
        this.sendFirstPersistedFile();
      }, adjustedDelay);
      this.retryTimer.unref();
    }
  }
  isRetriableRestError(error) {
    const restErrorTypes = Object.values(import_Constants.RetriableRestErrorTypes);
    if (error && error.code && restErrorTypes.includes(error.code)) {
      return true;
    }
    return false;
  }
  // Normalize location extraction for redirects; mirrors core HttpHeaders behavior
  getLocationFromHeaders(headers) {
    return headers?.get("location") ?? headers?.toJSON?.().location;
  }
  // Silence noisy failures from statsbeat OTel metric readers unless logging is explicitly enabled
  buildExportResult(result) {
    const shouldSurfaceStatsbeatFailures = !!process.env[import_Constants.ENV_APPLICATIONINSIGHTS_SDK_STATS_LOGGING];
    if (this.isStatsbeatSender && result.code === import_core.ExportResultCode.FAILED) {
      return shouldSurfaceStatsbeatFailures ? result : { code: import_core.ExportResultCode.SUCCESS };
    }
    return result;
  }
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  BaseSender
});
