// @azure/monitor-opentelemetry-exporter/dist/commonjs/platform/nodejs/httpSender.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var httpSender_exports = {};
__export(httpSender_exports, {
  HttpSender: () => HttpSender
});
module.exports = __toCommonJS(httpSender_exports);
var import_node_url = __toESM(require("node:url"));
var import_api = require("@opentelemetry/api");
var import_core_rest_pipeline = require("@azure/core-rest-pipeline");
var import_generated = require("../../generated/index.js");
var import_baseSender = require("./baseSender.js");
var import_breezeUtils = require("../../utils/breezeUtils.js");
const applicationInsightsResource = "https://monitor.azure.com/.default";
class HttpSender extends import_baseSender.BaseSender {
  appInsightsClient;
  credential;
  appInsightsClientOptions;
  constructor(options) {
    super(options);
    this.appInsightsClientOptions = {
      host: options.endpointUrl,
      ...options.exporterOptions
    };
    if (this.appInsightsClientOptions.credential) {
      const scopes = options.aadAudience ? [options.aadAudience] : [applicationInsightsResource];
      this.appInsightsClientOptions.credentials = { scopes };
    } else if (!this.appInsightsClientOptions.credentials?.scopes && this.appInsightsClientOptions.credentialScopes) {
      const legacy = this.appInsightsClientOptions.credentialScopes;
      this.appInsightsClientOptions.credentials = {
        scopes: Array.isArray(legacy) ? legacy : [legacy]
      };
    }
    const { credential, ...clientOptions } = this.appInsightsClientOptions;
    this.credential = credential;
    this.appInsightsClient = this.createClient(clientOptions);
  }
  createClient(clientOptions) {
    const userPipeline = clientOptions.pipeline;
    const client = new import_generated.ApplicationInsightsClient(this.credential, clientOptions);
    client.host = clientOptions.host;
    if (userPipeline) {
      for (const policy of userPipeline.getOrderedPolicies()) {
        if (!client.pipeline.getOrderedPolicies().some((p) => p.name === policy.name)) {
          client.pipeline.addPolicy(policy);
        }
      }
    }
    if (!this.appInsightsClientOptions.credential) {
      client.pipeline.removePolicy({ name: import_core_rest_pipeline.bearerTokenAuthenticationPolicyName });
    }
    client.pipeline.removePolicy({ name: import_core_rest_pipeline.redirectPolicyName });
    return client;
  }
  /**
   * Send Azure envelopes
   * @internal
   */
  async send(envelopes) {
    let response;
    await this.appInsightsClient.track(envelopes, {
      onResponse(rawResponse) {
        response = rawResponse;
      }
    });
    return {
      statusCode: response?.status,
      result: response?.bodyAsText ?? "",
      retryAfterMs: (0, import_breezeUtils.parseRetryAfterHeader)(response?.headers?.get("retry-after"))
    };
  }
  /**
   * Shutdown sender
   * @internal
   */
  async shutdown() {
    import_api.diag.info("HttpSender shutting down");
  }
  handlePermanentRedirect(location) {
    if (location) {
      const locUrl = new import_node_url.default.URL(location);
      if (locUrl && locUrl.host) {
        this.appInsightsClientOptions.host = "https://" + locUrl.host;
        const { credential, ...clientOptions } = this.appInsightsClientOptions;
        this.appInsightsClient = this.createClient(clientOptions);
      }
    }
  }
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  HttpSender
});
