// @azure/monitor-opentelemetry-exporter/dist/commonjs/export/statsbeat/customerSDKStats.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var customerSDKStats_exports = {};
__export(customerSDKStats_exports, {
  CustomerSDKStatsMetrics: () => CustomerSDKStatsMetrics
});
module.exports = __toCommonJS(customerSDKStats_exports);
var import_api = require("@opentelemetry/api");
var import_sdk_metrics = require("@opentelemetry/sdk-metrics");
var ai = __toESM(require("../../utils/constants/applicationinsights.js"));
var import_statsbeatMetrics = require("./statsbeatMetrics.js");
var import_types = require("./types.js");
var import_types2 = require("./types.js");
var import_metricUtils = require("../../utils/metricUtils.js");
var import_types3 = require("../../types.js");
class CustomerSDKStatsMetrics extends import_statsbeatMetrics.StatsbeatMetrics {
  static _instance;
  statsCollectionInterval = 9e5;
  // 15 minutes
  customerSDKStatsMeter;
  customerSDKStatsMeterProvider;
  customerSDKStatsExporter;
  customerSDKStatsCounter;
  customerSDKStatsMetricReader;
  isInitialized = false;
  // Custom dimensions
  language;
  version;
  attach = (0, import_metricUtils.getAttachType)();
  // Observable Gauges
  itemSuccessCountGauge;
  itemDropCountGauge;
  itemRetryCountGauge;
  // Customer SDK Stats properties
  customerProperties;
  constructor(options, exporter) {
    super();
    this.customerSDKStatsExporter = exporter;
    const customerMetricReaderOptions = {
      exporter: this.customerSDKStatsExporter,
      exportIntervalMillis: options.networkCollectionInterval || this.statsCollectionInterval
    };
    this.customerSDKStatsMetricReader = new import_sdk_metrics.PeriodicExportingMetricReader(
      customerMetricReaderOptions
    );
    this.customerSDKStatsMeterProvider = new import_sdk_metrics.MeterProvider({
      readers: [this.customerSDKStatsMetricReader]
    });
    this.customerSDKStatsMeter = this.customerSDKStatsMeterProvider.getMeter(
      "Azure Monitor Customer SDK Stats"
    );
    this.language = import_types2.STATSBEAT_LANGUAGE;
    this.version = ai.packageVersion;
    this.itemSuccessCountGauge = this.customerSDKStatsMeter.createObservableGauge(
      import_types2.CustomSDKStatsCounter.ITEM_SUCCESS_COUNT
    );
    this.itemDropCountGauge = this.customerSDKStatsMeter.createObservableGauge(
      import_types2.CustomSDKStatsCounter.ITEM_DROP_COUNT
    );
    this.itemRetryCountGauge = this.customerSDKStatsMeter.createObservableGauge(
      import_types2.CustomSDKStatsCounter.ITEM_RETRY_COUNT
    );
    if (!this.isInitialized) {
      this.initialize();
    }
    this.isInitialized = true;
    this.customerSDKStatsCounter = new import_types.CustomerSDKStats();
    this.customerProperties = {
      language: this.language,
      version: this.version,
      computeType: this.attach
    };
  }
  /**
   * Get singleton instance of CustomerSDKStatsMetrics
   * @param options - Configuration options for customer SDK Stats metrics
   * @returns Promise of the singleton instance
   */
  static async getInstance(options) {
    if (!CustomerSDKStatsMetrics._instance) {
      const { AzureMonitorStatsbeatExporter } = await import("./statsbeatExporter.js");
      const customerStatsExporterConfig = {
        connectionString: `InstrumentationKey=${options.instrumentationKey};IngestionEndpoint=${options.endpointUrl}`
      };
      const exporter = new AzureMonitorStatsbeatExporter(customerStatsExporterConfig);
      CustomerSDKStatsMetrics._instance = new CustomerSDKStatsMetrics(options, exporter);
    }
    return CustomerSDKStatsMetrics._instance;
  }
  /**
   * Shutdown the singleton instance
   * Used for cleanup and complete shutdown
   */
  static shutdown() {
    if (CustomerSDKStatsMetrics._instance) {
      const shutdownPromise = CustomerSDKStatsMetrics._instance.shutdown();
      CustomerSDKStatsMetrics._instance = void 0;
      return shutdownPromise;
    }
    return void 0;
  }
  /**
   * Shuts down the customer SDK Stats metrics provider
   * @returns Promise<void>
   */
  shutdown() {
    return this.customerSDKStatsMeterProvider.shutdown();
  }
  /**
   * Initializes the customer SDK Stats metrics
   * Sets up the resource provider and adds observable callbacks for each metric
   * @returns Promise<void>
   */
  async initialize() {
    try {
      await super.getResourceProvider();
      this.customerSDKStatsMeter.addBatchObservableCallback(this.itemSuccessCallback.bind(this), [
        this.itemSuccessCountGauge
      ]);
      this.customerSDKStatsMeter.addBatchObservableCallback(this.itemDropCallback.bind(this), [
        this.itemDropCountGauge
      ]);
      this.customerSDKStatsMeter.addBatchObservableCallback(this.itemRetryCallback.bind(this), [
        this.itemRetryCountGauge
      ]);
    } catch (error) {
      import_api.diag.debug("Call to get the resource provider failed for customer SDK Stats metrics.");
    }
  }
  // Observable gauge callbacks
  itemSuccessCallback(observableResult) {
    const counter = this.customerSDKStatsCounter;
    const attributes = { ...this.customerProperties, telemetry_type: import_types2.TelemetryType.UNKNOWN };
    for (const [telemetry_type, count] of counter.totalItemSuccessCount.entries()) {
      if (count > 0) {
        attributes.telemetry_type = telemetry_type;
        observableResult.observe(this.itemSuccessCountGauge, count, {
          ...attributes
        });
        counter.totalItemSuccessCount.set(telemetry_type, 0);
      }
    }
  }
  itemDropCallback(observableResult) {
    const counter = this.customerSDKStatsCounter;
    const baseAttributes = {
      ...this.customerProperties,
      "drop.code": import_types.DropCode.UNKNOWN,
      telemetry_type: import_types2.TelemetryType.UNKNOWN
    };
    for (const [telemetryType, dropCodeMap] of counter.totalItemDropCount.entries()) {
      for (const [dropCode, reasonMap] of dropCodeMap.entries()) {
        for (const [reason, successMap] of reasonMap.entries()) {
          for (const [success, count] of successMap.entries()) {
            const attributes = { ...baseAttributes };
            attributes.telemetry_type = telemetryType;
            attributes["drop.code"] = dropCode;
            if (reason) {
              attributes["drop.reason"] = reason;
            }
            if ((telemetryType === import_types2.TelemetryType.REQUEST || telemetryType === import_types2.TelemetryType.DEPENDENCY) && success !== null) {
              attributes["telemetry_success"] = success;
            }
            if (count > 0) {
              observableResult.observe(this.itemDropCountGauge, count, {
                ...attributes
              });
            }
            successMap.set(success, 0);
          }
        }
      }
    }
  }
  itemRetryCallback(observableResult) {
    const counter = this.customerSDKStatsCounter;
    const baseAttributes = {
      ...this.customerProperties,
      "retry.code": import_types.RetryCode.UNKNOWN,
      telemetry_type: import_types2.TelemetryType.UNKNOWN
    };
    for (const [telemetryType, retryCodeMap] of counter.totalItemRetryCount.entries()) {
      for (const [retryCode, reasonMap] of retryCodeMap.entries()) {
        for (const [reason, count] of reasonMap.entries()) {
          const attributes = { ...baseAttributes };
          attributes.telemetry_type = telemetryType;
          attributes["retry.code"] = retryCode;
          if (reason) {
            attributes["retry.reason"] = reason;
          }
          if (count > 0) {
            observableResult.observe(this.itemRetryCountGauge, count, {
              ...attributes
            });
          }
          reasonMap.set(reason, 0);
        }
      }
    }
  }
  // Public methods to track metrics
  /**
   * Tracks succcessful items
   * @param envelopes - Number of successful envelopes
   * @param telemetry_type - The type of telemetry being tracked
   */
  countSuccessfulItems(envelopes) {
    const counter = this.customerSDKStatsCounter;
    let telemetry_type;
    for (const envelope of envelopes) {
      telemetry_type = this.getTelemetryTypeFromEnvelope(envelope);
      const currentCount = counter.totalItemSuccessCount.get(telemetry_type) || 0;
      counter.totalItemSuccessCount.set(telemetry_type, currentCount + 1);
    }
  }
  /**
   * Tracks dropped items
   * @param envelopes - Array of envelopes dropped
   * @param dropCode - The drop code indicating the reason for drop
   * @param exceptionMessage - Optional exception message when dropCode is CLIENT_EXCEPTION
   * @param exceptionType - Optional explicit exception type override when dropCode is CLIENT_EXCEPTION
   */
  countDroppedItems(envelopes, dropCode, exceptionMessage, exceptionType) {
    const counter = this.customerSDKStatsCounter;
    let telemetry_type;
    for (const envelope of envelopes) {
      telemetry_type = this.getTelemetryTypeFromEnvelope(envelope);
      let dropCodeMap = counter.totalItemDropCount.get(telemetry_type);
      if (!dropCodeMap) {
        dropCodeMap = /* @__PURE__ */ new Map();
        counter.totalItemDropCount.set(telemetry_type, dropCodeMap);
      }
      let reasonMap = dropCodeMap.get(dropCode);
      if (!reasonMap) {
        reasonMap = /* @__PURE__ */ new Map();
        dropCodeMap.set(dropCode, reasonMap);
      }
      const reason = this.getDropReason(dropCode, exceptionMessage, exceptionType);
      let successMap = reasonMap.get(reason);
      if (!successMap) {
        successMap = /* @__PURE__ */ new Map();
        reasonMap.set(reason, successMap);
      }
      const individualTelemetrySuccess = this.getTelemetrySuccessFromEnvelope(envelope);
      const successKey = (telemetry_type === import_types2.TelemetryType.REQUEST || telemetry_type === import_types2.TelemetryType.DEPENDENCY) && individualTelemetrySuccess !== void 0 ? individualTelemetrySuccess : null;
      const currentCount = successMap.get(successKey) || 0;
      successMap.set(successKey, currentCount + 1);
    }
  }
  /**
   * Generates a low-cardinality, informative description for drop reasons
   * @param dropCode - The drop code (enum value or status code number)
   * @param exceptionMessage - Optional exception message for CLIENT_EXCEPTION
   * @param exceptionType - Optional explicit exception type override for CLIENT_EXCEPTION
   * @returns A descriptive reason string with low cardinality
   */
  getDropReason(dropCode, exceptionMessage, exceptionType) {
    if (dropCode === import_types.DropCode.CLIENT_EXCEPTION) {
      if (exceptionType) {
        return exceptionType;
      }
      if (exceptionMessage) {
        return this.categorizeExceptionMessage(exceptionMessage);
      }
      return import_types.ExceptionType.CLIENT_EXCEPTION;
    }
    if (typeof dropCode === "number") {
      return this.categorizeStatusCode(dropCode);
    }
    switch (dropCode) {
      case import_types.DropCode.CLIENT_READONLY:
        return import_types.DropReason.CLIENT_READONLY;
      case import_types.DropCode.CLIENT_PERSISTENCE_CAPACITY:
        return import_types.DropReason.CLIENT_PERSISTENCE_CAPACITY;
      case import_types.DropCode.CLIENT_STORAGE_DISABLED:
        return import_types.DropReason.CLIENT_STORAGE_DISABLED;
      case import_types.DropCode.UNKNOWN:
      default:
        return import_types.DropReason.UNKNOWN;
    }
  }
  /**
   * Categorizes exception messages into well-known exception categories
   * @param exceptionMessage - The exception message to categorize
   * @returns A well-known exception category string
   */
  categorizeExceptionMessage(exceptionMessage) {
    const message = exceptionMessage.toLowerCase();
    if (message.includes("timeout") || message.includes("timed out")) {
      return import_types.ExceptionType.TIMEOUT_EXCEPTION;
    }
    if (message.includes("network") || message.includes("connection") || message.includes("dns") || message.includes("socket")) {
      return import_types.ExceptionType.NETWORK_EXCEPTION;
    }
    if (message.includes("disk") || message.includes("storage") || message.includes("file") || message.includes("persist")) {
      return import_types.ExceptionType.STORAGE_EXCEPTION;
    }
    return import_types.ExceptionType.CLIENT_EXCEPTION;
  }
  /**
   * Categorizes HTTP status codes into informative descriptions
   * @param statusCode - The HTTP status code
   * @returns A descriptive category string
   */
  categorizeStatusCode(statusCode) {
    if (statusCode >= 400 && statusCode < 500) {
      switch (statusCode) {
        case 400:
          return "Bad request";
        case 401:
          return "Unauthorized";
        case 403:
          return "Forbidden";
        case 404:
          return "Not found";
        case 408:
          return "Request timeout";
        case 413:
          return "Payload too large";
        case 429:
          return "Too many requests";
        default:
          return "Client error 4xx";
      }
    }
    if (statusCode >= 500 && statusCode < 600) {
      switch (statusCode) {
        case 500:
          return "Internal server error";
        case 502:
          return "Bad gateway";
        case 503:
          return "Service unavailable";
        case 504:
          return "Gateway timeout";
        default:
          return "Server error 5xx";
      }
    }
    return `status_${statusCode}`;
  }
  /**
   * Tracks retried envelopes
   * @param envelopes - Number of envelopes retried
   * @param retryCode - The retry code indicating the reason for retry
   * @param exceptionMessage - Optional exception message when retryCode is CLIENT_EXCEPTION
   * @param exceptionType - Optional explicit exception type override when retryCode is CLIENT_EXCEPTION
   */
  countRetryItems(envelopes, retryCode, exceptionMessage, exceptionType) {
    const counter = this.customerSDKStatsCounter;
    let telemetry_type;
    for (const envelope of envelopes) {
      telemetry_type = this.getTelemetryTypeFromEnvelope(envelope);
      let retryCodeMap = counter.totalItemRetryCount.get(telemetry_type);
      if (!retryCodeMap) {
        retryCodeMap = /* @__PURE__ */ new Map();
        counter.totalItemRetryCount.set(telemetry_type, retryCodeMap);
      }
      let reasonMap = retryCodeMap.get(retryCode);
      if (!reasonMap) {
        reasonMap = /* @__PURE__ */ new Map();
        retryCodeMap.set(retryCode, reasonMap);
      }
      const reason = this.getRetryReason(retryCode, exceptionMessage, exceptionType);
      const currentCount = reasonMap.get(reason) || 0;
      reasonMap.set(reason, currentCount + 1);
    }
  }
  /**
   * Generates a low-cardinality, informative description for retry reasons
   * @param retryCode - The retry code (enum value or status code number)
   * @param exceptionMessage - Optional exception message for CLIENT_EXCEPTION
   * @param exceptionType - Optional explicit exception type override for CLIENT_EXCEPTION
   * @returns A descriptive reason string with low cardinality
   */
  getRetryReason(retryCode, exceptionMessage, exceptionType) {
    if (retryCode === import_types.RetryCode.CLIENT_EXCEPTION) {
      if (exceptionType) {
        return exceptionType;
      }
      if (exceptionMessage) {
        return this.categorizeExceptionMessage(exceptionMessage);
      }
      return import_types.ExceptionType.CLIENT_EXCEPTION;
    }
    if (typeof retryCode === "number") {
      return this.categorizeStatusCode(retryCode);
    }
    switch (retryCode) {
      case import_types.RetryCode.CLIENT_TIMEOUT:
        return import_types.RetryReason.CLIENT_TIMEOUT;
      case import_types.RetryCode.UNKNOWN:
      default:
        return import_types.RetryReason.UNKNOWN;
    }
  }
  /**
   * Check if a metric name corresponds to a performance counter
   * @param metricName - The name of the metric to check
   * @returns true if the metric name is a performance counter, false otherwise
   */
  isPerformanceCounterMetric(metricName) {
    return Object.values(import_types3.BreezePerformanceCounterNames).includes(
      metricName
    );
  }
  /**
   * Extract telemetry type from an envelope based on its baseType
   * @param envelope - The envelope to extract telemetry type from
   * @returns The corresponding telemetry type
   */
  getTelemetryTypeFromEnvelope(envelope) {
    if (envelope.data && envelope.data.baseType) {
      switch (envelope.data.baseType) {
        case "MessageData":
          return import_types2.TelemetryType.TRACE;
        case "AvailabilityData":
          return import_types2.TelemetryType.AVAILABILITY;
        case "TelemetryEventData":
          return import_types2.TelemetryType.CUSTOM_EVENT;
        case "TelemetryExceptionData":
          return import_types2.TelemetryType.EXCEPTION;
        case "PageViewData":
          return import_types2.TelemetryType.PAGE_VIEW;
        case "RemoteDependencyData":
          return import_types2.TelemetryType.DEPENDENCY;
        case "RequestData":
          return import_types2.TelemetryType.REQUEST;
        case "MetricData": {
          const metricsData = envelope.data.baseData;
          if (metricsData && metricsData.metrics && metricsData.metrics.length > 0) {
            const hasPerformanceCounter = metricsData.metrics.some(
              (metric) => this.isPerformanceCounterMetric(metric.name)
            );
            return hasPerformanceCounter ? import_types2.TelemetryType.PERFORMANCE_COUNTER : import_types2.TelemetryType.CUSTOM_METRIC;
          }
          return import_types2.TelemetryType.CUSTOM_METRIC;
        }
        default:
          return import_types2.TelemetryType.UNKNOWN;
      }
    }
    return import_types2.TelemetryType.UNKNOWN;
  }
  /**
   * Extract telemetry success value from an envelope for REQUEST and DEPENDENCY telemetry types
   * @param envelope - The envelope to extract success value from
   * @returns The success value if available, undefined otherwise
   */
  getTelemetrySuccessFromEnvelope(envelope) {
    if (!envelope.data || !envelope.data.baseData) {
      return void 0;
    }
    const baseType = envelope.data.baseType;
    if (baseType === "RequestData") {
      const requestData = envelope.data.baseData;
      return requestData.success;
    } else if (baseType === "RemoteDependencyData") {
      const dependencyData = envelope.data.baseData;
      return dependencyData.success;
    }
    return void 0;
  }
  /**
   * Checks if the given error is a timeout-related error
   * @param error - The error to check
   * @returns true if the error is timeout-related, false otherwise
   */
  isTimeoutError(error) {
    const timeoutErrorCodes = [
      "ETIMEDOUT",
      // Connection timed out
      "ESOCKETTIMEDOUT",
      // Socket timeout
      "ECONNRESET",
      // Connection reset (often due to timeout)
      "ENOTFOUND"
      // DNS lookup failed/timeout
    ];
    if (error && error.code && timeoutErrorCodes.includes(error.code)) {
      return true;
    }
    if (error && error.message) {
      const timeoutKeywords = ["timeout", "timed out", "connection reset"];
      const errorMessage = error.message.toLowerCase();
      return timeoutKeywords.some((keyword) => errorMessage.includes(keyword));
    }
    return false;
  }
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  CustomerSDKStatsMetrics
});
