// @azure/monitor-opentelemetry-exporter/dist/commonjs/export/statsbeat/longIntervalStatsbeatMetrics.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var longIntervalStatsbeatMetrics_exports = {};
__export(longIntervalStatsbeatMetrics_exports, {
  LongIntervalStatsbeatMetrics: () => LongIntervalStatsbeatMetrics
});
module.exports = __toCommonJS(longIntervalStatsbeatMetrics_exports);
var import_api = require("@opentelemetry/api");
var import_core = require("@opentelemetry/core");
var import_sdk_metrics = require("@opentelemetry/sdk-metrics");
var ai = __toESM(require("../../utils/constants/applicationinsights.js"));
var import_statsbeatMetrics = require("./statsbeatMetrics.js");
var import_types = require("./types.js");
var import_statsbeatExporter = require("./statsbeatExporter.js");
var import_metricUtils = require("../../utils/metricUtils.js");
class LongIntervalStatsbeatMetrics extends import_statsbeatMetrics.StatsbeatMetrics {
  static instance = null;
  statsCollectionLongInterval = 864e5;
  // 1 day
  // Custom dimensions
  cikey;
  runtimeVersion;
  language;
  version;
  attach = (0, import_metricUtils.getAttachType)();
  commonProperties;
  attachProperties;
  feature = 0;
  instrumentation = 0;
  longIntervalStatsbeatMeterProvider;
  longIntervalAzureExporter;
  longIntervalMetricReader;
  longIntervalStatsbeatMeter;
  // Network Attributes
  connectionString;
  // Observable Gauges
  featureStatsbeatGauge;
  attachStatsbeatGauge;
  isInitialized = false;
  constructor(options) {
    super();
    this.connectionString = super.getConnectionString(options.endpointUrl);
    const exporterConfig = {
      connectionString: this.connectionString,
      disableOfflineStorage: options.disableOfflineStorage
    };
    this.setFeatures();
    this.longIntervalAzureExporter = new import_statsbeatExporter.AzureMonitorStatsbeatExporter(exporterConfig);
    const longIntervalMetricReaderOptions = {
      exporter: this.longIntervalAzureExporter,
      exportIntervalMillis: Number(process.env.LONG_INTERVAL_EXPORT_MILLIS) || this.statsCollectionLongInterval
      // 1 day
    };
    this.longIntervalMetricReader = new import_sdk_metrics.PeriodicExportingMetricReader(
      longIntervalMetricReaderOptions
    );
    this.longIntervalStatsbeatMeterProvider = new import_sdk_metrics.MeterProvider({
      readers: [this.longIntervalMetricReader]
    });
    this.longIntervalStatsbeatMeter = this.longIntervalStatsbeatMeterProvider.getMeter(
      "Azure Monitor Long Interval Statsbeat"
    );
    this.runtimeVersion = process.version;
    this.language = import_types.STATSBEAT_LANGUAGE;
    this.version = ai.packageVersion;
    this.cikey = options.instrumentationKey;
    this.featureStatsbeatGauge = this.longIntervalStatsbeatMeter.createObservableGauge(
      import_types.StatsbeatCounter.FEATURE
    );
    this.attachStatsbeatGauge = this.longIntervalStatsbeatMeter.createObservableGauge(
      import_types.StatsbeatCounter.ATTACH
    );
    this.isInitialized = true;
    this.initialize();
    this.commonProperties = {
      os: this.os,
      rp: this.resourceProvider,
      cikey: this.cikey,
      runtimeVersion: this.runtimeVersion,
      language: this.language,
      version: this.version,
      attach: this.attach
    };
    this.attachProperties = {
      rpId: this.resourceIdentifier
    };
  }
  async initialize() {
    try {
      await this.getResourceProvider();
      this.attachStatsbeatGauge.addCallback(this.attachCallback.bind(this));
      this.longIntervalStatsbeatMeter.addBatchObservableCallback(
        this.getEnvironmentStatus.bind(this),
        [this.featureStatsbeatGauge]
      );
      setTimeout(async () => {
        try {
          const collectionResult = await this.longIntervalMetricReader.collect();
          if (collectionResult) {
            this.longIntervalAzureExporter.export(
              collectionResult.resourceMetrics,
              (result) => {
                if (result.code !== import_core.ExportResultCode.SUCCESS) {
                  import_api.diag.debug(
                    `LongIntervalStatsbeat: metrics export failed (error ${result.error})`
                  );
                }
              }
            );
          } else {
            import_api.diag.debug("LongIntervalStatsbeat: No metrics collected");
          }
        } catch (error) {
          import_api.diag.debug(`LongIntervalStatsbeat: Error collecting metrics: ${error}`);
        }
      }, 15e3);
    } catch (error) {
      import_api.diag.debug("Call to get the resource provider failed.");
    }
  }
  getEnvironmentStatus(observableResult) {
    this.setFeatures();
    let attributes;
    if (this.instrumentation > 0) {
      attributes = {
        ...this.commonProperties,
        feature: this.instrumentation,
        type: import_types.StatsbeatFeatureType.INSTRUMENTATION
      };
      observableResult.observe(this.featureStatsbeatGauge, 1, { ...attributes });
    }
    if (this.feature > 0) {
      attributes = {
        ...this.commonProperties,
        feature: this.feature,
        type: import_types.StatsbeatFeatureType.FEATURE
      };
      observableResult.observe(this.featureStatsbeatGauge, 1, { ...attributes });
    }
  }
  setFeatures() {
    const statsbeatFeatures = process.env.AZURE_MONITOR_STATSBEAT_FEATURES;
    if (statsbeatFeatures) {
      try {
        this.feature = JSON.parse(statsbeatFeatures).feature;
        this.instrumentation = JSON.parse(statsbeatFeatures).instrumentation;
      } catch (error) {
        import_api.diag.debug(
          `LongIntervalStatsbeat: Failed to parse features/instrumentations (error ${error})`
        );
      }
    }
  }
  attachCallback(observableResult) {
    const attributes = { ...this.commonProperties, ...this.attachProperties };
    observableResult.observe(1, attributes);
  }
  shutdown() {
    return this.longIntervalStatsbeatMeterProvider.shutdown();
  }
  /**
   * Singleton LongIntervalStatsbeatMetrics instance.
   * @internal
   */
  static getInstance(options) {
    if (!LongIntervalStatsbeatMetrics.instance) {
      LongIntervalStatsbeatMetrics.instance = new LongIntervalStatsbeatMetrics(options);
    }
    return LongIntervalStatsbeatMetrics.instance;
  }
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  LongIntervalStatsbeatMetrics
});
