// @azure/monitor-opentelemetry-exporter/dist/commonjs/export/statsbeat/networkStatsbeatMetrics.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var networkStatsbeatMetrics_exports = {};
__export(networkStatsbeatMetrics_exports, {
  NetworkStatsbeatMetrics: () => NetworkStatsbeatMetrics
});
module.exports = __toCommonJS(networkStatsbeatMetrics_exports);
var import_api = require("@opentelemetry/api");
var import_sdk_metrics = require("@opentelemetry/sdk-metrics");
var ai = __toESM(require("../../utils/constants/applicationinsights.js"));
var import_statsbeatMetrics = require("./statsbeatMetrics.js");
var import_types = require("./types.js");
var import_statsbeatExporter = require("./statsbeatExporter.js");
var import_Constants = require("../../Declarations/Constants.js");
var import_metricUtils = require("../../utils/metricUtils.js");
class NetworkStatsbeatMetrics extends import_statsbeatMetrics.StatsbeatMetrics {
  static instance = null;
  disableNonEssentialStatsbeat = !!process.env[import_Constants.ENV_DISABLE_STATSBEAT];
  commonProperties;
  networkProperties;
  isInitialized = false;
  statsCollectionShortInterval = 9e5;
  // 15 minutes
  networkStatsbeatCollection = [];
  networkStatsbeatMeter;
  networkStatsbeatMeterProvider;
  networkAzureExporter;
  // Custom dimensions
  cikey;
  runtimeVersion;
  language;
  version;
  attach = (0, import_metricUtils.getAttachType)();
  // Observable Gauges
  successCountGauge;
  failureCountGauge;
  retryCountGauge;
  throttleCountGauge;
  exceptionCountGauge;
  averageDurationGauge;
  // Non-essential Statsbeat Gauges
  readFailureGauge;
  writeFailureGauge;
  // Network attributes
  connectionString;
  endpointUrl;
  host;
  constructor(options) {
    super();
    this.connectionString = super.getConnectionString(options.endpointUrl);
    const exporterConfig = {
      connectionString: this.connectionString
    };
    this.networkAzureExporter = new import_statsbeatExporter.AzureMonitorStatsbeatExporter(exporterConfig);
    const networkMetricReaderOptions = {
      exporter: this.networkAzureExporter,
      exportIntervalMillis: options.networkCollectionInterval || this.statsCollectionShortInterval
      // 15 minutes
    };
    this.networkStatsbeatMeterProvider = new import_sdk_metrics.MeterProvider({
      readers: [new import_sdk_metrics.PeriodicExportingMetricReader(networkMetricReaderOptions)]
    });
    this.networkStatsbeatMeter = this.networkStatsbeatMeterProvider.getMeter(
      "Azure Monitor Network Statsbeat"
    );
    this.endpointUrl = options.endpointUrl;
    this.runtimeVersion = process.version;
    this.language = import_types.STATSBEAT_LANGUAGE;
    this.version = ai.packageVersion;
    this.host = this.getShortHost(options.endpointUrl);
    this.cikey = options.instrumentationKey;
    this.successCountGauge = this.networkStatsbeatMeter.createObservableGauge(
      import_types.StatsbeatCounter.SUCCESS_COUNT
    );
    this.failureCountGauge = this.networkStatsbeatMeter.createObservableGauge(
      import_types.StatsbeatCounter.FAILURE_COUNT
    );
    this.retryCountGauge = this.networkStatsbeatMeter.createObservableGauge(
      import_types.StatsbeatCounter.RETRY_COUNT
    );
    this.throttleCountGauge = this.networkStatsbeatMeter.createObservableGauge(
      import_types.StatsbeatCounter.THROTTLE_COUNT
    );
    this.exceptionCountGauge = this.networkStatsbeatMeter.createObservableGauge(
      import_types.StatsbeatCounter.EXCEPTION_COUNT
    );
    this.averageDurationGauge = this.networkStatsbeatMeter.createObservableGauge(
      import_types.StatsbeatCounter.AVERAGE_DURATION
    );
    if (!this.disableNonEssentialStatsbeat) {
      this.readFailureGauge = this.networkStatsbeatMeter.createObservableGauge(
        import_types.StatsbeatCounter.READ_FAILURE_COUNT
      );
      this.writeFailureGauge = this.networkStatsbeatMeter.createObservableGauge(
        import_types.StatsbeatCounter.WRITE_FAILURE_COUNT
      );
    }
    this.isInitialized = true;
    this.initialize();
    this.commonProperties = {
      os: this.os,
      rp: this.resourceProvider,
      cikey: this.cikey,
      runtimeVersion: this.runtimeVersion,
      language: this.language,
      version: this.version,
      attach: this.attach
    };
    this.networkProperties = {
      endpoint: this.endpointUrl,
      host: this.host
    };
  }
  shutdown() {
    return this.networkStatsbeatMeterProvider.shutdown();
  }
  async initialize() {
    try {
      await super.getResourceProvider();
      this.successCountGauge.addCallback(this.successCallback.bind(this));
      this.networkStatsbeatMeter.addBatchObservableCallback(this.failureCallback.bind(this), [
        this.failureCountGauge
      ]);
      this.networkStatsbeatMeter.addBatchObservableCallback(this.retryCallback.bind(this), [
        this.retryCountGauge
      ]);
      this.networkStatsbeatMeter.addBatchObservableCallback(this.throttleCallback.bind(this), [
        this.throttleCountGauge
      ]);
      this.networkStatsbeatMeter.addBatchObservableCallback(this.exceptionCallback.bind(this), [
        this.exceptionCountGauge
      ]);
      if (!this.disableNonEssentialStatsbeat) {
        this.readFailureGauge?.addCallback(this.readFailureCallback.bind(this));
        this.writeFailureGauge?.addCallback(this.writeFailureCallback.bind(this));
      }
      this.averageDurationGauge.addCallback(this.durationCallback.bind(this));
    } catch (error) {
      import_api.diag.debug("Call to get the resource provider failed.");
    }
  }
  // Observable gauge callbacks
  successCallback(observableResult) {
    const counter = this.getNetworkStatsbeatCounter(this.endpointUrl, this.host);
    if (counter.totalSuccessfulRequestCount > 0) {
      const attributes = { ...this.commonProperties, ...this.networkProperties };
      observableResult.observe(counter.totalSuccessfulRequestCount, attributes);
      counter.totalSuccessfulRequestCount = 0;
    }
  }
  failureCallback(observableResult) {
    const counter = this.getNetworkStatsbeatCounter(this.endpointUrl, this.host);
    const attributes = { ...this.networkProperties, ...this.commonProperties, statusCode: 0 };
    for (let i = 0; i < counter.totalFailedRequestCount.length; i++) {
      if (counter.totalFailedRequestCount[i].count > 0) {
        attributes.statusCode = counter.totalFailedRequestCount[i].statusCode;
        observableResult.observe(this.failureCountGauge, counter.totalFailedRequestCount[i].count, {
          ...attributes
        });
        counter.totalFailedRequestCount[i].count = 0;
      }
    }
  }
  retryCallback(observableResult) {
    const counter = this.getNetworkStatsbeatCounter(this.endpointUrl, this.host);
    const attributes = { ...this.networkProperties, ...this.commonProperties, statusCode: 0 };
    for (let i = 0; i < counter.retryCount.length; i++) {
      if (counter.retryCount[i].count > 0) {
        attributes.statusCode = counter.retryCount[i].statusCode;
        observableResult.observe(this.retryCountGauge, counter.retryCount[i].count, {
          ...attributes
        });
        counter.retryCount[i].count = 0;
      }
    }
  }
  throttleCallback(observableResult) {
    const counter = this.getNetworkStatsbeatCounter(this.endpointUrl, this.host);
    const attributes = { ...this.networkProperties, ...this.commonProperties, statusCode: 0 };
    for (let i = 0; i < counter.throttleCount.length; i++) {
      if (counter.throttleCount[i].count > 0) {
        attributes.statusCode = counter.throttleCount[i].statusCode;
        observableResult.observe(this.throttleCountGauge, counter.throttleCount[i].count, {
          ...attributes
        });
        counter.throttleCount[i].count = 0;
      }
    }
  }
  exceptionCallback(observableResult) {
    const counter = this.getNetworkStatsbeatCounter(this.endpointUrl, this.host);
    const attributes = { ...this.networkProperties, ...this.commonProperties, exceptionType: "" };
    for (let i = 0; i < counter.exceptionCount.length; i++) {
      if (counter.exceptionCount[i].count > 0) {
        attributes.exceptionType = counter.exceptionCount[i].exceptionType;
        observableResult.observe(this.exceptionCountGauge, counter.exceptionCount[i].count, {
          ...attributes
        });
        counter.exceptionCount[i].count = 0;
      }
    }
  }
  durationCallback(observableResult) {
    const counter = this.getNetworkStatsbeatCounter(this.endpointUrl, this.host);
    const attributes = { ...this.networkProperties, ...this.commonProperties };
    for (let i = 0; i < this.networkStatsbeatCollection.length; i++) {
      const currentCounter = this.networkStatsbeatCollection[i];
      currentCounter.time = Number(/* @__PURE__ */ new Date());
      const intervalRequests = currentCounter.totalRequestCount - currentCounter.lastRequestCount || 0;
      if (intervalRequests > 0) {
        currentCounter.averageRequestExecutionTime = (currentCounter.intervalRequestExecutionTime - currentCounter.lastIntervalRequestExecutionTime) / intervalRequests || 0;
      } else {
        currentCounter.averageRequestExecutionTime = 0;
      }
      currentCounter.lastIntervalRequestExecutionTime = currentCounter.intervalRequestExecutionTime;
      currentCounter.lastRequestCount = currentCounter.totalRequestCount;
      currentCounter.lastTime = currentCounter.time;
    }
    if (counter.averageRequestExecutionTime > 0) {
      observableResult.observe(counter.averageRequestExecutionTime, attributes);
      counter.averageRequestExecutionTime = 0;
    }
  }
  readFailureCallback(observableResult) {
    const counter = this.getNetworkStatsbeatCounter(this.endpointUrl, this.host);
    if (counter.totalReadFailureCount > 0) {
      const attributes = { ...this.commonProperties, ...this.networkProperties };
      observableResult.observe(counter.totalReadFailureCount, attributes);
      counter.totalReadFailureCount = 0;
    }
  }
  writeFailureCallback(observableResult) {
    const counter = this.getNetworkStatsbeatCounter(this.endpointUrl, this.host);
    if (counter.totalWriteFailureCount > 0) {
      const attributes = { ...this.commonProperties, ...this.networkProperties };
      observableResult.observe(counter.totalWriteFailureCount, attributes);
      counter.totalWriteFailureCount = 0;
    }
  }
  // Public methods to increase counters
  countSuccess(duration) {
    if (!this.isInitialized) {
      return;
    }
    const counter = this.getNetworkStatsbeatCounter(this.endpointUrl, this.host);
    counter.totalRequestCount++;
    counter.totalSuccessfulRequestCount++;
    counter.intervalRequestExecutionTime += duration;
  }
  countFailure(duration, statusCode) {
    if (!this.isInitialized) {
      return;
    }
    const counter = this.getNetworkStatsbeatCounter(this.endpointUrl, this.host);
    const currentStatusCounter = counter.totalFailedRequestCount.find(
      (statusCounter) => statusCode === statusCounter.statusCode
    );
    if (currentStatusCounter) {
      currentStatusCounter.count++;
    } else {
      counter.totalFailedRequestCount.push({ statusCode, count: 1 });
    }
    counter.totalRequestCount++;
    counter.intervalRequestExecutionTime += duration;
  }
  countRetry(statusCode) {
    if (!this.isInitialized) {
      return;
    }
    const counter = this.getNetworkStatsbeatCounter(this.endpointUrl, this.host);
    const currentStatusCounter = counter.retryCount.find(
      (statusCounter) => statusCode === statusCounter.statusCode
    );
    if (currentStatusCounter) {
      currentStatusCounter.count++;
    } else {
      counter.retryCount.push({ statusCode, count: 1 });
    }
  }
  countThrottle(statusCode) {
    if (!this.isInitialized) {
      return;
    }
    const counter = this.getNetworkStatsbeatCounter(this.endpointUrl, this.host);
    const currentStatusCounter = counter.throttleCount.find(
      (statusCounter) => statusCode === statusCounter.statusCode
    );
    if (currentStatusCounter) {
      currentStatusCounter.count++;
    } else {
      counter.throttleCount.push({ statusCode, count: 1 });
    }
  }
  countReadFailure() {
    if (!this.isInitialized || this.disableNonEssentialStatsbeat) {
      return;
    }
    const counter = this.getNetworkStatsbeatCounter(this.endpointUrl, this.host);
    counter.totalReadFailureCount++;
  }
  countWriteFailure() {
    if (!this.isInitialized || this.disableNonEssentialStatsbeat) {
      return;
    }
    const counter = this.getNetworkStatsbeatCounter(this.endpointUrl, this.host);
    counter.totalWriteFailureCount++;
  }
  countException(exceptionType) {
    if (!this.isInitialized) {
      return;
    }
    const counter = this.getNetworkStatsbeatCounter(this.endpointUrl, this.host);
    const currentErrorCounter = counter.exceptionCount.find(
      (exceptionCounter) => exceptionType.name === exceptionCounter.exceptionType
    );
    if (currentErrorCounter) {
      currentErrorCounter.count++;
    } else {
      counter.exceptionCount.push({ exceptionType: exceptionType.name, count: 1 });
    }
  }
  // Gets a networkStatsbeat counter if one exists for the given endpoint
  getNetworkStatsbeatCounter(endpoint, host) {
    for (let i = 0; i < this.networkStatsbeatCollection.length; i++) {
      if (endpoint === this.networkStatsbeatCollection[i].endpoint && host === this.networkStatsbeatCollection[i].host) {
        return this.networkStatsbeatCollection[i];
      }
    }
    const newCounter = new import_types.NetworkStatsbeat(endpoint, host);
    this.networkStatsbeatCollection.push(newCounter);
    return newCounter;
  }
  getShortHost(originalHost) {
    let shortHost = originalHost;
    try {
      const hostRegex = new RegExp(/^https?:\/\/(?:www\.)?([^/.-]+)/);
      const res = hostRegex.exec(originalHost);
      if (res !== null && res.length > 1) {
        shortHost = res[1];
      }
      shortHost = shortHost.replace(".in.applicationinsights.azure.com", "");
    } catch (error) {
      import_api.diag.debug("Failed to get the short host name.");
    }
    return shortHost;
  }
  /**
   * Singleton Network Statsbeat Metrics instance.
   * @internal
   */
  static getInstance(options) {
    if (!NetworkStatsbeatMetrics.instance) {
      NetworkStatsbeatMetrics.instance = new NetworkStatsbeatMetrics(options);
    }
    return NetworkStatsbeatMetrics.instance;
  }
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  NetworkStatsbeatMetrics
});
