// @azure/monitor-opentelemetry-exporter/dist/commonjs/export/statsbeat/statsbeatExporter.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var statsbeatExporter_exports = {};
__export(statsbeatExporter_exports, {
  AzureMonitorStatsbeatExporter: () => AzureMonitorStatsbeatExporter
});
module.exports = __toCommonJS(statsbeatExporter_exports);
var import_api = require("@opentelemetry/api");
var import_core = require("@opentelemetry/core");
var import_metricUtils = require("../../utils/metricUtils.js");
var import_base = require("../base.js");
class AzureMonitorStatsbeatExporter extends import_base.AzureMonitorBaseExporter {
  /**
   * Flag to determine if the Exporter is shutdown.
   */
  _isShutdown = false;
  _sender;
  _senderOptions;
  /**
   * Initializes a new instance of the AzureMonitorStatsbeatExporter class.
   * @param options - Exporter configuration
   */
  constructor(options) {
    super(options, true);
    this._senderOptions = {
      endpointUrl: this.endpointUrl,
      instrumentationKey: this.instrumentationKey,
      trackStatsbeat: this.trackStatsbeat,
      exporterOptions: options,
      isStatsbeatSender: true
    };
  }
  /**
   * Lazily initialize the sender to avoid circular dependency
   */
  async _getSender() {
    if (!this._sender) {
      const { HttpSender } = await import("../../platform/nodejs/httpSender.js");
      this._sender = new HttpSender(this._senderOptions);
    }
    return this._sender;
  }
  /**
   * Filter out envelopes with zero metric values to prevent exporting zero counts.
   * This ensures zero counts are observed for internal cleanup but not exported to Azure Monitor.
   * @param envelopes - Array of telemetry envelopes to filter
   * @returns Filtered array of envelopes with non-zero metric values
   */
  filterZeroValueMetrics(envelopes) {
    return envelopes.filter((envelope) => {
      if (envelope.data?.baseType === "MetricData") {
        const baseData = envelope.data?.baseData;
        const metrics = baseData.metrics;
        return metrics.some((metric) => metric.value !== 0);
      }
      return true;
    });
  }
  /**
   * Export Statsbeat metrics.
   */
  async export(metrics, resultCallback) {
    if (this._isShutdown) {
      setTimeout(() => resultCallback({ code: import_core.ExportResultCode.FAILED }), 0);
      return;
    }
    const envelopes = (0, import_metricUtils.resourceMetricsToEnvelope)(
      metrics,
      this.instrumentationKey,
      true
      // isStatsbeat flag passed to create a Statsbeat envelope.
    );
    const filteredEnvelopes = this.filterZeroValueMetrics(envelopes);
    await import_api.context.with((0, import_core.suppressTracing)(import_api.context.active()), async () => {
      const sender = await this._getSender();
      resultCallback(await sender.exportEnvelopes(filteredEnvelopes));
    });
  }
  /**
   * Shutdown AzureMonitorStatsbeatExporter.
   */
  async shutdown() {
    this._isShutdown = true;
    if (this._sender) {
      return this._sender.shutdown();
    }
  }
  /**
   * Force flush.
   */
  async forceFlush() {
    return Promise.resolve();
  }
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  AzureMonitorStatsbeatExporter
});
