// @azure/monitor-opentelemetry-exporter/dist/commonjs/export/statsbeat/statsbeatMetrics.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var statsbeatMetrics_exports = {};
__export(statsbeatMetrics_exports, {
  StatsbeatMetrics: () => StatsbeatMetrics
});
module.exports = __toCommonJS(statsbeatMetrics_exports);
var import_core_rest_pipeline = require("@azure/core-rest-pipeline");
var import_api = require("@opentelemetry/api");
var import_types = require("./types.js");
var import_node_os = __toESM(require("node:os"));
class StatsbeatMetrics {
  resourceProvider = import_types.StatsbeatResourceProvider.unknown;
  vmInfo = {};
  os = import_node_os.default.type();
  resourceIdentifier = "";
  async getResourceProvider() {
    this.resourceProvider = import_types.StatsbeatResourceProvider.unknown;
    if (process.env.AKS_ARM_NAMESPACE_ID || process.env.KUBERNETES_SERVICE_HOST) {
      this.resourceProvider = import_types.StatsbeatResourceProvider.aks;
      this.resourceIdentifier = process.env.AKS_ARM_NAMESPACE_ID || process.env.KUBERNETES_SERVICE_HOST || "";
    } else if (process.env.WEBSITE_SITE_NAME && !process.env.FUNCTIONS_WORKER_RUNTIME) {
      this.resourceProvider = import_types.StatsbeatResourceProvider.appsvc;
      this.resourceIdentifier = process.env.WEBSITE_SITE_NAME;
      if (process.env.WEBSITE_HOME_STAMPNAME) {
        this.resourceIdentifier += "/" + process.env.WEBSITE_HOME_STAMPNAME;
      }
    } else if (process.env.FUNCTIONS_WORKER_RUNTIME) {
      this.resourceProvider = import_types.StatsbeatResourceProvider.functions;
      if (process.env.WEBSITE_HOSTNAME) {
        this.resourceIdentifier = process.env.WEBSITE_HOSTNAME;
      }
    } else if (await this.getAzureComputeMetadata()) {
      this.resourceProvider = import_types.StatsbeatResourceProvider.vm;
      this.resourceIdentifier = this.vmInfo.id + "/" + this.vmInfo.subscriptionId;
      if (this.vmInfo.osType) {
        this.os = this.vmInfo.osType;
      }
    } else {
      this.resourceProvider = import_types.StatsbeatResourceProvider.unknown;
    }
  }
  async getAzureComputeMetadata() {
    const httpClient = (0, import_core_rest_pipeline.createDefaultHttpClient)();
    const method = "GET";
    const options = {
      url: `${import_types.AIMS_URI}?${import_types.AIMS_API_VERSION}&${import_types.AIMS_FORMAT}`,
      timeout: 5e3,
      // 5 seconds
      method,
      allowInsecureConnection: true
    };
    const request = (0, import_core_rest_pipeline.createPipelineRequest)(options);
    await httpClient.sendRequest(request).then((res) => {
      if (res.status === 200) {
        this.vmInfo.isVM = true;
        let virtualMachineData = "";
        res.on("data", (data) => {
          virtualMachineData += data;
        });
        res.on("end", () => {
          try {
            const data = JSON.parse(virtualMachineData);
            this.vmInfo.id = data["vmId"] || "";
            this.vmInfo.subscriptionId = data["subscriptionId"] || "";
            this.vmInfo.osType = data["osType"] || "";
          } catch (error) {
            import_api.diag.debug("Failed to parse JSON: ", error);
          }
        });
        return true;
      } else {
        return false;
      }
    }).catch(() => {
      return false;
    });
    return false;
  }
  getConnectionString(endpointUrl) {
    const currentEndpoint = endpointUrl;
    for (let i = 0; i < import_types.EU_ENDPOINTS.length; i++) {
      if (currentEndpoint.includes(import_types.EU_ENDPOINTS[i])) {
        return import_types.EU_CONNECTION_STRING;
      }
    }
    return import_types.NON_EU_CONNECTION_STRING;
  }
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  StatsbeatMetrics
});
