// @azure/monitor-opentelemetry-exporter/dist/commonjs/export/statsbeat/types.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var types_exports = {};
__export(types_exports, {
  AIMS_API_VERSION: () => AIMS_API_VERSION,
  AIMS_FORMAT: () => AIMS_FORMAT,
  AIMS_URI: () => AIMS_URI,
  AZURE_MONITOR_AUTO_ATTACH: () => AZURE_MONITOR_AUTO_ATTACH,
  AttachTypeName: () => AttachTypeName,
  CustomSDKStatsCounter: () => CustomSDKStatsCounter,
  CustomStatsbeatCounter: () => CustomStatsbeatCounter,
  CustomerSDKStats: () => CustomerSDKStats,
  CustomerStatsbeat: () => CustomerStatsbeat,
  DropCode: () => DropCode,
  DropReason: () => DropReason,
  EU_CONNECTION_STRING: () => EU_CONNECTION_STRING,
  EU_ENDPOINTS: () => EU_ENDPOINTS,
  ExceptionType: () => ExceptionType,
  MAX_STATSBEAT_FAILURES: () => MAX_STATSBEAT_FAILURES,
  NON_EU_CONNECTION_STRING: () => NON_EU_CONNECTION_STRING,
  NetworkStatsbeat: () => NetworkStatsbeat,
  RetryCode: () => RetryCode,
  RetryReason: () => RetryReason,
  STATSBEAT_LANGUAGE: () => STATSBEAT_LANGUAGE,
  StatsbeatCounter: () => StatsbeatCounter,
  StatsbeatFeatureType: () => StatsbeatFeatureType,
  StatsbeatResourceProvider: () => StatsbeatResourceProvider,
  TelemetryType: () => TelemetryType,
  isStatsbeatShutdownStatus: () => isStatsbeatShutdownStatus
});
module.exports = __toCommonJS(types_exports);
class NetworkStatsbeat {
  time;
  lastTime;
  endpoint;
  host;
  totalRequestCount;
  lastRequestCount;
  totalSuccessfulRequestCount;
  totalReadFailureCount;
  totalWriteFailureCount;
  totalFailedRequestCount;
  retryCount;
  exceptionCount;
  throttleCount;
  intervalRequestExecutionTime;
  lastIntervalRequestExecutionTime;
  averageRequestExecutionTime;
  constructor(endpoint, host) {
    this.endpoint = endpoint;
    this.host = host;
    this.totalRequestCount = 0;
    this.totalSuccessfulRequestCount = 0;
    this.totalReadFailureCount = 0;
    this.totalWriteFailureCount = 0;
    this.totalFailedRequestCount = [];
    this.retryCount = [];
    this.exceptionCount = [];
    this.throttleCount = [];
    this.intervalRequestExecutionTime = 0;
    this.lastIntervalRequestExecutionTime = 0;
    this.lastTime = +/* @__PURE__ */ new Date();
    this.lastRequestCount = 0;
    this.averageRequestExecutionTime = 0;
  }
}
class CustomerSDKStats {
  totalItemSuccessCount;
  // telemetry_type -> drop.code -> drop.reason -> success -> count
  // success can be true/false for request/dependency telemetry, or null for other types
  totalItemDropCount;
  // Nested Map structure: telemetry_type -> retry.code -> retry.reason -> count
  totalItemRetryCount;
  constructor() {
    this.totalItemSuccessCount = /* @__PURE__ */ new Map();
    this.totalItemDropCount = /* @__PURE__ */ new Map();
    this.totalItemRetryCount = /* @__PURE__ */ new Map();
  }
}
const CustomerStatsbeat = CustomerSDKStats;
const STATSBEAT_LANGUAGE = "node";
const AZURE_MONITOR_AUTO_ATTACH = "AZURE_MONITOR_AUTO_ATTACH";
const MAX_STATSBEAT_FAILURES = 3;
const StatsbeatResourceProvider = {
  appsvc: "appsvc",
  aks: "aks",
  functions: "functions",
  vm: "vm",
  unknown: "unknown"
};
var AttachTypeName = /* @__PURE__ */ ((AttachTypeName2) => {
  AttachTypeName2["INTEGRATED_AUTO"] = "IntegratedAuto";
  AttachTypeName2["MANUAL"] = "Manual";
  return AttachTypeName2;
})(AttachTypeName || {});
var StatsbeatCounter = /* @__PURE__ */ ((StatsbeatCounter2) => {
  StatsbeatCounter2["SUCCESS_COUNT"] = "Request_Success_Count";
  StatsbeatCounter2["FAILURE_COUNT"] = "Request_Failure_Count";
  StatsbeatCounter2["RETRY_COUNT"] = "Retry_Count";
  StatsbeatCounter2["THROTTLE_COUNT"] = "Throttle_Count";
  StatsbeatCounter2["EXCEPTION_COUNT"] = "Exception_Count";
  StatsbeatCounter2["AVERAGE_DURATION"] = "Request_Duration";
  StatsbeatCounter2["READ_FAILURE_COUNT"] = "Read_Failure_Count";
  StatsbeatCounter2["WRITE_FAILURE_COUNT"] = "Write_Failure_Count";
  StatsbeatCounter2["ATTACH"] = "Attach";
  StatsbeatCounter2["FEATURE"] = "Feature";
  return StatsbeatCounter2;
})(StatsbeatCounter || {});
var CustomSDKStatsCounter = /* @__PURE__ */ ((CustomSDKStatsCounter2) => {
  CustomSDKStatsCounter2["ITEM_SUCCESS_COUNT"] = "Item_Success_Count";
  CustomSDKStatsCounter2["ITEM_DROP_COUNT"] = "Item_Dropped_Count";
  CustomSDKStatsCounter2["ITEM_RETRY_COUNT"] = "Item_Retry_Count";
  return CustomSDKStatsCounter2;
})(CustomSDKStatsCounter || {});
const CustomStatsbeatCounter = CustomSDKStatsCounter;
const AIMS_URI = "http://169.254.169.254/metadata/instance/compute";
const AIMS_API_VERSION = "api-version=2017-12-01";
const AIMS_FORMAT = "format=json";
const NON_EU_CONNECTION_STRING = "InstrumentationKey=c4a29126-a7cb-47e5-b348-11414998b11e;IngestionEndpoint=https://westus-0.in.applicationinsights.azure.com";
const EU_CONNECTION_STRING = "InstrumentationKey=7dc56bab-3c0c-4e9f-9ebb-d1acadee8d0f;IngestionEndpoint=https://westeurope-5.in.applicationinsights.azure.com";
const EU_ENDPOINTS = [
  "westeurope",
  "northeurope",
  "francecentral",
  "francesouth",
  "germanywestcentral",
  "norwayeast",
  "norwaywest",
  "swedencentral",
  "switzerlandnorth",
  "switzerlandwest",
  "uksouth",
  "ukwest"
];
var TelemetryType = /* @__PURE__ */ ((TelemetryType2) => {
  TelemetryType2["AVAILABILITY"] = "AVAILABILITY";
  TelemetryType2["CUSTOM_EVENT"] = "CUSTOM_EVENT";
  TelemetryType2["CUSTOM_METRIC"] = "CUSTOM_METRIC";
  TelemetryType2["DEPENDENCY"] = "DEPENDENCY";
  TelemetryType2["EXCEPTION"] = "EXCEPTION";
  TelemetryType2["PAGE_VIEW"] = "PAGE_VIEW";
  TelemetryType2["PERFORMANCE_COUNTER"] = "PERFORMANCE_COUNTER";
  TelemetryType2["REQUEST"] = "REQUEST";
  TelemetryType2["TRACE"] = "TRACE";
  TelemetryType2["UNKNOWN"] = "UNKNOWN";
  return TelemetryType2;
})(TelemetryType || {});
var DropCode = /* @__PURE__ */ ((DropCode2) => {
  DropCode2["CLIENT_EXCEPTION"] = "CLIENT_EXCEPTION";
  DropCode2["CLIENT_READONLY"] = "CLIENT_READONLY";
  DropCode2["CLIENT_PERSISTENCE_CAPACITY"] = "CLIENT_PERSISTENCE_CAPACITY";
  DropCode2["CLIENT_STORAGE_DISABLED"] = "CLIENT_STORAGE_DISABLED";
  DropCode2["UNKNOWN"] = "UNKNOWN";
  return DropCode2;
})(DropCode || {});
var RetryCode = /* @__PURE__ */ ((RetryCode2) => {
  RetryCode2["CLIENT_EXCEPTION"] = "CLIENT_EXCEPTION";
  RetryCode2["CLIENT_TIMEOUT"] = "CLIENT_TIMEOUT";
  RetryCode2["UNKNOWN"] = "UNKNOWN";
  return RetryCode2;
})(RetryCode || {});
var StatsbeatFeatureType = /* @__PURE__ */ ((StatsbeatFeatureType2) => {
  StatsbeatFeatureType2[StatsbeatFeatureType2["FEATURE"] = 0] = "FEATURE";
  StatsbeatFeatureType2[StatsbeatFeatureType2["INSTRUMENTATION"] = 1] = "INSTRUMENTATION";
  return StatsbeatFeatureType2;
})(StatsbeatFeatureType || {});
var ExceptionType = /* @__PURE__ */ ((ExceptionType2) => {
  ExceptionType2["CLIENT_EXCEPTION"] = "Client exception";
  ExceptionType2["NETWORK_EXCEPTION"] = "Network exception";
  ExceptionType2["STORAGE_EXCEPTION"] = "Storage exception";
  ExceptionType2["TIMEOUT_EXCEPTION"] = "Timeout exception";
  return ExceptionType2;
})(ExceptionType || {});
var DropReason = /* @__PURE__ */ ((DropReason2) => {
  DropReason2["CLIENT_READONLY"] = "Client readonly";
  DropReason2["CLIENT_PERSISTENCE_CAPACITY"] = "Client persistence capacity";
  DropReason2["CLIENT_STORAGE_DISABLED"] = "Client local storage disabled";
  DropReason2["UNKNOWN"] = "Unknown reason";
  return DropReason2;
})(DropReason || {});
var RetryReason = /* @__PURE__ */ ((RetryReason2) => {
  RetryReason2["CLIENT_TIMEOUT"] = "Client timeout";
  RetryReason2["UNKNOWN"] = "Unknown reason";
  return RetryReason2;
})(RetryReason || {});
function isStatsbeatShutdownStatus(statusCode) {
  return statusCode === 401 || // Unauthorized
  statusCode === 403 || // Forbidden
  statusCode === 503;
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  AIMS_API_VERSION,
  AIMS_FORMAT,
  AIMS_URI,
  AZURE_MONITOR_AUTO_ATTACH,
  AttachTypeName,
  CustomSDKStatsCounter,
  CustomStatsbeatCounter,
  CustomerSDKStats,
  CustomerStatsbeat,
  DropCode,
  DropReason,
  EU_CONNECTION_STRING,
  EU_ENDPOINTS,
  ExceptionType,
  MAX_STATSBEAT_FAILURES,
  NON_EU_CONNECTION_STRING,
  NetworkStatsbeat,
  RetryCode,
  RetryReason,
  STATSBEAT_LANGUAGE,
  StatsbeatCounter,
  StatsbeatFeatureType,
  StatsbeatResourceProvider,
  TelemetryType,
  isStatsbeatShutdownStatus
});
