// @azure/monitor-opentelemetry-exporter/dist/esm/generated/models/models.js
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.
import { serializeRecord } from "../static-helpers/serialization/serialize-record.js";
export function telemetryItemSerializer(item) {
    return {
        ver: item["version"],
        name: item["name"],
        time: item["time"].toISOString(),
        sampleRate: item["sampleRate"],
        seq: item["sequence"],
        iKey: item["instrumentationKey"],
        tags: item["tags"],
        data: !item["data"] ? item["data"] : monitorBaseSerializer(item["data"]),
    };
}
export function monitorBaseSerializer(item) {
    return {
        baseType: item["baseType"],
        baseData: !item["baseData"] ? item["baseData"] : monitorDomainUnionSerializer(item["baseData"]),
    };
}
export function monitorDomainSerializer(item) {
    return { ...serializeRecord(item.additionalProperties ?? {}), ver: item["version"] };
}
export function monitorDomainDeserializer(item) {
    return {
        additionalProperties: serializeRecord(item, ["version", "kind"]),
        version: item["ver"],
        kind: item["kind"],
    };
}
export function monitorDomainUnionSerializer(item) {
    switch (item.kind) {
        case "AvailabilityData":
            return availabilityDataSerializer(item);
        case "EventData":
            return telemetryEventDataSerializer(item);
        case "ExceptionData":
            return telemetryExceptionDataSerializer(item);
        case "MessageData":
            return messageDataSerializer(item);
        case "MetricsData":
            return metricsDataSerializer(item);
        case "PageViewData":
            return pageViewDataSerializer(item);
        case "PageViewPerfData":
            return pageViewPerfDataSerializer(item);
        case "RemoteDependencyData":
            return remoteDependencyDataSerializer(item);
        case "RequestData":
            return requestDataSerializer(item);
        default:
            return monitorDomainSerializer(item);
    }
}
export function monitorDomainUnionDeserializer(item) {
    switch (item.kind) {
        case "AvailabilityData":
            return availabilityDataDeserializer(item);
        case "EventData":
            return telemetryEventDataDeserializer(item);
        case "ExceptionData":
            return telemetryExceptionDataDeserializer(item);
        case "MessageData":
            return messageDataDeserializer(item);
        case "MetricsData":
            return metricsDataDeserializer(item);
        case "PageViewData":
            return pageViewDataDeserializer(item);
        case "PageViewPerfData":
            return pageViewPerfDataDeserializer(item);
        case "RemoteDependencyData":
            return remoteDependencyDataDeserializer(item);
        case "RequestData":
            return requestDataDeserializer(item);
        default:
            return monitorDomainDeserializer(item);
    }
}
/** Identifies the specific telemetry data type. */
export var KnownMonitorDomainKind;
(function (KnownMonitorDomainKind) {
    /** AvailabilityData type. */
    KnownMonitorDomainKind["AvailabilityData"] = "AvailabilityData";
    /** EventData type. */
    KnownMonitorDomainKind["EventData"] = "EventData";
    /** ExceptionData type. */
    KnownMonitorDomainKind["ExceptionData"] = "ExceptionData";
    /** MessageData type. */
    KnownMonitorDomainKind["MessageData"] = "MessageData";
    /** MetricsData type. */
    KnownMonitorDomainKind["MetricsData"] = "MetricsData";
    /** PageViewData type. */
    KnownMonitorDomainKind["PageViewData"] = "PageViewData";
    /** PageViewPerfData type. */
    KnownMonitorDomainKind["PageViewPerfData"] = "PageViewPerfData";
    /** RemoteDependencyData type. */
    KnownMonitorDomainKind["RemoteDependencyData"] = "RemoteDependencyData";
    /** RequestData type. */
    KnownMonitorDomainKind["RequestData"] = "RequestData";
})(KnownMonitorDomainKind || (KnownMonitorDomainKind = {}));
export function availabilityDataSerializer(item) {
    return {
        ...serializeRecord(item.additionalProperties ?? {}),
        ver: item["version"],
        id: item["id"],
        name: item["name"],
        duration: item["duration"],
        success: item["success"],
        runLocation: item["runLocation"],
        message: item["message"],
        properties: item["properties"],
        measurements: item["measurements"],
    };
}
export function availabilityDataDeserializer(item) {
    return {
        additionalProperties: serializeRecord(item, [
            "version",
            "kind",
            "id",
            "name",
            "duration",
            "success",
            "runLocation",
            "message",
            "properties",
            "measurements",
        ]),
        version: item["ver"],
        kind: item["kind"],
        id: item["id"],
        name: item["name"],
        duration: item["duration"],
        success: item["success"],
        runLocation: item["runLocation"],
        message: item["message"],
        properties: !item["properties"]
            ? item["properties"]
            : Object.fromEntries(Object.entries(item["properties"]).map(([k, p]) => [k, p])),
        measurements: !item["measurements"]
            ? item["measurements"]
            : Object.fromEntries(Object.entries(item["measurements"]).map(([k, p]) => [k, p])),
    };
}
export function telemetryEventDataSerializer(item) {
    return {
        ...serializeRecord(item.additionalProperties ?? {}),
        ver: item["version"],
        name: item["name"],
        properties: item["properties"],
        measurements: item["measurements"],
    };
}
export function telemetryEventDataDeserializer(item) {
    return {
        additionalProperties: serializeRecord(item, [
            "version",
            "kind",
            "name",
            "properties",
            "measurements",
        ]),
        version: item["ver"],
        kind: item["kind"],
        name: item["name"],
        properties: !item["properties"]
            ? item["properties"]
            : Object.fromEntries(Object.entries(item["properties"]).map(([k, p]) => [k, p])),
        measurements: !item["measurements"]
            ? item["measurements"]
            : Object.fromEntries(Object.entries(item["measurements"]).map(([k, p]) => [k, p])),
    };
}
export function telemetryExceptionDataSerializer(item) {
    return {
        ...serializeRecord(item.additionalProperties ?? {}),
        ver: item["version"],
        exceptions: telemetryExceptionDetailsArraySerializer(item["exceptions"]),
        severityLevel: item["severityLevel"],
        problemId: item["problemId"],
        properties: item["properties"],
        measurements: item["measurements"],
    };
}
export function telemetryExceptionDataDeserializer(item) {
    return {
        additionalProperties: serializeRecord(item, [
            "version",
            "kind",
            "exceptions",
            "severityLevel",
            "problemId",
            "properties",
            "measurements",
        ]),
        version: item["ver"],
        kind: item["kind"],
        exceptions: telemetryExceptionDetailsArrayDeserializer(item["exceptions"]),
        severityLevel: item["severityLevel"],
        problemId: item["problemId"],
        properties: !item["properties"]
            ? item["properties"]
            : Object.fromEntries(Object.entries(item["properties"]).map(([k, p]) => [k, p])),
        measurements: !item["measurements"]
            ? item["measurements"]
            : Object.fromEntries(Object.entries(item["measurements"]).map(([k, p]) => [k, p])),
    };
}
export function telemetryExceptionDetailsArraySerializer(result) {
    return result.map((item) => {
        return telemetryExceptionDetailsSerializer(item);
    });
}
export function telemetryExceptionDetailsArrayDeserializer(result) {
    return result.map((item) => {
        return telemetryExceptionDetailsDeserializer(item);
    });
}
export function telemetryExceptionDetailsSerializer(item) {
    return {
        id: item["id"],
        outerId: item["outerId"],
        typeName: item["typeName"],
        message: item["message"],
        hasFullStack: item["hasFullStack"],
        stack: item["stack"],
        parsedStack: !item["parsedStack"]
            ? item["parsedStack"]
            : stackFrameArraySerializer(item["parsedStack"]),
    };
}
export function telemetryExceptionDetailsDeserializer(item) {
    return {
        id: item["id"],
        outerId: item["outerId"],
        typeName: item["typeName"],
        message: item["message"],
        hasFullStack: item["hasFullStack"],
        stack: item["stack"],
        parsedStack: !item["parsedStack"]
            ? item["parsedStack"]
            : stackFrameArrayDeserializer(item["parsedStack"]),
    };
}
export function stackFrameArraySerializer(result) {
    return result.map((item) => {
        return stackFrameSerializer(item);
    });
}
export function stackFrameArrayDeserializer(result) {
    return result.map((item) => {
        return stackFrameDeserializer(item);
    });
}
export function stackFrameSerializer(item) {
    return {
        level: item["level"],
        method: item["method"],
        assembly: item["assembly"],
        fileName: item["fileName"],
        line: item["line"],
    };
}
export function stackFrameDeserializer(item) {
    return {
        level: item["level"],
        method: item["method"],
        assembly: item["assembly"],
        fileName: item["fileName"],
        line: item["line"],
    };
}
/** Defines the level of severity for the event. */
export var KnownSeverityLevel;
(function (KnownSeverityLevel) {
    /** Verbose level. */
    KnownSeverityLevel["Verbose"] = "Verbose";
    /** Information level. */
    KnownSeverityLevel["Information"] = "Information";
    /** Warning level. */
    KnownSeverityLevel["Warning"] = "Warning";
    /** Error level. */
    KnownSeverityLevel["Error"] = "Error";
    /** Critical level. */
    KnownSeverityLevel["Critical"] = "Critical";
})(KnownSeverityLevel || (KnownSeverityLevel = {}));
export function messageDataSerializer(item) {
    return {
        ...serializeRecord(item.additionalProperties ?? {}),
        ver: item["version"],
        message: item["message"],
        severityLevel: item["severityLevel"],
        properties: item["properties"],
        measurements: item["measurements"],
    };
}
export function messageDataDeserializer(item) {
    return {
        additionalProperties: serializeRecord(item, [
            "version",
            "kind",
            "message",
            "severityLevel",
            "properties",
            "measurements",
        ]),
        version: item["ver"],
        kind: item["kind"],
        message: item["message"],
        severityLevel: item["severityLevel"],
        properties: !item["properties"]
            ? item["properties"]
            : Object.fromEntries(Object.entries(item["properties"]).map(([k, p]) => [k, p])),
        measurements: !item["measurements"]
            ? item["measurements"]
            : Object.fromEntries(Object.entries(item["measurements"]).map(([k, p]) => [k, p])),
    };
}
export function metricsDataSerializer(item) {
    return {
        ...serializeRecord(item.additionalProperties ?? {}),
        ver: item["version"],
        metrics: metricDataPointArraySerializer(item["metrics"]),
        properties: item["properties"],
    };
}
export function metricsDataDeserializer(item) {
    return {
        additionalProperties: serializeRecord(item, ["version", "kind", "metrics", "properties"]),
        version: item["ver"],
        kind: item["kind"],
        metrics: metricDataPointArrayDeserializer(item["metrics"]),
        properties: !item["properties"]
            ? item["properties"]
            : Object.fromEntries(Object.entries(item["properties"]).map(([k, p]) => [k, p])),
    };
}
export function metricDataPointArraySerializer(result) {
    return result.map((item) => {
        return metricDataPointSerializer(item);
    });
}
export function metricDataPointArrayDeserializer(result) {
    return result.map((item) => {
        return metricDataPointDeserializer(item);
    });
}
export function metricDataPointSerializer(item) {
    return {
        ns: item["namespace"],
        name: item["name"],
        kind: item["dataPointType"],
        value: item["value"],
        count: item["count"],
        min: item["min"],
        max: item["max"],
        stdDev: item["stdDev"],
    };
}
export function metricDataPointDeserializer(item) {
    return {
        namespace: item["ns"],
        name: item["name"],
        dataPointType: item["kind"],
        value: item["value"],
        count: item["count"],
        min: item["min"],
        max: item["max"],
        stdDev: item["stdDev"],
    };
}
/** Type of the metric data. */
export var KnownDataPointType;
(function (KnownDataPointType) {
    /** Single measurement. */
    KnownDataPointType["Measurement"] = "Measurement";
    /** Aggregated value. */
    KnownDataPointType["Aggregation"] = "Aggregation";
})(KnownDataPointType || (KnownDataPointType = {}));
export function pageViewDataSerializer(item) {
    return {
        ...serializeRecord(item.additionalProperties ?? {}),
        ver: item["version"],
        id: item["id"],
        name: item["name"],
        url: item["url"],
        duration: item["duration"],
        referredUri: item["referredUri"],
        properties: item["properties"],
        measurements: item["measurements"],
    };
}
export function pageViewDataDeserializer(item) {
    return {
        additionalProperties: serializeRecord(item, [
            "version",
            "kind",
            "id",
            "name",
            "url",
            "duration",
            "referredUri",
            "properties",
            "measurements",
        ]),
        version: item["ver"],
        kind: item["kind"],
        id: item["id"],
        name: item["name"],
        url: item["url"],
        duration: item["duration"],
        referredUri: item["referredUri"],
        properties: !item["properties"]
            ? item["properties"]
            : Object.fromEntries(Object.entries(item["properties"]).map(([k, p]) => [k, p])),
        measurements: !item["measurements"]
            ? item["measurements"]
            : Object.fromEntries(Object.entries(item["measurements"]).map(([k, p]) => [k, p])),
    };
}
export function pageViewPerfDataSerializer(item) {
    return {
        ...serializeRecord(item.additionalProperties ?? {}),
        ver: item["version"],
        id: item["id"],
        name: item["name"],
        url: item["url"],
        duration: item["duration"],
        perfTotal: item["perfTotal"],
        networkConnect: item["networkConnect"],
        sentRequest: item["sentRequest"],
        receivedResponse: item["receivedResponse"],
        domProcessing: item["domProcessing"],
        properties: item["properties"],
        measurements: item["measurements"],
    };
}
export function pageViewPerfDataDeserializer(item) {
    return {
        additionalProperties: serializeRecord(item, [
            "version",
            "kind",
            "id",
            "name",
            "url",
            "duration",
            "perfTotal",
            "networkConnect",
            "sentRequest",
            "receivedResponse",
            "domProcessing",
            "properties",
            "measurements",
        ]),
        version: item["ver"],
        kind: item["kind"],
        id: item["id"],
        name: item["name"],
        url: item["url"],
        duration: item["duration"],
        perfTotal: item["perfTotal"],
        networkConnect: item["networkConnect"],
        sentRequest: item["sentRequest"],
        receivedResponse: item["receivedResponse"],
        domProcessing: item["domProcessing"],
        properties: !item["properties"]
            ? item["properties"]
            : Object.fromEntries(Object.entries(item["properties"]).map(([k, p]) => [k, p])),
        measurements: !item["measurements"]
            ? item["measurements"]
            : Object.fromEntries(Object.entries(item["measurements"]).map(([k, p]) => [k, p])),
    };
}
export function remoteDependencyDataSerializer(item) {
    return {
        ...serializeRecord(item.additionalProperties ?? {}),
        ver: item["version"],
        id: item["id"],
        name: item["name"],
        resultCode: item["resultCode"],
        data: item["data"],
        type: item["type"],
        target: item["target"],
        duration: item["duration"],
        success: item["success"],
        properties: item["properties"],
        measurements: item["measurements"],
    };
}
export function remoteDependencyDataDeserializer(item) {
    return {
        additionalProperties: serializeRecord(item, [
            "version",
            "kind",
            "id",
            "name",
            "resultCode",
            "data",
            "type",
            "target",
            "duration",
            "success",
            "properties",
            "measurements",
        ]),
        version: item["ver"],
        kind: item["kind"],
        id: item["id"],
        name: item["name"],
        resultCode: item["resultCode"],
        data: item["data"],
        type: item["type"],
        target: item["target"],
        duration: item["duration"],
        success: item["success"],
        properties: !item["properties"]
            ? item["properties"]
            : Object.fromEntries(Object.entries(item["properties"]).map(([k, p]) => [k, p])),
        measurements: !item["measurements"]
            ? item["measurements"]
            : Object.fromEntries(Object.entries(item["measurements"]).map(([k, p]) => [k, p])),
    };
}
export function requestDataSerializer(item) {
    return {
        ...serializeRecord(item.additionalProperties ?? {}),
        ver: item["version"],
        id: item["id"],
        name: item["name"],
        duration: item["duration"],
        success: item["success"],
        responseCode: item["responseCode"],
        source: item["source"],
        url: item["url"],
        properties: item["properties"],
        measurements: item["measurements"],
    };
}
export function requestDataDeserializer(item) {
    return {
        additionalProperties: serializeRecord(item, [
            "version",
            "kind",
            "id",
            "name",
            "duration",
            "success",
            "responseCode",
            "source",
            "url",
            "properties",
            "measurements",
        ]),
        version: item["ver"],
        kind: item["kind"],
        id: item["id"],
        name: item["name"],
        duration: item["duration"],
        success: item["success"],
        responseCode: item["responseCode"],
        source: item["source"],
        url: item["url"],
        properties: !item["properties"]
            ? item["properties"]
            : Object.fromEntries(Object.entries(item["properties"]).map(([k, p]) => [k, p])),
        measurements: !item["measurements"]
            ? item["measurements"]
            : Object.fromEntries(Object.entries(item["measurements"]).map(([k, p]) => [k, p])),
    };
}
export function trackResponseDeserializer(item) {
    return {
        itemsReceived: item["itemsReceived"],
        itemsAccepted: item["itemsAccepted"],
        errors: !item["errors"]
            ? item["errors"]
            : telemetryErrorDetailsArrayDeserializer(item["errors"]),
    };
}
export function telemetryErrorDetailsArrayDeserializer(result) {
    return result.map((item) => {
        return telemetryErrorDetailsDeserializer(item);
    });
}
export function telemetryErrorDetailsDeserializer(item) {
    return {
        index: item["index"],
        statusCode: item["statusCode"],
        message: item["message"],
    };
}
/** The context tag keys. */
export var KnownContextTagKeys;
(function (KnownContextTagKeys) {
    /** Application version. */
    KnownContextTagKeys["AiApplicationVer"] = "ai.application.ver";
    /** Device ID. */
    KnownContextTagKeys["AiDeviceId"] = "ai.device.id";
    /** Device locale. */
    KnownContextTagKeys["AiDeviceLocale"] = "ai.device.locale";
    /** Device model. */
    KnownContextTagKeys["AiDeviceModel"] = "ai.device.model";
    /** Device OEM name. */
    KnownContextTagKeys["AiDeviceOemName"] = "ai.device.oemName";
    /** Device OS version. */
    KnownContextTagKeys["AiDeviceOSVersion"] = "ai.device.osVersion";
    /** Device type. */
    KnownContextTagKeys["AiDeviceType"] = "ai.device.type";
    /** Location IP. */
    KnownContextTagKeys["AiLocationIp"] = "ai.location.ip";
    /** Location country. */
    KnownContextTagKeys["AiLocationCountry"] = "ai.location.country";
    /** Location province. */
    KnownContextTagKeys["AiLocationProvince"] = "ai.location.province";
    /** Location city. */
    KnownContextTagKeys["AiLocationCity"] = "ai.location.city";
    /** Operation ID. */
    KnownContextTagKeys["AiOperationId"] = "ai.operation.id";
    /** Operation name. */
    KnownContextTagKeys["AiOperationName"] = "ai.operation.name";
    /** Operation parent ID. */
    KnownContextTagKeys["AiOperationParentId"] = "ai.operation.parentId";
    /** Operation synthetic source. */
    KnownContextTagKeys["AiOperationSyntheticSource"] = "ai.operation.syntheticSource";
    /** Operation correlation vector. */
    KnownContextTagKeys["AiOperationCorrelationVector"] = "ai.operation.correlationVector";
    /** Session ID. */
    KnownContextTagKeys["AiSessionId"] = "ai.session.id";
    /** If session is the first one. */
    KnownContextTagKeys["AiSessionIsFirst"] = "ai.session.isFirst";
    /** User account ID. */
    KnownContextTagKeys["AiUserAccountId"] = "ai.user.accountId";
    /** User ID. */
    KnownContextTagKeys["AiUserId"] = "ai.user.id";
    /** Authenticated user ID. */
    KnownContextTagKeys["AiUserAuthUserId"] = "ai.user.authUserId";
    /** Cloud role. */
    KnownContextTagKeys["AiCloudRole"] = "ai.cloud.role";
    /** Cloud role version. */
    KnownContextTagKeys["AiCloudRoleVer"] = "ai.cloud.roleVer";
    /** Cloud role instance. */
    KnownContextTagKeys["AiCloudRoleInstance"] = "ai.cloud.roleInstance";
    /** Cloud location. */
    KnownContextTagKeys["AiCloudLocation"] = "ai.cloud.location";
    /** Internal SDK version. */
    KnownContextTagKeys["AiInternalSdkVersion"] = "ai.internal.sdkVersion";
    /** Internal agent version. */
    KnownContextTagKeys["AiInternalAgentVersion"] = "ai.internal.agentVersion";
    /** Internal node name. */
    KnownContextTagKeys["AiInternalNodeName"] = "ai.internal.nodeName";
})(KnownContextTagKeys || (KnownContextTagKeys = {}));
export function telemetryItemArraySerializer(result) {
    return result.map((item) => {
        return telemetryItemSerializer(item);
    });
}
