// @azure/monitor-opentelemetry/dist/commonjs/generated/models/models.js
"use strict";
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.
Object.defineProperty(exports, "__esModule", { value: true });
exports.KnownVersions = void 0;
exports.monitoringDataPointSerializer = monitoringDataPointSerializer;
exports.metricPointArraySerializer = metricPointArraySerializer;
exports.metricPointSerializer = metricPointSerializer;
exports.documentIngressUnionArraySerializer = documentIngressUnionArraySerializer;
exports.documentIngressSerializer = documentIngressSerializer;
exports.documentIngressUnionSerializer = documentIngressUnionSerializer;
exports.keyValuePairStringStringArraySerializer = keyValuePairStringStringArraySerializer;
exports.keyValuePairStringStringSerializer = keyValuePairStringStringSerializer;
exports.requestSerializer = requestSerializer;
exports.remoteDependencySerializer = remoteDependencySerializer;
exports.exceptionSerializer = exceptionSerializer;
exports.eventSerializer = eventSerializer;
exports.traceSerializer = traceSerializer;
exports.processCpuDataArraySerializer = processCpuDataArraySerializer;
exports.processCpuDataSerializer = processCpuDataSerializer;
exports.collectionConfigurationErrorArraySerializer = collectionConfigurationErrorArraySerializer;
exports.collectionConfigurationErrorSerializer = collectionConfigurationErrorSerializer;
exports.collectionConfigurationInfoSerializer = collectionConfigurationInfoSerializer;
exports.collectionConfigurationInfoDeserializer = collectionConfigurationInfoDeserializer;
exports.derivedMetricInfoArraySerializer = derivedMetricInfoArraySerializer;
exports.derivedMetricInfoArrayDeserializer = derivedMetricInfoArrayDeserializer;
exports.derivedMetricInfoSerializer = derivedMetricInfoSerializer;
exports.derivedMetricInfoDeserializer = derivedMetricInfoDeserializer;
exports.filterConjunctionGroupInfoArraySerializer = filterConjunctionGroupInfoArraySerializer;
exports.filterConjunctionGroupInfoArrayDeserializer = filterConjunctionGroupInfoArrayDeserializer;
exports.filterConjunctionGroupInfoSerializer = filterConjunctionGroupInfoSerializer;
exports.filterConjunctionGroupInfoDeserializer = filterConjunctionGroupInfoDeserializer;
exports.filterInfoArraySerializer = filterInfoArraySerializer;
exports.filterInfoArrayDeserializer = filterInfoArrayDeserializer;
exports.filterInfoSerializer = filterInfoSerializer;
exports.filterInfoDeserializer = filterInfoDeserializer;
exports.documentStreamInfoArraySerializer = documentStreamInfoArraySerializer;
exports.documentStreamInfoArrayDeserializer = documentStreamInfoArrayDeserializer;
exports.documentStreamInfoSerializer = documentStreamInfoSerializer;
exports.documentStreamInfoDeserializer = documentStreamInfoDeserializer;
exports.documentFilterConjunctionGroupInfoArraySerializer = documentFilterConjunctionGroupInfoArraySerializer;
exports.documentFilterConjunctionGroupInfoArrayDeserializer = documentFilterConjunctionGroupInfoArrayDeserializer;
exports.documentFilterConjunctionGroupInfoSerializer = documentFilterConjunctionGroupInfoSerializer;
exports.documentFilterConjunctionGroupInfoDeserializer = documentFilterConjunctionGroupInfoDeserializer;
exports.quotaConfigurationInfoSerializer = quotaConfigurationInfoSerializer;
exports.quotaConfigurationInfoDeserializer = quotaConfigurationInfoDeserializer;
exports.serviceErrorDeserializer = serviceErrorDeserializer;
exports.monitoringDataPointArraySerializer = monitoringDataPointArraySerializer;
function monitoringDataPointSerializer(item) {
    return {
        Version: item["version"],
        InvariantVersion: item["invariantVersion"],
        Instance: item["instance"],
        RoleName: item["roleName"],
        MachineName: item["machineName"],
        StreamId: item["streamId"],
        Timestamp: !item["timestamp"] ? item["timestamp"] : item["timestamp"].toISOString(),
        TransmissionTime: !item["transmissionTime"]
            ? item["transmissionTime"]
            : item["transmissionTime"].toISOString(),
        IsWebApp: item["isWebApp"],
        PerformanceCollectionSupported: item["performanceCollectionSupported"],
        Metrics: !item["metrics"] ? item["metrics"] : metricPointArraySerializer(item["metrics"]),
        Documents: !item["documents"]
            ? item["documents"]
            : documentIngressUnionArraySerializer(item["documents"]),
        TopCpuProcesses: !item["topCpuProcesses"]
            ? item["topCpuProcesses"]
            : processCpuDataArraySerializer(item["topCpuProcesses"]),
        CollectionConfigurationErrors: !item["collectionConfigurationErrors"]
            ? item["collectionConfigurationErrors"]
            : collectionConfigurationErrorArraySerializer(item["collectionConfigurationErrors"]),
    };
}
function metricPointArraySerializer(result) {
    return result.map((item) => {
        return metricPointSerializer(item);
    });
}
function metricPointSerializer(item) {
    return { Name: item["name"], Value: item["value"], Weight: item["weight"] };
}
function documentIngressUnionArraySerializer(result) {
    return result.map((item) => {
        return documentIngressUnionSerializer(item);
    });
}
function documentIngressSerializer(item) {
    return {
        DocumentType: item["documentType"],
        DocumentStreamIds: !item["documentStreamIds"]
            ? item["documentStreamIds"]
            : item["documentStreamIds"].map((p) => {
                return p;
            }),
        Properties: !item["properties"]
            ? item["properties"]
            : keyValuePairStringStringArraySerializer(item["properties"]),
    };
}
function documentIngressUnionSerializer(item) {
    switch (item.documentType) {
        case "Request":
            return requestSerializer(item);
        case "RemoteDependency":
            return remoteDependencySerializer(item);
        case "Exception":
            return exceptionSerializer(item);
        case "Event":
            return eventSerializer(item);
        case "Trace":
            return traceSerializer(item);
        default:
            return documentIngressSerializer(item);
    }
}
function keyValuePairStringStringArraySerializer(result) {
    return result.map((item) => {
        return keyValuePairStringStringSerializer(item);
    });
}
function keyValuePairStringStringSerializer(item) {
    return { key: item["key"], value: item["value"] };
}
function requestSerializer(item) {
    return {
        DocumentType: item["documentType"],
        DocumentStreamIds: !item["documentStreamIds"]
            ? item["documentStreamIds"]
            : item["documentStreamIds"].map((p) => {
                return p;
            }),
        Properties: !item["properties"]
            ? item["properties"]
            : keyValuePairStringStringArraySerializer(item["properties"]),
        Name: item["name"],
        Url: item["url"],
        ResponseCode: item["responseCode"],
        Duration: item["duration"],
    };
}
function remoteDependencySerializer(item) {
    return {
        DocumentType: item["documentType"],
        DocumentStreamIds: !item["documentStreamIds"]
            ? item["documentStreamIds"]
            : item["documentStreamIds"].map((p) => {
                return p;
            }),
        Properties: !item["properties"]
            ? item["properties"]
            : keyValuePairStringStringArraySerializer(item["properties"]),
        Name: item["name"],
        CommandName: item["commandName"],
        ResultCode: item["resultCode"],
        Duration: item["duration"],
    };
}
function exceptionSerializer(item) {
    return {
        DocumentType: item["documentType"],
        DocumentStreamIds: !item["documentStreamIds"]
            ? item["documentStreamIds"]
            : item["documentStreamIds"].map((p) => {
                return p;
            }),
        Properties: !item["properties"]
            ? item["properties"]
            : keyValuePairStringStringArraySerializer(item["properties"]),
        ExceptionType: item["exceptionType"],
        ExceptionMessage: item["exceptionMessage"],
    };
}
function eventSerializer(item) {
    return {
        DocumentType: item["documentType"],
        DocumentStreamIds: !item["documentStreamIds"]
            ? item["documentStreamIds"]
            : item["documentStreamIds"].map((p) => {
                return p;
            }),
        Properties: !item["properties"]
            ? item["properties"]
            : keyValuePairStringStringArraySerializer(item["properties"]),
        Name: item["name"],
    };
}
function traceSerializer(item) {
    return {
        DocumentType: item["documentType"],
        DocumentStreamIds: !item["documentStreamIds"]
            ? item["documentStreamIds"]
            : item["documentStreamIds"].map((p) => {
                return p;
            }),
        Properties: !item["properties"]
            ? item["properties"]
            : keyValuePairStringStringArraySerializer(item["properties"]),
        Message: item["message"],
    };
}
function processCpuDataArraySerializer(result) {
    return result.map((item) => {
        return processCpuDataSerializer(item);
    });
}
function processCpuDataSerializer(item) {
    return { ProcessName: item["processName"], CpuPercentage: item["cpuPercentage"] };
}
function collectionConfigurationErrorArraySerializer(result) {
    return result.map((item) => {
        return collectionConfigurationErrorSerializer(item);
    });
}
function collectionConfigurationErrorSerializer(item) {
    return {
        CollectionConfigurationErrorType: item["collectionConfigurationErrorType"],
        Message: item["message"],
        FullException: item["fullException"],
        Data: keyValuePairStringStringArraySerializer(item["data"]),
    };
}
function collectionConfigurationInfoSerializer(item) {
    return {
        ETag: item["eTag"],
        Metrics: derivedMetricInfoArraySerializer(item["metrics"]),
        DocumentStreams: documentStreamInfoArraySerializer(item["documentStreams"]),
        QuotaInfo: !item["quotaInfo"]
            ? item["quotaInfo"]
            : quotaConfigurationInfoSerializer(item["quotaInfo"]),
    };
}
function collectionConfigurationInfoDeserializer(item) {
    return {
        eTag: item["ETag"],
        metrics: derivedMetricInfoArrayDeserializer(item["Metrics"]),
        documentStreams: documentStreamInfoArrayDeserializer(item["DocumentStreams"]),
        quotaInfo: !item["QuotaInfo"]
            ? item["QuotaInfo"]
            : quotaConfigurationInfoDeserializer(item["QuotaInfo"]),
    };
}
function derivedMetricInfoArraySerializer(result) {
    return result.map((item) => {
        return derivedMetricInfoSerializer(item);
    });
}
function derivedMetricInfoArrayDeserializer(result) {
    return result.map((item) => {
        return derivedMetricInfoDeserializer(item);
    });
}
function derivedMetricInfoSerializer(item) {
    return {
        Id: item["id"],
        TelemetryType: item["telemetryType"],
        FilterGroups: filterConjunctionGroupInfoArraySerializer(item["filterGroups"]),
        Projection: item["projection"],
        Aggregation: item["aggregation"],
        BackEndAggregation: item["backEndAggregation"],
    };
}
function derivedMetricInfoDeserializer(item) {
    return {
        id: item["Id"],
        telemetryType: item["TelemetryType"],
        filterGroups: filterConjunctionGroupInfoArrayDeserializer(item["FilterGroups"]),
        projection: item["Projection"],
        aggregation: item["Aggregation"],
        backEndAggregation: item["BackEndAggregation"],
    };
}
function filterConjunctionGroupInfoArraySerializer(result) {
    return result.map((item) => {
        return filterConjunctionGroupInfoSerializer(item);
    });
}
function filterConjunctionGroupInfoArrayDeserializer(result) {
    return result.map((item) => {
        return filterConjunctionGroupInfoDeserializer(item);
    });
}
function filterConjunctionGroupInfoSerializer(item) {
    return { Filters: filterInfoArraySerializer(item["filters"]) };
}
function filterConjunctionGroupInfoDeserializer(item) {
    return {
        filters: filterInfoArrayDeserializer(item["Filters"]),
    };
}
function filterInfoArraySerializer(result) {
    return result.map((item) => {
        return filterInfoSerializer(item);
    });
}
function filterInfoArrayDeserializer(result) {
    return result.map((item) => {
        return filterInfoDeserializer(item);
    });
}
function filterInfoSerializer(item) {
    return {
        FieldName: item["fieldName"],
        Predicate: item["predicate"],
        Comparand: item["comparand"],
    };
}
function filterInfoDeserializer(item) {
    return {
        fieldName: item["FieldName"],
        predicate: item["Predicate"],
        comparand: item["Comparand"],
    };
}
function documentStreamInfoArraySerializer(result) {
    return result.map((item) => {
        return documentStreamInfoSerializer(item);
    });
}
function documentStreamInfoArrayDeserializer(result) {
    return result.map((item) => {
        return documentStreamInfoDeserializer(item);
    });
}
function documentStreamInfoSerializer(item) {
    return {
        Id: item["id"],
        DocumentFilterGroups: documentFilterConjunctionGroupInfoArraySerializer(item["documentFilterGroups"]),
    };
}
function documentStreamInfoDeserializer(item) {
    return {
        id: item["Id"],
        documentFilterGroups: documentFilterConjunctionGroupInfoArrayDeserializer(item["DocumentFilterGroups"]),
    };
}
function documentFilterConjunctionGroupInfoArraySerializer(result) {
    return result.map((item) => {
        return documentFilterConjunctionGroupInfoSerializer(item);
    });
}
function documentFilterConjunctionGroupInfoArrayDeserializer(result) {
    return result.map((item) => {
        return documentFilterConjunctionGroupInfoDeserializer(item);
    });
}
function documentFilterConjunctionGroupInfoSerializer(item) {
    return {
        TelemetryType: item["telemetryType"],
        Filters: filterConjunctionGroupInfoSerializer(item["filters"]),
    };
}
function documentFilterConjunctionGroupInfoDeserializer(item) {
    return {
        telemetryType: item["TelemetryType"],
        filters: filterConjunctionGroupInfoDeserializer(item["Filters"]),
    };
}
function quotaConfigurationInfoSerializer(item) {
    return {
        InitialQuota: item["initialQuota"],
        MaxQuota: item["maxQuota"],
        QuotaAccrualRatePerSec: item["quotaAccrualRatePerSec"],
    };
}
function quotaConfigurationInfoDeserializer(item) {
    return {
        initialQuota: item["InitialQuota"],
        maxQuota: item["MaxQuota"],
        quotaAccrualRatePerSec: item["QuotaAccrualRatePerSec"],
    };
}
function serviceErrorDeserializer(item) {
    return {
        requestId: item["RequestId"],
        responseDateTime: item["ResponseDateTime"],
        code: item["Code"],
        message: item["Message"],
        exception: item["Exception"],
    };
}
/** Live Metrics service versions. */
var KnownVersions;
(function (KnownVersions) {
    /** The 2024-04-01-preview version of the Live Metrics service. */
    KnownVersions["V20240401Preview"] = "2024-04-01-preview";
})(KnownVersions || (exports.KnownVersions = KnownVersions = {}));
function monitoringDataPointArraySerializer(result) {
    return result.map((item) => {
        return monitoringDataPointSerializer(item);
    });
}
