// @azure/monitor-opentelemetry-exporter/dist/commonjs/index.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var src_exports = {};
__export(src_exports, {
  AI_OPERATION_NAME: () => import_Constants2.AI_OPERATION_NAME,
  ApplicationInsightsSampler: () => import_percentageSampler.ApplicationInsightsSampler,
  AzureMonitorBaseExporter: () => import_base.AzureMonitorBaseExporter,
  AzureMonitorLogExporter: () => import_log.AzureMonitorLogExporter,
  AzureMonitorMetricExporter: () => import_metric.AzureMonitorMetricExporter,
  AzureMonitorTraceExporter: () => import_trace.AzureMonitorTraceExporter,
  RateLimitedSampler: () => import_rateLimitedSampler.RateLimitedSampler,
  ServiceApiVersion: () => import_Constants.ServiceApiVersion
});
module.exports = __toCommonJS(src_exports);
var import_percentageSampler = require("./sampling/percentageSampler.js");
var import_rateLimitedSampler = require("./sampling/rateLimitedSampler.js");
var import_base = require("./export/base.js");
var import_trace = require("./export/trace.js");
var import_metric = require("./export/metric.js");
var import_log = require("./export/log.js");
var import_Constants = require("./Declarations/Constants.js");
var import_Constants2 = require("./Declarations/Constants.js");
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  AI_OPERATION_NAME,
  ApplicationInsightsSampler,
  AzureMonitorBaseExporter,
  AzureMonitorLogExporter,
  AzureMonitorMetricExporter,
  AzureMonitorTraceExporter,
  RateLimitedSampler,
  ServiceApiVersion
});
