// @azure/monitor-opentelemetry-exporter/dist/commonjs/types.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var types_exports = {};
__export(types_exports, {
  BreezePerformanceCounterNames: () => BreezePerformanceCounterNames,
  MaxPropertyLengths: () => MaxPropertyLengths,
  OTelPerformanceCounterNames: () => OTelPerformanceCounterNames,
  allowedMicrosoftAttributes: () => allowedMicrosoftAttributes,
  experimentalOpenTelemetryValues: () => experimentalOpenTelemetryValues,
  httpSemanticValues: () => httpSemanticValues,
  internalMicrosoftAttributes: () => internalMicrosoftAttributes,
  legacySemanticValues: () => legacySemanticValues
});
module.exports = __toCommonJS(types_exports);
var import_semantic_conventions = require("@opentelemetry/semantic-conventions");
var OTelPerformanceCounterNames = /* @__PURE__ */ ((OTelPerformanceCounterNames2) => {
  OTelPerformanceCounterNames2["PRIVATE_BYTES"] = "Private_Bytes";
  OTelPerformanceCounterNames2["AVAILABLE_BYTES"] = "Available_Bytes";
  OTelPerformanceCounterNames2["PROCESSOR_TIME"] = "Processor_Time";
  OTelPerformanceCounterNames2["PROCESS_TIME_STANDARD"] = "Process_Time_Standard";
  OTelPerformanceCounterNames2["REQUEST_RATE"] = "Request_Rate";
  OTelPerformanceCounterNames2["REQUEST_DURATION"] = "Request_Execution_Time";
  OTelPerformanceCounterNames2["PROCESS_TIME_NORMALIZED"] = "Process_Time_Normalized";
  OTelPerformanceCounterNames2["EXCEPTION_RATE"] = "Exception_Rate";
  return OTelPerformanceCounterNames2;
})(OTelPerformanceCounterNames || {});
var BreezePerformanceCounterNames = /* @__PURE__ */ ((BreezePerformanceCounterNames2) => {
  BreezePerformanceCounterNames2["PRIVATE_BYTES"] = "\\Process(??APP_WIN32_PROC??)\\Private Bytes";
  BreezePerformanceCounterNames2["AVAILABLE_BYTES"] = "\\Memory\\Available Bytes";
  BreezePerformanceCounterNames2["PROCESSOR_TIME"] = "\\Processor(_Total)\\% Processor Time";
  BreezePerformanceCounterNames2["PROCESS_TIME_STANDARD"] = "\\Process(??APP_WIN32_PROC??)\\% Processor Time";
  BreezePerformanceCounterNames2["REQUEST_RATE"] = "\\ASP.NET Applications(??APP_W3SVC_PROC??)\\Requests/Sec";
  BreezePerformanceCounterNames2["REQUEST_DURATION"] = "\\ASP.NET Applications(??APP_W3SVC_PROC??)\\Request Execution Time";
  BreezePerformanceCounterNames2["PROCESS_TIME_NORMALIZED"] = "\\Process(??APP_WIN32_PROC??)\\% Processor Time Normalized";
  BreezePerformanceCounterNames2["EXCEPTION_RATE"] = "\\.NET CLR Exceptions(??APP_CLR_PROC??)\\# of Exceps Thrown / sec";
  return BreezePerformanceCounterNames2;
})(BreezePerformanceCounterNames || {});
var MaxPropertyLengths = /* @__PURE__ */ ((MaxPropertyLengths2) => {
  MaxPropertyLengths2[MaxPropertyLengths2["NINE_BIT"] = 512] = "NINE_BIT";
  MaxPropertyLengths2[MaxPropertyLengths2["TEN_BIT"] = 1024] = "TEN_BIT";
  MaxPropertyLengths2[MaxPropertyLengths2["THIRTEEN_BIT"] = 8192] = "THIRTEEN_BIT";
  MaxPropertyLengths2[MaxPropertyLengths2["FIFTEEN_BIT"] = 32768] = "FIFTEEN_BIT";
  MaxPropertyLengths2[MaxPropertyLengths2["SIXTEEN_BIT"] = 65536] = "SIXTEEN_BIT";
  MaxPropertyLengths2[MaxPropertyLengths2["EIGHTEEN_BIT"] = 262144] = "EIGHTEEN_BIT";
  return MaxPropertyLengths2;
})(MaxPropertyLengths || {});
const legacySemanticValues = [
  import_semantic_conventions.SEMATTRS_NET_PEER_IP,
  import_semantic_conventions.SEMATTRS_NET_PEER_NAME,
  import_semantic_conventions.SEMATTRS_NET_HOST_IP,
  import_semantic_conventions.SEMATTRS_PEER_SERVICE,
  import_semantic_conventions.SEMATTRS_HTTP_USER_AGENT,
  import_semantic_conventions.SEMATTRS_HTTP_METHOD,
  import_semantic_conventions.SEMATTRS_HTTP_URL,
  import_semantic_conventions.SEMATTRS_HTTP_STATUS_CODE,
  import_semantic_conventions.SEMATTRS_HTTP_ROUTE,
  import_semantic_conventions.SEMATTRS_HTTP_HOST,
  import_semantic_conventions.SEMATTRS_DB_SYSTEM,
  import_semantic_conventions.SEMATTRS_DB_STATEMENT,
  import_semantic_conventions.SEMATTRS_DB_OPERATION,
  import_semantic_conventions.SEMATTRS_DB_NAME,
  import_semantic_conventions.SEMATTRS_RPC_SYSTEM,
  import_semantic_conventions.SEMATTRS_RPC_GRPC_STATUS_CODE,
  import_semantic_conventions.SEMATTRS_EXCEPTION_TYPE,
  import_semantic_conventions.SEMATTRS_EXCEPTION_MESSAGE,
  import_semantic_conventions.SEMATTRS_EXCEPTION_STACKTRACE,
  import_semantic_conventions.SEMATTRS_HTTP_SCHEME,
  import_semantic_conventions.SEMATTRS_HTTP_TARGET,
  import_semantic_conventions.SEMATTRS_HTTP_FLAVOR,
  import_semantic_conventions.SEMATTRS_NET_TRANSPORT,
  import_semantic_conventions.SEMATTRS_NET_HOST_NAME,
  import_semantic_conventions.SEMATTRS_NET_HOST_PORT,
  import_semantic_conventions.SEMATTRS_NET_PEER_PORT,
  import_semantic_conventions.SEMATTRS_HTTP_CLIENT_IP,
  import_semantic_conventions.SEMATTRS_ENDUSER_ID,
  "http.status_text"
];
var experimentalOpenTelemetryValues = /* @__PURE__ */ ((experimentalOpenTelemetryValues2) => {
  experimentalOpenTelemetryValues2["SYNTHETIC_TYPE"] = "user_agent.synthetic.type";
  experimentalOpenTelemetryValues2["ATTR_ENDUSER_PSEUDO_ID"] = "enduser.pseudo.id";
  experimentalOpenTelemetryValues2["ATTR_ENDUSER_ID"] = "enduser.id";
  return experimentalOpenTelemetryValues2;
})(experimentalOpenTelemetryValues || {});
const httpSemanticValues = [
  import_semantic_conventions.ATTR_CLIENT_ADDRESS,
  import_semantic_conventions.ATTR_CLIENT_PORT,
  import_semantic_conventions.ATTR_SERVER_ADDRESS,
  import_semantic_conventions.ATTR_SERVER_PORT,
  import_semantic_conventions.ATTR_URL_FULL,
  import_semantic_conventions.ATTR_URL_PATH,
  import_semantic_conventions.ATTR_URL_QUERY,
  import_semantic_conventions.ATTR_URL_SCHEME,
  import_semantic_conventions.ATTR_ERROR_TYPE,
  import_semantic_conventions.ATTR_NETWORK_LOCAL_ADDRESS,
  import_semantic_conventions.ATTR_NETWORK_LOCAL_PORT,
  import_semantic_conventions.ATTR_NETWORK_PROTOCOL_NAME,
  import_semantic_conventions.ATTR_NETWORK_PEER_ADDRESS,
  import_semantic_conventions.ATTR_NETWORK_PEER_PORT,
  import_semantic_conventions.ATTR_NETWORK_PROTOCOL_VERSION,
  import_semantic_conventions.ATTR_NETWORK_TRANSPORT,
  import_semantic_conventions.ATTR_USER_AGENT_ORIGINAL,
  import_semantic_conventions.ATTR_HTTP_REQUEST_METHOD,
  import_semantic_conventions.ATTR_HTTP_RESPONSE_STATUS_CODE,
  import_semantic_conventions.ATTR_EXCEPTION_TYPE,
  import_semantic_conventions.ATTR_EXCEPTION_MESSAGE,
  import_semantic_conventions.ATTR_EXCEPTION_STACKTRACE,
  "enduser.id" /* ATTR_ENDUSER_ID */,
  "enduser.pseudo.id" /* ATTR_ENDUSER_PSEUDO_ID */,
  "user_agent.synthetic.type" /* SYNTHETIC_TYPE */
];
const internalMicrosoftAttributes = ["_MS.ProcessedByMetricExtractors"];
const allowedMicrosoftAttributes = [
  "microsoft.gen_ai.main_agent.name",
  "microsoft.gen_ai.main_agent.id",
  "microsoft.gen_ai.main_agent.version",
  "microsoft.gen_ai.main_agent.conversation_id"
];
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  BreezePerformanceCounterNames,
  MaxPropertyLengths,
  OTelPerformanceCounterNames,
  allowedMicrosoftAttributes,
  experimentalOpenTelemetryValues,
  httpSemanticValues,
  internalMicrosoftAttributes,
  legacySemanticValues
});
