// @azure/monitor-opentelemetry-exporter/dist/commonjs/utils/constants/applicationinsights.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var applicationinsights_exports = {};
__export(applicationinsights_exports, {
  ApplicationInsightsAvailabilityBaseType: () => ApplicationInsightsAvailabilityBaseType,
  ApplicationInsightsAvailabilityName: () => ApplicationInsightsAvailabilityName,
  ApplicationInsightsBaseType: () => ApplicationInsightsBaseType,
  ApplicationInsightsCustomEventName: () => ApplicationInsightsCustomEventName,
  ApplicationInsightsEventBaseType: () => ApplicationInsightsEventBaseType,
  ApplicationInsightsEventName: () => ApplicationInsightsEventName,
  ApplicationInsightsExceptionBaseType: () => ApplicationInsightsExceptionBaseType,
  ApplicationInsightsExceptionName: () => ApplicationInsightsExceptionName,
  ApplicationInsightsMessageBaseType: () => ApplicationInsightsMessageBaseType,
  ApplicationInsightsMessageName: () => ApplicationInsightsMessageName,
  ApplicationInsightsPageViewBaseType: () => ApplicationInsightsPageViewBaseType,
  ApplicationInsightsPageViewName: () => ApplicationInsightsPageViewName,
  AzureMonitorSampleRate: () => AzureMonitorSampleRate,
  DEFAULT_BREEZE_DATA_VERSION: () => DEFAULT_BREEZE_DATA_VERSION,
  DependencyTypes: () => DependencyTypes,
  ENQUEUED_TIME: () => ENQUEUED_TIME,
  MS_LINKS: () => MS_LINKS,
  MicrosoftClientIp: () => MicrosoftClientIp,
  TIME_SINCE_ENQUEUED: () => TIME_SINCE_ENQUEUED,
  packageVersion: () => packageVersion
});
module.exports = __toCommonJS(applicationinsights_exports);
const MS_LINKS = "_MS.links";
const ENQUEUED_TIME = "enqueuedTime";
const TIME_SINCE_ENQUEUED = "timeSinceEnqueued";
const packageVersion = "1.0.0-beta.42";
const DEFAULT_BREEZE_DATA_VERSION = 2;
var DependencyTypes = /* @__PURE__ */ ((DependencyTypes2) => {
  DependencyTypes2["InProc"] = "InProc";
  DependencyTypes2["QueueMessage"] = "Queue Message";
  DependencyTypes2["Sql"] = "SQL";
  DependencyTypes2["Http"] = "Http";
  DependencyTypes2["Grpc"] = "GRPC";
  DependencyTypes2["Wcf"] = "WCF Service";
  return DependencyTypes2;
})(DependencyTypes || {});
const AzureMonitorSampleRate = "microsoft.sample_rate";
const ApplicationInsightsBaseType = "_MS.baseType";
const ApplicationInsightsCustomEventName = "microsoft.custom_event.name";
const MicrosoftClientIp = "microsoft.client.ip";
const ApplicationInsightsMessageName = "Microsoft.ApplicationInsights.Message";
const ApplicationInsightsExceptionName = "Microsoft.ApplicationInsights.Exception";
const ApplicationInsightsPageViewName = "Microsoft.ApplicationInsights.PageView";
const ApplicationInsightsAvailabilityName = "Microsoft.ApplicationInsights.Availability";
const ApplicationInsightsEventName = "Microsoft.ApplicationInsights.Event";
const ApplicationInsightsMessageBaseType = "MessageData";
const ApplicationInsightsExceptionBaseType = "ExceptionData";
const ApplicationInsightsPageViewBaseType = "PageViewData";
const ApplicationInsightsAvailabilityBaseType = "AvailabilityData";
const ApplicationInsightsEventBaseType = "EventData";
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  ApplicationInsightsAvailabilityBaseType,
  ApplicationInsightsAvailabilityName,
  ApplicationInsightsBaseType,
  ApplicationInsightsCustomEventName,
  ApplicationInsightsEventBaseType,
  ApplicationInsightsEventName,
  ApplicationInsightsExceptionBaseType,
  ApplicationInsightsExceptionName,
  ApplicationInsightsMessageBaseType,
  ApplicationInsightsMessageName,
  ApplicationInsightsPageViewBaseType,
  ApplicationInsightsPageViewName,
  AzureMonitorSampleRate,
  DEFAULT_BREEZE_DATA_VERSION,
  DependencyTypes,
  ENQUEUED_TIME,
  MS_LINKS,
  MicrosoftClientIp,
  TIME_SINCE_ENQUEUED,
  packageVersion
});
