// @azure/monitor-opentelemetry-exporter/dist/esm/utils/metricUtils.js
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.
import { DataPointType } from "@opentelemetry/sdk-metrics";
import { createTagsFromResource, truncateCustomDimensions } from "./common.js";
import { BreezePerformanceCounterNames, OTelPerformanceCounterNames } from "../types.js";
import { DEFAULT_BREEZE_DATA_VERSION } from "./constants/applicationinsights.js";
import { ENV_OTEL_METRICS_EXPORTER, ENV_OTLP_METRICS_ENDPOINT, ENV_AZURE_MONITOR_AUTO_ATTACH, ENV_APPLICATIONINSIGHTS_METRICS_TO_LOGANALYTICS_ENABLED, isEnvVarTrue, } from "../Declarations/Constants.js";
import { AttachTypeName, AZURE_MONITOR_AUTO_ATTACH } from "../export/statsbeat/types.js";
import { getInstance } from "../platform/index.js";
const breezePerformanceCountersMap = new Map([
    [OTelPerformanceCounterNames.PRIVATE_BYTES, BreezePerformanceCounterNames.PRIVATE_BYTES],
    [OTelPerformanceCounterNames.AVAILABLE_BYTES, BreezePerformanceCounterNames.AVAILABLE_BYTES],
    [OTelPerformanceCounterNames.PROCESSOR_TIME, BreezePerformanceCounterNames.PROCESSOR_TIME],
    [
        OTelPerformanceCounterNames.PROCESS_TIME_STANDARD,
        BreezePerformanceCounterNames.PROCESS_TIME_STANDARD,
    ],
    [
        OTelPerformanceCounterNames.PROCESS_TIME_NORMALIZED,
        BreezePerformanceCounterNames.PROCESS_TIME_NORMALIZED,
    ],
    [OTelPerformanceCounterNames.REQUEST_RATE, BreezePerformanceCounterNames.REQUEST_RATE],
    [OTelPerformanceCounterNames.REQUEST_DURATION, BreezePerformanceCounterNames.REQUEST_DURATION],
    [OTelPerformanceCounterNames.EXCEPTION_RATE, BreezePerformanceCounterNames.EXCEPTION_RATE],
]);
function createPropertiesFromMetricAttributes(attributes) {
    const properties = {};
    if (attributes) {
        for (const key of Object.keys(attributes)) {
            properties[key] = attributes[key];
        }
    }
    return properties;
}
/**
 * Metric to Azure envelope parsing.
 * @internal
 */
export function resourceMetricsToEnvelope(metrics, ikey, isStatsbeat) {
    const envelopes = [];
    const time = new Date();
    const instrumentationKey = ikey;
    let tags;
    let envelopeName;
    if (isStatsbeat) {
        envelopeName = "Microsoft.ApplicationInsights.Statsbeat";
        const context = getInstance();
        tags = { ...context.tags };
    }
    else {
        envelopeName = "Microsoft.ApplicationInsights.Metric";
        tags = createTagsFromResource(metrics.resource);
    }
    metrics.scopeMetrics.forEach((scopeMetric) => {
        scopeMetric.metrics.forEach((metric) => {
            metric.dataPoints.forEach((dataPoint) => {
                const baseData = {
                    kind: "MetricsData",
                    metrics: [],
                    version: DEFAULT_BREEZE_DATA_VERSION,
                    properties: {},
                };
                baseData.properties = createPropertiesFromMetricAttributes(dataPoint.attributes);
                // If we're not exporting statsbeat, the metric is *not* a standard metric and the env var is set to false, we should not send the metric.
                if (shouldSendToOtlp() &&
                    isAksAttach() &&
                    !isStandardMetric(dataPoint) &&
                    !isEnvVarTrue(ENV_APPLICATIONINSIGHTS_METRICS_TO_LOGANALYTICS_ENABLED) &&
                    !isStatsbeat) {
                    return;
                }
                if (shouldSendToOtlp() && isAksAttach() && !isStatsbeat) {
                    baseData.properties["_MS.SentToAMW"] = "True";
                }
                else if (isAksAttach() && !isStatsbeat) {
                    baseData.properties["_MS.SentToAMW"] = "False";
                }
                let perfCounterName;
                if (breezePerformanceCountersMap.has(metric.descriptor.name)) {
                    perfCounterName = breezePerformanceCountersMap.get(metric.descriptor.name);
                }
                const metricDataPoint = {
                    name: perfCounterName ? perfCounterName : metric.descriptor.name,
                    value: 0,
                    dataPointType: "Aggregation",
                };
                if (metric.dataPointType === DataPointType.SUM ||
                    metric.dataPointType === DataPointType.GAUGE) {
                    metricDataPoint.value = dataPoint.value;
                    metricDataPoint.count = 1;
                }
                else {
                    metricDataPoint.value = dataPoint.value.sum || 0;
                    metricDataPoint.count = dataPoint.value.count;
                    metricDataPoint.max = dataPoint.value.max;
                    metricDataPoint.min = dataPoint.value.min;
                }
                baseData.metrics.push(metricDataPoint);
                const envelope = {
                    name: envelopeName,
                    time: time,
                    sampleRate: 100, // Metrics are never sampled
                    instrumentationKey: instrumentationKey,
                    tags: tags,
                    version: 1,
                    data: {
                        baseType: "MetricData",
                        baseData: {
                            ...baseData,
                            properties: truncateCustomDimensions(baseData.properties || {}),
                        },
                    },
                };
                envelopes.push(envelope);
            });
        });
    });
    return envelopes;
}
export function isAksAttach() {
    return !!(isEnvVarTrue(ENV_AZURE_MONITOR_AUTO_ATTACH) && process.env.AKS_ARM_NAMESPACE_ID);
}
export function shouldSendToOtlp() {
    return !!(process.env[ENV_OTLP_METRICS_ENDPOINT] &&
        process.env[ENV_OTEL_METRICS_EXPORTER]?.includes("otlp"));
}
export function isStandardMetric(dataPoint) {
    return dataPoint.attributes?.["_MS.IsAutocollected"] === "True";
}
export function getAttachType() {
    if (isEnvVarTrue(AZURE_MONITOR_AUTO_ATTACH)) {
        return AttachTypeName.INTEGRATED_AUTO;
    }
    return AttachTypeName.MANUAL;
}
