// @azure/monitor-opentelemetry-exporter/dist/commonjs/Declarations/Constants.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var Constants_exports = {};
__export(Constants_exports, {
  AI_OPERATION_NAME: () => AI_OPERATION_NAME,
  APPLICATION_ID_RESOURCE_KEY: () => APPLICATION_ID_RESOURCE_KEY,
  CUSTOM_DIMENSIONS_GENAI_KEYS: () => CUSTOM_DIMENSIONS_GENAI_KEYS,
  DEFAULT_BREEZE_API_VERSION: () => DEFAULT_BREEZE_API_VERSION,
  DEFAULT_BREEZE_ENDPOINT: () => DEFAULT_BREEZE_ENDPOINT,
  DEFAULT_LIVEMETRICS_ENDPOINT: () => DEFAULT_LIVEMETRICS_ENDPOINT,
  DEFAULT_LIVEMETRICS_HOST: () => DEFAULT_LIVEMETRICS_HOST,
  ENV_APPLICATIONINSIGHTS_METRICS_TO_LOGANALYTICS_ENABLED: () => ENV_APPLICATIONINSIGHTS_METRICS_TO_LOGANALYTICS_ENABLED,
  ENV_APPLICATIONINSIGHTS_SDKSTATS_EXPORT_INTERVAL: () => ENV_APPLICATIONINSIGHTS_SDKSTATS_EXPORT_INTERVAL,
  ENV_APPLICATIONINSIGHTS_SDK_STATS_LOGGING: () => ENV_APPLICATIONINSIGHTS_SDK_STATS_LOGGING,
  ENV_APPLICATIONINSIGHTS_SHIM_VERSION: () => ENV_APPLICATIONINSIGHTS_SHIM_VERSION,
  ENV_AZURE_MONITOR_AUTO_ATTACH: () => ENV_AZURE_MONITOR_AUTO_ATTACH,
  ENV_AZURE_MONITOR_DISTRO_VERSION: () => ENV_AZURE_MONITOR_DISTRO_VERSION,
  ENV_AZURE_MONITOR_PREFIX: () => ENV_AZURE_MONITOR_PREFIX,
  ENV_CONNECTION_STRING: () => ENV_CONNECTION_STRING,
  ENV_DISABLE_SDKSTATS: () => ENV_DISABLE_SDKSTATS,
  ENV_DISABLE_STATSBEAT: () => ENV_DISABLE_STATSBEAT,
  ENV_INSTRUMENTATION_KEY: () => ENV_INSTRUMENTATION_KEY,
  ENV_MICROSOFT_OPENTELEMETRY_VERSION: () => ENV_MICROSOFT_OPENTELEMETRY_VERSION,
  ENV_OPENTELEMETRY_RESOURCE_METRIC_DISABLED: () => ENV_OPENTELEMETRY_RESOURCE_METRIC_DISABLED,
  ENV_OTEL_METRICS_EXPORTER: () => ENV_OTEL_METRICS_EXPORTER,
  ENV_OTLP_METRICS_ENDPOINT: () => ENV_OTLP_METRICS_ENDPOINT,
  LEGACY_ENV_DISABLE_STATSBEAT: () => LEGACY_ENV_DISABLE_STATSBEAT,
  PerformanceCounter: () => PerformanceCounter,
  PerformanceToQuickPulseCounter: () => PerformanceToQuickPulseCounter,
  QuickPulseCounter: () => QuickPulseCounter,
  RetriableRestErrorTypes: () => RetriableRestErrorTypes,
  ServiceApiVersion: () => ServiceApiVersion,
  isEnvVarTrue: () => isEnvVarTrue
});
module.exports = __toCommonJS(Constants_exports);
var ServiceApiVersion = /* @__PURE__ */ ((ServiceApiVersion2) => {
  ServiceApiVersion2["V2"] = "2020-09-15_Preview";
  return ServiceApiVersion2;
})(ServiceApiVersion || {});
const AI_OPERATION_NAME = "ai.operation.name";
const DEFAULT_BREEZE_ENDPOINT = "https://dc.services.visualstudio.com";
const DEFAULT_BREEZE_API_VERSION = "2020-09-15_Preview" /* V2 */;
const DEFAULT_LIVEMETRICS_ENDPOINT = "https://rt.services.visualstudio.com";
const DEFAULT_LIVEMETRICS_HOST = "rt.services.visualstudio.com";
const ENV_CONNECTION_STRING = "APPLICATIONINSIGHTS_CONNECTION_STRING";
const ENV_INSTRUMENTATION_KEY = "APPINSIGHTS_INSTRUMENTATIONKEY";
const ENV_DISABLE_STATSBEAT = "APPLICATIONINSIGHTS_STATSBEAT_DISABLED";
const LEGACY_ENV_DISABLE_STATSBEAT = "APPLICATION_INSIGHTS_NO_STATSBEAT";
const ENV_OPENTELEMETRY_RESOURCE_METRIC_DISABLED = "APPLICATIONINSIGHTS_OPENTELEMETRY_RESOURCE_METRIC_DISABLED";
const ENV_OTLP_METRICS_ENDPOINT = "OTEL_EXPORTER_OTLP_METRICS_ENDPOINT";
const ENV_OTEL_METRICS_EXPORTER = "OTEL_METRICS_EXPORTER";
const ENV_AZURE_MONITOR_AUTO_ATTACH = "AZURE_MONITOR_AUTO_ATTACH";
const ENV_APPLICATIONINSIGHTS_METRICS_TO_LOGANALYTICS_ENABLED = "APPLICATIONINSIGHTS_METRICS_TO_LOGANALYTICS_ENABLED";
function isEnvVarTrue(envVarName) {
  return process.env[envVarName]?.trim().toLowerCase() === "true";
}
var RetriableRestErrorTypes = /* @__PURE__ */ ((RetriableRestErrorTypes2) => {
  RetriableRestErrorTypes2["REQUEST_SEND_ERROR"] = "REQUEST_SEND_ERROR";
  RetriableRestErrorTypes2["DNS_LOOKUP_TIMEOUT"] = "EAI_AGAIN";
  return RetriableRestErrorTypes2;
})(RetriableRestErrorTypes || {});
const ENV_APPLICATIONINSIGHTS_SHIM_VERSION = "APPLICATION_INSIGHTS_SHIM_VERSION";
const ENV_AZURE_MONITOR_PREFIX = "AZURE_MONITOR_PREFIX";
const ENV_AZURE_MONITOR_DISTRO_VERSION = "AZURE_MONITOR_DISTRO_VERSION";
const ENV_MICROSOFT_OPENTELEMETRY_VERSION = "MICROSOFT_OPENTELEMETRY_VERSION";
const APPLICATION_ID_RESOURCE_KEY = "microsoft.applicationId";
const ENV_DISABLE_SDKSTATS = "APPLICATIONINSIGHTS_SDKSTATS_DISABLED";
const ENV_APPLICATIONINSIGHTS_SDKSTATS_EXPORT_INTERVAL = "APPLICATIONINSIGHTS_SDKSTATS_EXPORT_INTERVAL";
const ENV_APPLICATIONINSIGHTS_SDK_STATS_LOGGING = "APPLICATIONINSIGHTS_SDK_STATS_LOGGING";
const CUSTOM_DIMENSIONS_GENAI_KEYS = /* @__PURE__ */ new Set([
  "gen_ai.input.messages",
  "gen_ai.output.messages",
  "gen_ai.system_instructions",
  "gen_ai.tool.definitions",
  "gen_ai.tool.call.arguments",
  "gen_ai.tool.call.result",
  "gen_ai.evaluation.explanation"
]);
var QuickPulseCounter = /* @__PURE__ */ ((QuickPulseCounter2) => {
  QuickPulseCounter2["COMMITTED_BYTES"] = "\\Memory\\Committed Bytes";
  QuickPulseCounter2["PROCESSOR_TIME"] = "\\Processor(_Total)\\% Processor Time";
  QuickPulseCounter2["REQUEST_RATE"] = "\\ApplicationInsights\\Requests/Sec";
  QuickPulseCounter2["REQUEST_FAILURE_RATE"] = "\\ApplicationInsights\\Requests Failed/Sec";
  QuickPulseCounter2["REQUEST_DURATION"] = "\\ApplicationInsights\\Request Duration";
  QuickPulseCounter2["DEPENDENCY_RATE"] = "\\ApplicationInsights\\Dependency Calls/Sec";
  QuickPulseCounter2["DEPENDENCY_FAILURE_RATE"] = "\\ApplicationInsights\\Dependency Calls Failed/Sec";
  QuickPulseCounter2["DEPENDENCY_DURATION"] = "\\ApplicationInsights\\Dependency Call Duration";
  QuickPulseCounter2["EXCEPTION_RATE"] = "\\ApplicationInsights\\Exceptions/Sec";
  return QuickPulseCounter2;
})(QuickPulseCounter || {});
var PerformanceCounter = /* @__PURE__ */ ((PerformanceCounter2) => {
  PerformanceCounter2["PRIVATE_BYTES"] = "\\Process(??APP_WIN32_PROC??)\\Private Bytes";
  PerformanceCounter2["AVAILABLE_BYTES"] = "\\Memory\\Available Bytes";
  PerformanceCounter2["PROCESSOR_TIME"] = "\\Processor(_Total)\\% Processor Time";
  PerformanceCounter2["PROCESS_TIME_STANDARD"] = "\\Process(??APP_WIN32_PROC??)\\% Processor Time";
  PerformanceCounter2["PROCESS_TIME_NORMALIZED"] = "\\Process(??APP_WIN32_PROC??)\\% Processor Time Normalized";
  PerformanceCounter2["REQUEST_RATE"] = "\\ASP.NET Applications(??APP_W3SVC_PROC??)\\Requests/Sec";
  PerformanceCounter2["REQUEST_DURATION"] = "\\ASP.NET Applications(??APP_W3SVC_PROC??)\\Request Execution Time";
  PerformanceCounter2["EXCEPTION_RATE"] = "\\.NET CLR Exceptions(??APP_CLR_PROC??)\\# of Exceps Thrown / sec";
  PerformanceCounter2["IO_RATE"] = "\\Process(??APP_WIN32_PROC??)\\IO Data Bytes/sec";
  return PerformanceCounter2;
})(PerformanceCounter || {});
const PerformanceToQuickPulseCounter = {
  ["\\Processor(_Total)\\% Processor Time" /* PROCESSOR_TIME */]: "\\Processor(_Total)\\% Processor Time" /* PROCESSOR_TIME */,
  ["\\ASP.NET Applications(??APP_W3SVC_PROC??)\\Requests/Sec" /* REQUEST_RATE */]: "\\ApplicationInsights\\Requests/Sec" /* REQUEST_RATE */,
  ["\\ASP.NET Applications(??APP_W3SVC_PROC??)\\Request Execution Time" /* REQUEST_DURATION */]: "\\ApplicationInsights\\Request Duration" /* REQUEST_DURATION */,
  // Remap quick pulse only counters
  ["\\Memory\\Committed Bytes" /* COMMITTED_BYTES */]: "\\Memory\\Committed Bytes" /* COMMITTED_BYTES */,
  ["\\ApplicationInsights\\Requests Failed/Sec" /* REQUEST_FAILURE_RATE */]: "\\ApplicationInsights\\Requests Failed/Sec" /* REQUEST_FAILURE_RATE */,
  ["\\ApplicationInsights\\Dependency Calls/Sec" /* DEPENDENCY_RATE */]: "\\ApplicationInsights\\Dependency Calls/Sec" /* DEPENDENCY_RATE */,
  ["\\ApplicationInsights\\Dependency Calls Failed/Sec" /* DEPENDENCY_FAILURE_RATE */]: "\\ApplicationInsights\\Dependency Calls Failed/Sec" /* DEPENDENCY_FAILURE_RATE */,
  ["\\ApplicationInsights\\Dependency Call Duration" /* DEPENDENCY_DURATION */]: "\\ApplicationInsights\\Dependency Call Duration" /* DEPENDENCY_DURATION */,
  ["\\ApplicationInsights\\Exceptions/Sec" /* EXCEPTION_RATE */]: "\\ApplicationInsights\\Exceptions/Sec" /* EXCEPTION_RATE */
};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  AI_OPERATION_NAME,
  APPLICATION_ID_RESOURCE_KEY,
  CUSTOM_DIMENSIONS_GENAI_KEYS,
  DEFAULT_BREEZE_API_VERSION,
  DEFAULT_BREEZE_ENDPOINT,
  DEFAULT_LIVEMETRICS_ENDPOINT,
  DEFAULT_LIVEMETRICS_HOST,
  ENV_APPLICATIONINSIGHTS_METRICS_TO_LOGANALYTICS_ENABLED,
  ENV_APPLICATIONINSIGHTS_SDKSTATS_EXPORT_INTERVAL,
  ENV_APPLICATIONINSIGHTS_SDK_STATS_LOGGING,
  ENV_APPLICATIONINSIGHTS_SHIM_VERSION,
  ENV_AZURE_MONITOR_AUTO_ATTACH,
  ENV_AZURE_MONITOR_DISTRO_VERSION,
  ENV_AZURE_MONITOR_PREFIX,
  ENV_CONNECTION_STRING,
  ENV_DISABLE_SDKSTATS,
  ENV_DISABLE_STATSBEAT,
  ENV_INSTRUMENTATION_KEY,
  ENV_MICROSOFT_OPENTELEMETRY_VERSION,
  ENV_OPENTELEMETRY_RESOURCE_METRIC_DISABLED,
  ENV_OTEL_METRICS_EXPORTER,
  ENV_OTLP_METRICS_ENDPOINT,
  LEGACY_ENV_DISABLE_STATSBEAT,
  PerformanceCounter,
  PerformanceToQuickPulseCounter,
  QuickPulseCounter,
  RetriableRestErrorTypes,
  ServiceApiVersion,
  isEnvVarTrue
});
