// @azure/monitor-opentelemetry/dist/commonjs/metrics/quickpulse/export/sender.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.QuickpulseSender = void 0;
const tslib_1 = require("tslib");
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.
const url_1 = tslib_1.__importDefault(require("url"));
const core_rest_pipeline_1 = require("@azure/core-rest-pipeline");
const api_1 = require("@opentelemetry/api");
const index_js_1 = require("../../../generated/index.js");
const applicationInsightsResource = "https://monitor.azure.com/.default";
/**
 * Quickpulse sender class
 * @internal
 */
class QuickpulseSender {
    quickpulseClient;
    instrumentationKey;
    endpointUrl;
    credential;
    credentialScopes;
    constructor(options) {
        // Build endpoint using provided configuration or default values
        this.endpointUrl = options.endpointUrl;
        const clientOptions = {
            endpoint: this.endpointUrl,
        };
        this.instrumentationKey = options.instrumentationKey;
        this.credential = options.credential;
        // Configure credential scopes
        if (options.credentialScopes) {
            this.credentialScopes = Array.isArray(options.credentialScopes)
                ? options.credentialScopes
                : [options.credentialScopes];
        }
        else {
            this.credentialScopes = [applicationInsightsResource];
        }
        if (options.credential) {
            clientOptions.credentials = { scopes: this.credentialScopes };
        }
        this.quickpulseClient = this.createQuickpulseClient(clientOptions);
    }
    createQuickpulseClient(clientOptions) {
        const client = new index_js_1.QuickpulseClient(this.credential, clientOptions);
        // Handle redirects in HTTP Sender
        client.pipeline.removePolicy({ name: core_rest_pipeline_1.redirectPolicyName });
        return client;
    }
    /**
     * isSubscribed Quickpulse service
     * @internal
     */
    async isSubscribed(optionalParams) {
        try {
            let responseHeaders = {};
            const body = await this.quickpulseClient.isSubscribed(this.instrumentationKey, {
                ...optionalParams,
                onResponse: (rawResponse) => {
                    responseHeaders = rawResponse.headers.toJSON();
                },
            });
            return {
                ...body,
                xMsQpsSubscribed: responseHeaders["x-ms-qps-subscribed"],
                xMsQpsConfigurationEtag: responseHeaders["x-ms-qps-configuration-etag"],
                xMsQpsServicePollingIntervalHint: responseHeaders["x-ms-qps-service-polling-interval-hint"],
                xMsQpsServiceEndpointRedirectV2: responseHeaders["x-ms-qps-service-endpoint-redirect-v2"],
            };
        }
        catch (error) {
            const restError = error;
            api_1.diag.info("Failed to ping Quickpulse service", restError.message);
        }
        return;
    }
    /**
     * publish Quickpulse service
     * @internal
     */
    async publish(optionalParams) {
        try {
            let responseHeaders = {};
            const body = await this.quickpulseClient.publish(this.instrumentationKey, {
                ...optionalParams,
                onResponse: (rawResponse) => {
                    responseHeaders = rawResponse.headers.toJSON();
                },
            });
            return {
                ...body,
                xMsQpsSubscribed: responseHeaders["x-ms-qps-subscribed"],
                xMsQpsConfigurationEtag: responseHeaders["x-ms-qps-configuration-etag"],
            };
        }
        catch (error) {
            const restError = error;
            api_1.diag.warn("Failed to post Quickpulse service", restError.message);
        }
        return;
    }
    handlePermanentRedirect(location) {
        if (location) {
            const locUrl = new url_1.default.URL(location);
            if (locUrl && locUrl.host) {
                this.endpointUrl = "https://" + locUrl.host;
                // Recreate the client so subsequent requests use the new endpoint
                this.quickpulseClient = this.createQuickpulseClient({
                    endpoint: this.endpointUrl,
                    credentials: { scopes: this.credentialScopes },
                });
            }
        }
    }
}
exports.QuickpulseSender = QuickpulseSender;
