// @azure/monitor-opentelemetry-exporter/dist/commonjs/utils/common.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var common_exports = {};
__export(common_exports, {
  createResourceMetricEnvelope: () => createResourceMetricEnvelope,
  createTagsFromResource: () => createTagsFromResource,
  getDependencyTarget: () => getDependencyTarget,
  getUrl: () => getUrl,
  hrTimeToDate: () => hrTimeToDate,
  isSqlDB: () => isSqlDB,
  isSyntheticSource: () => isSyntheticSource,
  serializeAttribute: () => serializeAttribute,
  shouldCreateResourceMetric: () => shouldCreateResourceMetric,
  truncateCustomDimensions: () => truncateCustomDimensions
});
module.exports = __toCommonJS(common_exports);
var import_node_os = __toESM(require("node:os"));
var import_core_util = require("@azure/core-util");
var import_semantic_conventions = require("@opentelemetry/semantic-conventions");
var import_types = require("../types.js");
var import_platform = require("../platform/index.js");
var import_generated = require("../generated/index.js");
var import_api = require("@opentelemetry/api");
var import_core = require("@opentelemetry/core");
var import_Constants = require("../Declarations/Constants.js");
var import_applicationinsights = require("./constants/applicationinsights.js");
var import_spanUtils = require("./spanUtils.js");
function hrTimeToDate(hrTime) {
  return new Date((0, import_core.hrTimeToNanoseconds)(hrTime) / 1e6);
}
function createTagsFromResource(resource) {
  const context = (0, import_platform.getInstance)();
  const tags = { ...context.tags };
  const attributes = resource?.attributes;
  if (attributes) {
    tags[import_generated.KnownContextTagKeys.AiCloudRole] = getCloudRole(resource);
    tags[import_generated.KnownContextTagKeys.AiCloudRoleInstance] = getCloudRoleInstance(resource);
    if (attributes[import_semantic_conventions.SEMRESATTRS_DEVICE_ID]) {
      tags[import_generated.KnownContextTagKeys.AiDeviceId] = String(attributes[import_semantic_conventions.SEMRESATTRS_DEVICE_ID]);
    }
    if (attributes[import_semantic_conventions.SEMRESATTRS_DEVICE_MODEL_NAME]) {
      tags[import_generated.KnownContextTagKeys.AiDeviceModel] = String(attributes[import_semantic_conventions.SEMRESATTRS_DEVICE_MODEL_NAME]);
    }
    if (attributes[import_semantic_conventions.SEMRESATTRS_SERVICE_VERSION]) {
      tags[import_generated.KnownContextTagKeys.AiApplicationVer] = String(attributes[import_semantic_conventions.SEMRESATTRS_SERVICE_VERSION]);
    }
  }
  return tags;
}
function getCloudRole(resource) {
  let cloudRole = "";
  const serviceName = resource.attributes[import_semantic_conventions.SEMRESATTRS_SERVICE_NAME];
  const serviceNamespace = resource.attributes[import_semantic_conventions.SEMRESATTRS_SERVICE_NAMESPACE];
  if (serviceName) {
    if (!String(serviceName).startsWith("unknown_service")) {
      if (serviceNamespace) {
        return `${serviceNamespace}.${serviceName}`;
      } else {
        return String(serviceName);
      }
    } else {
      if (serviceNamespace) {
        cloudRole = `${serviceNamespace}.${serviceName}`;
      } else {
        cloudRole = String(serviceName);
      }
    }
  }
  const kubernetesDeploymentName = resource.attributes[import_semantic_conventions.SEMRESATTRS_K8S_DEPLOYMENT_NAME];
  if (kubernetesDeploymentName) {
    return String(kubernetesDeploymentName);
  }
  const kuberneteReplicasetName = resource.attributes[import_semantic_conventions.SEMRESATTRS_K8S_REPLICASET_NAME];
  if (kuberneteReplicasetName) {
    return String(kuberneteReplicasetName);
  }
  const kubernetesStatefulSetName = resource.attributes[import_semantic_conventions.SEMRESATTRS_K8S_STATEFULSET_NAME];
  if (kubernetesStatefulSetName) {
    return String(kubernetesStatefulSetName);
  }
  const kubernetesJobName = resource.attributes[import_semantic_conventions.SEMRESATTRS_K8S_JOB_NAME];
  if (kubernetesJobName) {
    return String(kubernetesJobName);
  }
  const kubernetesCronjobName = resource.attributes[import_semantic_conventions.SEMRESATTRS_K8S_CRONJOB_NAME];
  if (kubernetesCronjobName) {
    return String(kubernetesCronjobName);
  }
  const kubernetesDaemonsetName = resource.attributes[import_semantic_conventions.SEMRESATTRS_K8S_DAEMONSET_NAME];
  if (kubernetesDaemonsetName) {
    return String(kubernetesDaemonsetName);
  }
  return cloudRole;
}
function getCloudRoleInstance(resource) {
  const kubernetesPodName = resource.attributes[import_semantic_conventions.SEMRESATTRS_K8S_POD_NAME];
  if (kubernetesPodName) {
    return String(kubernetesPodName);
  }
  const serviceInstanceId = resource.attributes[import_semantic_conventions.SEMRESATTRS_SERVICE_INSTANCE_ID];
  if (serviceInstanceId) {
    return String(serviceInstanceId);
  }
  return import_node_os.default && import_node_os.default.hostname();
}
function isSqlDB(dbSystem) {
  return dbSystem === import_semantic_conventions.DBSYSTEMVALUES_DB2 || dbSystem === import_semantic_conventions.DBSYSTEMVALUES_DERBY || dbSystem === import_semantic_conventions.DBSYSTEMVALUES_MARIADB || dbSystem === import_semantic_conventions.DBSYSTEMVALUES_MSSQL || dbSystem === import_semantic_conventions.DBSYSTEMVALUES_ORACLE || dbSystem === import_semantic_conventions.DBSYSTEMVALUES_SQLITE || dbSystem === import_semantic_conventions.DBSYSTEMVALUES_OTHER_SQL || dbSystem === import_semantic_conventions.DBSYSTEMVALUES_HSQLDB || dbSystem === import_semantic_conventions.DBSYSTEMVALUES_H2;
}
function getUrl(attributes) {
  if (!attributes) {
    return "";
  }
  const httpMethod = (0, import_spanUtils.getHttpMethod)(attributes);
  if (httpMethod) {
    const httpUrl = (0, import_spanUtils.getHttpUrl)(attributes);
    if (httpUrl) {
      return String(httpUrl);
    } else {
      const httpScheme = (0, import_spanUtils.getHttpScheme)(attributes);
      const httpTarget = (0, import_spanUtils.getHttpTarget)(attributes);
      if (httpScheme && httpTarget) {
        const httpHost = (0, import_spanUtils.getHttpHost)(attributes);
        if (httpHost) {
          return `${httpScheme}://${httpHost}${httpTarget}`;
        } else {
          const netPeerPort = (0, import_spanUtils.getNetPeerPort)(attributes);
          if (netPeerPort) {
            const netPeerName = (0, import_spanUtils.getNetPeerName)(attributes);
            if (netPeerName) {
              return `${httpScheme}://${netPeerName}:${netPeerPort}${httpTarget}`;
            } else {
              const netPeerIp = (0, import_spanUtils.getPeerIp)(attributes);
              if (netPeerIp) {
                return `${httpScheme}://${netPeerIp}:${netPeerPort}${httpTarget}`;
              }
            }
          }
        }
      }
    }
  }
  return "";
}
function getDependencyTarget(attributes) {
  if (!attributes) {
    return "";
  }
  const peerService = attributes[import_semantic_conventions.SEMATTRS_PEER_SERVICE];
  const httpHost = (0, import_spanUtils.getHttpHost)(attributes);
  const httpUrl = (0, import_spanUtils.getHttpUrl)(attributes);
  const netPeerName = (0, import_spanUtils.getNetPeerName)(attributes);
  const netPeerIp = (0, import_spanUtils.getPeerIp)(attributes);
  if (peerService) {
    return String(peerService);
  } else if (httpHost) {
    return String(httpHost);
  } else if (httpUrl) {
    return String(httpUrl);
  } else if (netPeerName) {
    return String(netPeerName);
  } else if (netPeerIp) {
    return String(netPeerIp);
  }
  return "";
}
function createResourceMetricEnvelope(resource, instrumentationKey, applicationId) {
  const attributes = resource.attributes;
  if (attributes) {
    const tags = createTagsFromResource(resource);
    const resourceAttributes = {};
    if (applicationId && !attributes[import_Constants.APPLICATION_ID_RESOURCE_KEY]) {
      resourceAttributes[import_Constants.APPLICATION_ID_RESOURCE_KEY] = applicationId;
    }
    for (const key of Object.keys(attributes)) {
      if (!(key.startsWith("_MS.") || key === import_semantic_conventions.ATTR_TELEMETRY_SDK_VERSION || key === import_semantic_conventions.ATTR_TELEMETRY_SDK_LANGUAGE || key === import_semantic_conventions.ATTR_TELEMETRY_SDK_NAME)) {
        resourceAttributes[key] = attributes[key];
      }
    }
    if (Object.keys(resourceAttributes).length > 0) {
      const baseData = {
        kind: "MetricsData",
        version: import_applicationinsights.DEFAULT_BREEZE_DATA_VERSION,
        metrics: [{ name: "_OTELRESOURCE_", value: 1 }],
        properties: truncateCustomDimensions(resourceAttributes)
      };
      const envelope = {
        name: "Microsoft.ApplicationInsights.Metric",
        time: /* @__PURE__ */ new Date(),
        sampleRate: 100,
        // Metrics are never sampled
        instrumentationKey,
        version: 1,
        data: {
          baseType: "MetricData",
          baseData
        },
        tags
      };
      return envelope;
    }
  }
  return;
}
function serializeAttribute(value) {
  if (typeof value === "object") {
    if (value instanceof Error) {
      try {
        return JSON.stringify(value, Object.getOwnPropertyNames(value));
      } catch (err) {
        return String(value);
      }
    } else if (value instanceof Uint8Array) {
      return String(value);
    } else {
      try {
        return JSON.stringify(value);
      } catch (err) {
        return String(value);
      }
    }
  }
  return String(value);
}
function shouldCreateResourceMetric() {
  return !(0, import_Constants.isEnvVarTrue)(import_Constants.ENV_OPENTELEMETRY_RESOURCE_METRIC_DISABLED);
}
function isSyntheticSource(attributes) {
  return !!attributes[import_types.experimentalOpenTelemetryValues.SYNTHETIC_TYPE];
}
function truncateCustomDimensions(properties) {
  const defaultMaxSize = import_types.MaxPropertyLengths.THIRTEEN_BIT;
  const genaiMaxSize = import_types.MaxPropertyLengths.EIGHTEEN_BIT;
  const result = {};
  let truncated = false;
  for (const key of Object.keys(properties)) {
    let value = typeof properties[key] === "string" ? properties[key] : serializeAttribute(properties[key]);
    const maxSize = import_Constants.CUSTOM_DIMENSIONS_GENAI_KEYS.has(key) ? genaiMaxSize : defaultMaxSize;
    const valueBytes = (0, import_core_util.stringToUint8Array)(value, "utf-8");
    if (valueBytes.byteLength > maxSize) {
      value = (0, import_core_util.uint8ArrayToString)(valueBytes.subarray(0, maxSize), "utf-8");
      truncated = true;
    }
    result[key] = value;
  }
  if (truncated) {
    import_api.diag.debug("Custom dimension value exceeded size limit. Property value has been truncated.");
  }
  return result;
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  createResourceMetricEnvelope,
  createTagsFromResource,
  getDependencyTarget,
  getUrl,
  hrTimeToDate,
  isSqlDB,
  isSyntheticSource,
  serializeAttribute,
  shouldCreateResourceMetric,
  truncateCustomDimensions
});
