// @azure/monitor-opentelemetry-exporter/dist/commonjs/utils/logUtils.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var logUtils_exports = {};
__export(logUtils_exports, {
  logToEnvelope: () => logToEnvelope
});
module.exports = __toCommonJS(logUtils_exports);
var import_generated = require("../generated/index.js");
var import_common = require("./common.js");
var import_semantic_conventions = require("@opentelemetry/semantic-conventions");
var import_types = require("../types.js");
var import_types2 = require("../types.js");
var import_api = require("@opentelemetry/api");
var import_applicationinsights = require("./constants/applicationinsights.js");
function logToEnvelope(log, ikey) {
  const time = (0, import_common.hrTimeToDate)(log.hrTime);
  const sampleRate = 100;
  const instrumentationKey = ikey;
  const tags = createTagsFromLog(log);
  let [properties, measurements] = createPropertiesFromLog(log);
  let name;
  let baseType;
  let baseData;
  const exceptionStacktrace = log.attributes[import_semantic_conventions.ATTR_EXCEPTION_STACKTRACE];
  const exceptionType = log.attributes[import_semantic_conventions.ATTR_EXCEPTION_TYPE];
  const isExceptionType = !!(exceptionType && exceptionStacktrace) || false;
  const isMessageType = !log.attributes[import_applicationinsights.ApplicationInsightsBaseType] && !log.attributes[import_applicationinsights.ApplicationInsightsCustomEventName] && !exceptionType;
  if (isExceptionType) {
    const exceptionMessage = log.attributes[import_semantic_conventions.ATTR_EXCEPTION_MESSAGE];
    name = import_applicationinsights.ApplicationInsightsExceptionName;
    baseType = import_applicationinsights.ApplicationInsightsExceptionBaseType;
    const exceptionDetails = {
      typeName: String(exceptionType),
      message: String(exceptionMessage),
      hasFullStack: exceptionStacktrace ? true : false,
      stack: String(exceptionStacktrace)
    };
    const exceptionData = {
      kind: "ExceptionData",
      exceptions: [exceptionDetails],
      severityLevel: getSeverity(log.severityNumber),
      version: import_applicationinsights.DEFAULT_BREEZE_DATA_VERSION
    };
    baseData = exceptionData;
  } else if (log.attributes[import_applicationinsights.ApplicationInsightsCustomEventName]) {
    name = import_applicationinsights.ApplicationInsightsEventName;
    baseType = import_applicationinsights.ApplicationInsightsEventBaseType;
    const eventData = {
      kind: "EventData",
      name: String(log.attributes[import_applicationinsights.ApplicationInsightsCustomEventName]),
      version: import_applicationinsights.DEFAULT_BREEZE_DATA_VERSION
    };
    baseData = eventData;
    measurements = getLegacyApplicationInsightsMeasurements(log);
  } else if (isMessageType) {
    name = import_applicationinsights.ApplicationInsightsMessageName;
    baseType = import_applicationinsights.ApplicationInsightsMessageBaseType;
    const messageData = {
      kind: "MessageData",
      message: (0, import_common.serializeAttribute)(log.body),
      severityLevel: getSeverity(log.severityNumber),
      version: import_applicationinsights.DEFAULT_BREEZE_DATA_VERSION
    };
    baseData = messageData;
  } else {
    baseType = String(log.attributes[import_applicationinsights.ApplicationInsightsBaseType]);
    name = getLegacyApplicationInsightsName(log);
    const legacyBaseData = getLegacyApplicationInsightsBaseData(log);
    if (!legacyBaseData) {
      return;
    }
    measurements = getLegacyApplicationInsightsMeasurements(log);
    baseData = legacyBaseData;
  }
  if (baseData && "message" in baseData && baseData.message) {
    baseData.message = String(baseData.message).substring(0, import_types2.MaxPropertyLengths.FIFTEEN_BIT);
  }
  baseData.properties = (0, import_common.truncateCustomDimensions)(properties);
  baseData.measurements = measurements;
  return {
    name,
    sampleRate,
    time,
    instrumentationKey,
    tags,
    version: 1,
    data: {
      baseType,
      baseData
    }
  };
}
function createTagsFromLog(log) {
  const tags = (0, import_common.createTagsFromResource)(log.resource);
  if (log.spanContext?.traceId) {
    tags[import_generated.KnownContextTagKeys.AiOperationId] = log.spanContext.traceId;
  }
  if (log.spanContext?.spanId) {
    tags[import_generated.KnownContextTagKeys.AiOperationParentId] = log.spanContext.spanId;
  }
  if (log.attributes[import_generated.KnownContextTagKeys.AiOperationName]) {
    tags[import_generated.KnownContextTagKeys.AiOperationName] = log.attributes[import_generated.KnownContextTagKeys.AiOperationName];
  }
  if ((0, import_common.isSyntheticSource)(log.attributes)) {
    tags[import_generated.KnownContextTagKeys.AiOperationSyntheticSource] = "True";
  }
  const microsoftClientIp = log.attributes?.[import_applicationinsights.MicrosoftClientIp];
  if (microsoftClientIp) {
    tags[import_generated.KnownContextTagKeys.AiLocationIp] = String(microsoftClientIp);
  }
  const attributes = log.attributes;
  if (attributes[import_types.experimentalOpenTelemetryValues.ATTR_ENDUSER_ID]) {
    const endUserId = String(attributes[import_types.experimentalOpenTelemetryValues.ATTR_ENDUSER_ID]);
    if (endUserId && endUserId.length > 0) {
      tags[import_generated.KnownContextTagKeys.AiUserAuthUserId] = endUserId;
    }
  }
  if (attributes[import_types.experimentalOpenTelemetryValues.ATTR_ENDUSER_PSEUDO_ID]) {
    const endUserPseudoId = String(
      attributes[import_types.experimentalOpenTelemetryValues.ATTR_ENDUSER_PSEUDO_ID]
    );
    if (endUserPseudoId && endUserPseudoId.length > 0) {
      tags[import_generated.KnownContextTagKeys.AiUserId] = endUserPseudoId;
    }
  }
  return tags;
}
function createPropertiesFromLog(log) {
  const measurements = {};
  const properties = {};
  if (log.attributes) {
    for (const key of Object.keys(log.attributes)) {
      if (!(key.startsWith("_MS.") || key.startsWith("microsoft") || import_types2.legacySemanticValues.includes(key) || import_types2.httpSemanticValues.includes(key) || key === import_generated.KnownContextTagKeys.AiOperationName)) {
        properties[key] = (0, import_common.serializeAttribute)(log.attributes[key]);
      }
    }
  }
  return [properties, measurements];
}
function getSeverity(severityNumber) {
  if (severityNumber) {
    if (severityNumber > 0 && severityNumber < 9) {
      return import_generated.KnownSeverityLevel.Verbose;
    } else if (severityNumber >= 9 && severityNumber < 13) {
      return import_generated.KnownSeverityLevel.Information;
    } else if (severityNumber >= 13 && severityNumber < 17) {
      return import_generated.KnownSeverityLevel.Warning;
    } else if (severityNumber >= 17 && severityNumber < 21) {
      return import_generated.KnownSeverityLevel.Error;
    } else if (severityNumber >= 21 && severityNumber < 25) {
      return import_generated.KnownSeverityLevel.Critical;
    }
  }
  return;
}
function getLegacyApplicationInsightsName(log) {
  let name = "";
  switch (log.attributes[import_applicationinsights.ApplicationInsightsBaseType]) {
    case import_applicationinsights.ApplicationInsightsAvailabilityBaseType:
      name = import_applicationinsights.ApplicationInsightsAvailabilityName;
      break;
    case import_applicationinsights.ApplicationInsightsExceptionBaseType:
      name = import_applicationinsights.ApplicationInsightsExceptionName;
      break;
    case import_applicationinsights.ApplicationInsightsMessageBaseType:
      name = import_applicationinsights.ApplicationInsightsMessageName;
      break;
    case import_applicationinsights.ApplicationInsightsPageViewBaseType:
      name = import_applicationinsights.ApplicationInsightsPageViewName;
      break;
    case import_applicationinsights.ApplicationInsightsEventBaseType:
      name = import_applicationinsights.ApplicationInsightsEventName;
      break;
  }
  return name;
}
function getLegacyApplicationInsightsMeasurements(log) {
  const measurements = {};
  const body = log.body;
  if (body && typeof body === "object" && !Array.isArray(body) && "measurements" in body) {
    const bodyMeasurements = body.measurements;
    if (bodyMeasurements && typeof bodyMeasurements === "object" && !Array.isArray(bodyMeasurements)) {
      for (const [key, value] of Object.entries(bodyMeasurements)) {
        if (typeof value === "number" && Number.isFinite(value)) {
          measurements[key] = value;
        }
      }
    }
  }
  return measurements;
}
function getLegacyApplicationInsightsBaseData(log) {
  let baseData;
  if (log.body) {
    try {
      switch (log.attributes[import_applicationinsights.ApplicationInsightsBaseType]) {
        case import_applicationinsights.ApplicationInsightsAvailabilityBaseType:
          baseData = {
            ...log.body,
            kind: "AvailabilityData"
          };
          break;
        case import_applicationinsights.ApplicationInsightsExceptionBaseType:
          baseData = {
            ...log.body,
            kind: "ExceptionData"
          };
          break;
        case import_applicationinsights.ApplicationInsightsMessageBaseType:
          baseData = {
            ...log.body,
            kind: "MessageData"
          };
          if (typeof baseData.message === "object") {
            baseData.message = (0, import_common.serializeAttribute)(baseData.message);
          }
          break;
        case import_applicationinsights.ApplicationInsightsPageViewBaseType:
          baseData = {
            ...log.body,
            kind: "PageViewData"
          };
          break;
        case import_applicationinsights.ApplicationInsightsEventBaseType:
          baseData = {
            ...log.body,
            kind: "EventData"
          };
          break;
      }
      if (baseData && baseData.version === void 0) {
        baseData.version = import_applicationinsights.DEFAULT_BREEZE_DATA_VERSION;
      }
    } catch (err) {
      import_api.diag.error("AzureMonitorLogExporter failed to parse Application Insights Telemetry");
    }
  }
  return baseData;
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  logToEnvelope
});
