// @azure/monitor-opentelemetry-exporter/dist/commonjs/utils/metricUtils.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var metricUtils_exports = {};
__export(metricUtils_exports, {
  getAttachType: () => getAttachType,
  isAksAttach: () => isAksAttach,
  isStandardMetric: () => isStandardMetric,
  resourceMetricsToEnvelope: () => resourceMetricsToEnvelope,
  shouldSendToOtlp: () => shouldSendToOtlp
});
module.exports = __toCommonJS(metricUtils_exports);
var import_sdk_metrics = require("@opentelemetry/sdk-metrics");
var import_common = require("./common.js");
var import_types = require("../types.js");
var import_applicationinsights = require("./constants/applicationinsights.js");
var import_Constants = require("../Declarations/Constants.js");
var import_types2 = require("../export/statsbeat/types.js");
var import_platform = require("../platform/index.js");
const breezePerformanceCountersMap = /* @__PURE__ */ new Map([
  [import_types.OTelPerformanceCounterNames.PRIVATE_BYTES, import_types.BreezePerformanceCounterNames.PRIVATE_BYTES],
  [import_types.OTelPerformanceCounterNames.AVAILABLE_BYTES, import_types.BreezePerformanceCounterNames.AVAILABLE_BYTES],
  [import_types.OTelPerformanceCounterNames.PROCESSOR_TIME, import_types.BreezePerformanceCounterNames.PROCESSOR_TIME],
  [
    import_types.OTelPerformanceCounterNames.PROCESS_TIME_STANDARD,
    import_types.BreezePerformanceCounterNames.PROCESS_TIME_STANDARD
  ],
  [
    import_types.OTelPerformanceCounterNames.PROCESS_TIME_NORMALIZED,
    import_types.BreezePerformanceCounterNames.PROCESS_TIME_NORMALIZED
  ],
  [import_types.OTelPerformanceCounterNames.REQUEST_RATE, import_types.BreezePerformanceCounterNames.REQUEST_RATE],
  [import_types.OTelPerformanceCounterNames.REQUEST_DURATION, import_types.BreezePerformanceCounterNames.REQUEST_DURATION],
  [import_types.OTelPerformanceCounterNames.EXCEPTION_RATE, import_types.BreezePerformanceCounterNames.EXCEPTION_RATE]
]);
function createPropertiesFromMetricAttributes(attributes) {
  const properties = {};
  if (attributes) {
    for (const key of Object.keys(attributes)) {
      properties[key] = attributes[key];
    }
  }
  return properties;
}
function resourceMetricsToEnvelope(metrics, ikey, isStatsbeat) {
  const envelopes = [];
  const time = /* @__PURE__ */ new Date();
  const instrumentationKey = ikey;
  let tags;
  let envelopeName;
  if (isStatsbeat) {
    envelopeName = "Microsoft.ApplicationInsights.Statsbeat";
    const context = (0, import_platform.getInstance)();
    tags = { ...context.tags };
  } else {
    envelopeName = "Microsoft.ApplicationInsights.Metric";
    tags = (0, import_common.createTagsFromResource)(metrics.resource);
  }
  metrics.scopeMetrics.forEach((scopeMetric) => {
    scopeMetric.metrics.forEach((metric) => {
      metric.dataPoints.forEach((dataPoint) => {
        const baseData = {
          kind: "MetricsData",
          metrics: [],
          version: import_applicationinsights.DEFAULT_BREEZE_DATA_VERSION,
          properties: {}
        };
        baseData.properties = createPropertiesFromMetricAttributes(dataPoint.attributes);
        if (shouldSendToOtlp() && isAksAttach() && !isStandardMetric(dataPoint) && !(0, import_Constants.isEnvVarTrue)(import_Constants.ENV_APPLICATIONINSIGHTS_METRICS_TO_LOGANALYTICS_ENABLED) && !isStatsbeat) {
          return;
        }
        if (shouldSendToOtlp() && isAksAttach() && !isStatsbeat) {
          baseData.properties["_MS.SentToAMW"] = "True";
        } else if (isAksAttach() && !isStatsbeat) {
          baseData.properties["_MS.SentToAMW"] = "False";
        }
        let perfCounterName;
        if (breezePerformanceCountersMap.has(metric.descriptor.name)) {
          perfCounterName = breezePerformanceCountersMap.get(metric.descriptor.name);
        }
        const metricDataPoint = {
          name: perfCounterName ? perfCounterName : metric.descriptor.name,
          value: 0,
          dataPointType: "Aggregation"
        };
        if (metric.dataPointType === import_sdk_metrics.DataPointType.SUM || metric.dataPointType === import_sdk_metrics.DataPointType.GAUGE) {
          metricDataPoint.value = dataPoint.value;
          metricDataPoint.count = 1;
        } else {
          metricDataPoint.value = dataPoint.value.sum || 0;
          metricDataPoint.count = dataPoint.value.count;
          metricDataPoint.max = dataPoint.value.max;
          metricDataPoint.min = dataPoint.value.min;
        }
        baseData.metrics.push(metricDataPoint);
        const envelope = {
          name: envelopeName,
          time,
          sampleRate: 100,
          // Metrics are never sampled
          instrumentationKey,
          tags,
          version: 1,
          data: {
            baseType: "MetricData",
            baseData: {
              ...baseData,
              properties: (0, import_common.truncateCustomDimensions)(baseData.properties || {})
            }
          }
        };
        envelopes.push(envelope);
      });
    });
  });
  return envelopes;
}
function isAksAttach() {
  return !!((0, import_Constants.isEnvVarTrue)(import_Constants.ENV_AZURE_MONITOR_AUTO_ATTACH) && process.env.AKS_ARM_NAMESPACE_ID);
}
function shouldSendToOtlp() {
  return !!(process.env[import_Constants.ENV_OTLP_METRICS_ENDPOINT] && process.env[import_Constants.ENV_OTEL_METRICS_EXPORTER]?.includes("otlp"));
}
function isStandardMetric(dataPoint) {
  return dataPoint.attributes?.["_MS.IsAutocollected"] === "True";
}
function getAttachType() {
  if ((0, import_Constants.isEnvVarTrue)(import_types2.AZURE_MONITOR_AUTO_ATTACH)) {
    return import_types2.AttachTypeName.INTEGRATED_AUTO;
  }
  return import_types2.AttachTypeName.MANUAL;
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  getAttachType,
  isAksAttach,
  isStandardMetric,
  resourceMetricsToEnvelope,
  shouldSendToOtlp
});
