// @azure/monitor-opentelemetry-exporter/dist/commonjs/utils/spanUtils.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var spanUtils_exports = {};
__export(spanUtils_exports, {
  getHttpClientIp: () => getHttpClientIp,
  getHttpHost: () => getHttpHost,
  getHttpMethod: () => getHttpMethod,
  getHttpScheme: () => getHttpScheme,
  getHttpStatusCode: () => getHttpStatusCode,
  getHttpTarget: () => getHttpTarget,
  getHttpUrl: () => getHttpUrl,
  getLocationIp: () => getLocationIp,
  getNetPeerName: () => getNetPeerName,
  getNetPeerPort: () => getNetPeerPort,
  getPeerIp: () => getPeerIp,
  getUserAgent: () => getUserAgent,
  readableSpanToEnvelope: () => readableSpanToEnvelope,
  spanEventsToEnvelopes: () => spanEventsToEnvelopes
});
module.exports = __toCommonJS(spanUtils_exports);
var import_core = require("@opentelemetry/core");
var import_api = require("@opentelemetry/api");
var import_semantic_conventions = require("@opentelemetry/semantic-conventions");
var import_common = require("./common.js");
var import_types = require("../types.js");
var import_eventhub = require("./eventhub.js");
var import_applicationinsights = require("./constants/applicationinsights.js");
var import_azAttributes = require("./constants/span/azAttributes.js");
var import_generated = require("../generated/index.js");
var import_breezeUtils = require("./breezeUtils.js");
function createTagsFromSpan(span) {
  const tags = (0, import_common.createTagsFromResource)(span.resource);
  tags[import_generated.KnownContextTagKeys.AiOperationId] = span.spanContext().traceId;
  if (span.parentSpanContext?.spanId) {
    tags[import_generated.KnownContextTagKeys.AiOperationParentId] = span.parentSpanContext.spanId;
  }
  const endUserId = span.attributes[import_types.experimentalOpenTelemetryValues.ATTR_ENDUSER_ID];
  if (endUserId) {
    tags[import_generated.KnownContextTagKeys.AiUserAuthUserId] = String(endUserId);
  }
  const endUserPseudoId = span.attributes[import_types.experimentalOpenTelemetryValues.ATTR_ENDUSER_PSEUDO_ID];
  if (endUserPseudoId) {
    tags[import_generated.KnownContextTagKeys.AiUserId] = String(endUserPseudoId);
  }
  const httpUserAgent = getUserAgent(span.attributes);
  if (httpUserAgent) {
    tags["ai.user.userAgent"] = String(httpUserAgent);
  }
  if ((0, import_common.isSyntheticSource)(span.attributes)) {
    tags[import_generated.KnownContextTagKeys.AiOperationSyntheticSource] = "True";
  }
  const microsoftClientIp = span.attributes[import_applicationinsights.MicrosoftClientIp];
  if (microsoftClientIp) {
    tags[import_generated.KnownContextTagKeys.AiLocationIp] = String(microsoftClientIp);
  }
  if (span.kind === import_api.SpanKind.SERVER) {
    const httpMethod = getHttpMethod(span.attributes);
    if (!microsoftClientIp) {
      getLocationIp(tags, span.attributes);
    }
    if (httpMethod) {
      const httpRoute = span.attributes[import_semantic_conventions.ATTR_HTTP_ROUTE];
      const httpUrl = getHttpUrl(span.attributes);
      tags[import_generated.KnownContextTagKeys.AiOperationName] = span.name;
      if (httpRoute) {
        tags[import_generated.KnownContextTagKeys.AiOperationName] = String(
          `${httpMethod} ${httpRoute}`
        ).substring(0, import_types.MaxPropertyLengths.TEN_BIT);
      } else if (httpUrl) {
        try {
          const url = new URL(String(httpUrl));
          tags[import_generated.KnownContextTagKeys.AiOperationName] = String(
            `${httpMethod} ${url.pathname}`
          ).substring(0, import_types.MaxPropertyLengths.TEN_BIT);
        } catch {
        }
      }
    } else {
      tags[import_generated.KnownContextTagKeys.AiOperationName] = span.name;
    }
  } else {
    if (span.attributes[import_generated.KnownContextTagKeys.AiOperationName]) {
      tags[import_generated.KnownContextTagKeys.AiOperationName] = span.attributes[import_generated.KnownContextTagKeys.AiOperationName];
    }
  }
  return tags;
}
function createPropertiesFromSpanAttributes(attributes) {
  const properties = {};
  if (attributes) {
    for (const key of Object.keys(attributes)) {
      if (
        // We need to not ignore the _MS.ProcessedByMetricExtractors key as it's used to identify standard metrics
        !(key.startsWith("_MS.") && !import_types.internalMicrosoftAttributes.includes(key) || key.startsWith("microsoft.") && !import_types.allowedMicrosoftAttributes.includes(key) || import_types.legacySemanticValues.includes(key) || import_types.httpSemanticValues.includes(key) || key === import_generated.KnownContextTagKeys.AiOperationName)
      ) {
        properties[key] = (0, import_common.serializeAttribute)(attributes[key]);
      }
    }
  }
  return properties;
}
function createPropertiesFromSpan(span) {
  const properties = createPropertiesFromSpanAttributes(span.attributes);
  const measurements = {};
  const links = span.links.map((link) => ({
    operation_Id: link.context.traceId,
    id: link.context.spanId
  }));
  if (links.length > 0) {
    properties[import_applicationinsights.MS_LINKS] = JSON.stringify(links);
  }
  return [properties, measurements];
}
function createDependencyData(span) {
  const remoteDependencyData = {
    kind: "RemoteDependencyData",
    name: span.name,
    // Default
    id: `${span.spanContext().spanId}`,
    success: span.status?.code !== import_api.SpanStatusCode.ERROR,
    resultCode: "0",
    type: "Dependency",
    duration: (0, import_breezeUtils.msToTimeSpan)((0, import_core.hrTimeToMilliseconds)(span.duration)),
    version: import_applicationinsights.DEFAULT_BREEZE_DATA_VERSION
  };
  if (span.kind === import_api.SpanKind.PRODUCER) {
    remoteDependencyData.type = import_applicationinsights.DependencyTypes.QueueMessage;
  }
  if (span.kind === import_api.SpanKind.INTERNAL && span.parentSpanContext) {
    remoteDependencyData.type = import_applicationinsights.DependencyTypes.InProc;
  }
  const httpMethod = getHttpMethod(span.attributes);
  const dbSystem = span.attributes[import_semantic_conventions.SEMATTRS_DB_SYSTEM];
  const rpcSystem = span.attributes[import_semantic_conventions.SEMATTRS_RPC_SYSTEM];
  if (httpMethod) {
    const httpUrl = getHttpUrl(span.attributes);
    if (httpUrl) {
      try {
        const dependencyUrl = new URL(String(httpUrl));
        remoteDependencyData.name = `${httpMethod} ${dependencyUrl.pathname}`;
      } catch {
      }
    }
    remoteDependencyData.type = import_applicationinsights.DependencyTypes.Http;
    remoteDependencyData.data = (0, import_common.getUrl)(span.attributes);
    const httpStatusCode = getHttpStatusCode(span.attributes);
    if (httpStatusCode) {
      remoteDependencyData.resultCode = String(httpStatusCode);
    }
    let target = (0, import_common.getDependencyTarget)(span.attributes);
    if (target) {
      try {
        const portRegex = new RegExp(/(https?)(:\/\/.*)(:\d+)(\S*)/);
        const res = portRegex.exec(target);
        if (res !== null) {
          const protocol = res[1];
          const port = res[3];
          if (protocol === "https" && port === ":443" || protocol === "http" && port === ":80") {
            target = res[1] + res[2] + res[4];
          }
        }
      } catch {
      }
      remoteDependencyData.target = `${target}`;
    }
  } else if (dbSystem) {
    if (String(dbSystem) === import_semantic_conventions.DBSYSTEMVALUES_MYSQL) {
      remoteDependencyData.type = "mysql";
    } else if (String(dbSystem) === import_semantic_conventions.DBSYSTEMVALUES_POSTGRESQL) {
      remoteDependencyData.type = "postgresql";
    } else if (String(dbSystem) === import_semantic_conventions.DBSYSTEMVALUES_MONGODB) {
      remoteDependencyData.type = "mongodb";
    } else if (String(dbSystem) === import_semantic_conventions.DBSYSTEMVALUES_REDIS) {
      remoteDependencyData.type = "redis";
    } else if ((0, import_common.isSqlDB)(String(dbSystem))) {
      remoteDependencyData.type = "SQL";
    } else {
      remoteDependencyData.type = String(dbSystem);
    }
    const dbStatement = span.attributes[import_semantic_conventions.SEMATTRS_DB_STATEMENT];
    const dbOperation = span.attributes[import_semantic_conventions.SEMATTRS_DB_OPERATION];
    if (dbStatement) {
      remoteDependencyData.data = String(dbStatement);
    } else if (dbOperation) {
      remoteDependencyData.data = String(dbOperation);
    }
    const target = (0, import_common.getDependencyTarget)(span.attributes);
    const dbName = span.attributes[import_semantic_conventions.SEMATTRS_DB_NAME];
    if (target) {
      remoteDependencyData.target = dbName ? `${target}|${dbName}` : `${target}`;
    } else {
      remoteDependencyData.target = dbName ? `${dbName}` : `${dbSystem}`;
    }
  } else if (rpcSystem) {
    if (rpcSystem === import_applicationinsights.DependencyTypes.Wcf) {
      remoteDependencyData.type = import_applicationinsights.DependencyTypes.Wcf;
    } else {
      remoteDependencyData.type = import_applicationinsights.DependencyTypes.Grpc;
    }
    const grpcStatusCode = span.attributes[import_semantic_conventions.SEMATTRS_RPC_GRPC_STATUS_CODE];
    if (grpcStatusCode) {
      remoteDependencyData.resultCode = String(grpcStatusCode);
    }
    const target = (0, import_common.getDependencyTarget)(span.attributes);
    if (target) {
      remoteDependencyData.target = `${target}`;
    } else if (rpcSystem) {
      remoteDependencyData.target = String(rpcSystem);
    }
  }
  return remoteDependencyData;
}
function createRequestData(span) {
  const requestData = {
    kind: "RequestData",
    id: `${span.spanContext().spanId}`,
    success: span.status.code !== import_api.SpanStatusCode.UNSET ? span.status.code === import_api.SpanStatusCode.OK : (Number(getHttpStatusCode(span.attributes)) || 0) < 400,
    responseCode: "0",
    duration: (0, import_breezeUtils.msToTimeSpan)((0, import_core.hrTimeToMilliseconds)(span.duration)),
    version: import_applicationinsights.DEFAULT_BREEZE_DATA_VERSION,
    source: void 0
  };
  const httpMethod = getHttpMethod(span.attributes);
  const grpcStatusCode = span.attributes[import_semantic_conventions.SEMATTRS_RPC_GRPC_STATUS_CODE];
  if (httpMethod) {
    requestData.url = (0, import_common.getUrl)(span.attributes);
    const httpStatusCode = getHttpStatusCode(span.attributes);
    if (httpStatusCode) {
      requestData.responseCode = String(httpStatusCode);
    }
  } else if (grpcStatusCode) {
    requestData.responseCode = String(grpcStatusCode);
  }
  return requestData;
}
function readableSpanToEnvelope(span, ikey) {
  let name;
  let baseType;
  let baseData;
  const time = (0, import_common.hrTimeToDate)(span.startTime);
  const instrumentationKey = ikey;
  const tags = createTagsFromSpan(span);
  const [properties, measurements] = createPropertiesFromSpan(span);
  switch (span.kind) {
    case import_api.SpanKind.CLIENT:
    case import_api.SpanKind.PRODUCER:
    case import_api.SpanKind.INTERNAL:
      name = "Microsoft.ApplicationInsights.RemoteDependency";
      baseType = "RemoteDependencyData";
      baseData = createDependencyData(span);
      break;
    case import_api.SpanKind.SERVER:
    case import_api.SpanKind.CONSUMER:
      name = "Microsoft.ApplicationInsights.Request";
      baseType = "RequestData";
      baseData = createRequestData(span);
      baseData.name = tags[import_generated.KnownContextTagKeys.AiOperationName];
      break;
    default:
      import_api.diag.error(`Unsupported span kind ${span.kind}`);
      throw new Error(`Unsupported span kind ${span.kind}`);
  }
  let sampleRate = 100;
  if (span.attributes[import_applicationinsights.AzureMonitorSampleRate]) {
    sampleRate = Number(span.attributes[import_applicationinsights.AzureMonitorSampleRate]);
  }
  if (span.attributes[import_azAttributes.AzNamespace]) {
    if (span.kind === import_api.SpanKind.INTERNAL && baseData && "resultCode" in baseData) {
      baseData.type = `${import_applicationinsights.DependencyTypes.InProc} | ${span.attributes[import_azAttributes.AzNamespace]}`;
    }
    if (span.attributes[import_azAttributes.AzNamespace] === import_azAttributes.MicrosoftEventHub) {
      (0, import_eventhub.parseEventHubSpan)(span, baseData);
    }
  }
  if (baseData.id) {
    baseData.id = baseData.id.substring(0, import_types.MaxPropertyLengths.NINE_BIT);
  }
  if (baseData.name) {
    baseData.name = baseData.name.substring(0, import_types.MaxPropertyLengths.TEN_BIT);
  }
  if (baseData && "resultCode" in baseData) {
    const dependencyData = baseData;
    if (dependencyData.resultCode) {
      dependencyData.resultCode = String(dependencyData.resultCode).substring(
        0,
        import_types.MaxPropertyLengths.TEN_BIT
      );
    }
    if (dependencyData.data) {
      dependencyData.data = String(dependencyData.data).substring(
        0,
        import_types.MaxPropertyLengths.THIRTEEN_BIT
      );
    }
    if (dependencyData.type) {
      dependencyData.type = String(dependencyData.type).substring(0, import_types.MaxPropertyLengths.TEN_BIT);
    }
    if (dependencyData.target) {
      dependencyData.target = String(dependencyData.target).substring(
        0,
        import_types.MaxPropertyLengths.TEN_BIT
      );
    }
  }
  baseData.properties = (0, import_common.truncateCustomDimensions)(properties);
  baseData.measurements = measurements;
  return {
    name,
    sampleRate,
    time,
    instrumentationKey,
    tags,
    version: 1,
    data: {
      baseType,
      baseData
    }
  };
}
function spanEventsToEnvelopes(span, ikey) {
  const envelopes = [];
  if (span.events) {
    span.events.forEach((event) => {
      let baseType;
      const time = (0, import_common.hrTimeToDate)(event.time);
      let name = "";
      let baseData;
      const properties = createPropertiesFromSpanAttributes(event.attributes);
      const tags = (0, import_common.createTagsFromResource)(span.resource);
      tags[import_generated.KnownContextTagKeys.AiOperationId] = span.spanContext().traceId;
      const spanId = span.spanContext().spanId;
      if (spanId) {
        tags[import_generated.KnownContextTagKeys.AiOperationParentId] = spanId;
      }
      if (event.name === "exception") {
        name = "Microsoft.ApplicationInsights.Exception";
        baseType = "ExceptionData";
        let typeName = "";
        let message = "Exception";
        let stack = "";
        let hasFullStack = false;
        if (event.attributes) {
          typeName = String(event.attributes[import_semantic_conventions.SEMATTRS_EXCEPTION_TYPE]);
          stack = String(event.attributes[import_semantic_conventions.SEMATTRS_EXCEPTION_STACKTRACE]);
          if (stack) {
            hasFullStack = true;
          }
          const exceptionMsg = event.attributes[import_semantic_conventions.SEMATTRS_EXCEPTION_MESSAGE];
          if (exceptionMsg) {
            message = String(exceptionMsg);
          }
          const escaped = event.attributes[import_semantic_conventions.SEMATTRS_EXCEPTION_ESCAPED];
          if (escaped !== void 0) {
            properties[import_semantic_conventions.SEMATTRS_EXCEPTION_ESCAPED] = String(escaped);
          }
        }
        const exceptionDetails = {
          typeName,
          message,
          stack,
          hasFullStack
        };
        const exceptionData = {
          kind: "ExceptionData",
          exceptions: [exceptionDetails],
          version: import_applicationinsights.DEFAULT_BREEZE_DATA_VERSION,
          properties: (0, import_common.truncateCustomDimensions)(properties)
        };
        baseData = exceptionData;
      } else {
        name = "Microsoft.ApplicationInsights.Message";
        baseType = "MessageData";
        const messageData = {
          kind: "MessageData",
          message: event.name,
          version: import_applicationinsights.DEFAULT_BREEZE_DATA_VERSION,
          properties: (0, import_common.truncateCustomDimensions)(properties)
        };
        baseData = messageData;
      }
      let sampleRate = 100;
      if (span.attributes[import_applicationinsights.AzureMonitorSampleRate]) {
        sampleRate = Number(span.attributes[import_applicationinsights.AzureMonitorSampleRate]);
      }
      if (baseData && "message" in baseData && baseData.message) {
        baseData.message = String(baseData.message).substring(
          0,
          import_types.MaxPropertyLengths.FIFTEEN_BIT
        );
      }
      const env = {
        name,
        time,
        instrumentationKey: ikey,
        version: 1,
        sampleRate,
        data: {
          baseType,
          baseData
        },
        tags
      };
      envelopes.push(env);
    });
  }
  return envelopes;
}
function getPeerIp(attributes) {
  if (attributes) {
    return attributes[import_semantic_conventions.ATTR_NETWORK_PEER_ADDRESS] || attributes[import_semantic_conventions.SEMATTRS_NET_PEER_IP];
  }
  return;
}
function getLocationIp(tags, attributes) {
  if (attributes) {
    const httpClientIp = getHttpClientIp(attributes);
    const netPeerIp = getPeerIp(attributes);
    if (httpClientIp) {
      tags[import_generated.KnownContextTagKeys.AiLocationIp] = String(httpClientIp);
    } else if (netPeerIp) {
      tags[import_generated.KnownContextTagKeys.AiLocationIp] = String(netPeerIp);
    }
  }
}
function getHttpClientIp(attributes) {
  if (attributes) {
    return attributes[import_semantic_conventions.ATTR_CLIENT_ADDRESS] || attributes[import_semantic_conventions.SEMATTRS_HTTP_CLIENT_IP];
  }
  return;
}
function getUserAgent(attributes) {
  if (attributes) {
    return attributes[import_semantic_conventions.ATTR_USER_AGENT_ORIGINAL] || attributes[import_semantic_conventions.SEMATTRS_HTTP_USER_AGENT];
  }
  return;
}
function getHttpUrl(attributes) {
  if (attributes) {
    return attributes[import_semantic_conventions.ATTR_URL_FULL] || attributes[import_semantic_conventions.SEMATTRS_HTTP_URL];
  }
  return;
}
function getHttpMethod(attributes) {
  if (attributes) {
    return attributes[import_semantic_conventions.ATTR_HTTP_REQUEST_METHOD] || attributes[import_semantic_conventions.SEMATTRS_HTTP_METHOD];
  }
  return;
}
function getHttpStatusCode(attributes) {
  if (attributes) {
    return attributes[import_semantic_conventions.ATTR_HTTP_RESPONSE_STATUS_CODE] || attributes[import_semantic_conventions.SEMATTRS_HTTP_STATUS_CODE];
  }
  return;
}
function getHttpScheme(attributes) {
  if (attributes) {
    return attributes[import_semantic_conventions.ATTR_URL_SCHEME] || attributes[import_semantic_conventions.SEMATTRS_HTTP_SCHEME];
  }
  return;
}
function getHttpTarget(attributes) {
  if (attributes) {
    if (attributes[import_semantic_conventions.ATTR_URL_PATH]) {
      return attributes[import_semantic_conventions.ATTR_URL_PATH];
    }
    if (attributes[import_semantic_conventions.ATTR_URL_QUERY]) {
      return attributes[import_semantic_conventions.ATTR_URL_QUERY];
    }
    return attributes[import_semantic_conventions.SEMATTRS_HTTP_TARGET];
  }
  return;
}
function getHttpHost(attributes) {
  if (attributes) {
    return attributes[import_semantic_conventions.ATTR_SERVER_ADDRESS] || attributes[import_semantic_conventions.SEMATTRS_HTTP_HOST];
  }
  return;
}
function getNetPeerName(attributes) {
  if (attributes) {
    return attributes[import_semantic_conventions.ATTR_CLIENT_ADDRESS] || attributes[import_semantic_conventions.SEMATTRS_NET_PEER_NAME];
  }
  return;
}
function getNetPeerPort(attributes) {
  if (attributes) {
    return attributes[import_semantic_conventions.ATTR_CLIENT_PORT] || attributes[import_semantic_conventions.ATTR_SERVER_PORT] || attributes[import_semantic_conventions.SEMATTRS_NET_PEER_PORT];
  }
  return;
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  getHttpClientIp,
  getHttpHost,
  getHttpMethod,
  getHttpScheme,
  getHttpStatusCode,
  getHttpTarget,
  getHttpUrl,
  getLocationIp,
  getNetPeerName,
  getNetPeerPort,
  getPeerIp,
  getUserAgent,
  readableSpanToEnvelope,
  spanEventsToEnvelopes
});
