// @azure/monitor-opentelemetry-exporter/dist/commonjs/export/base.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var base_exports = {};
__export(base_exports, {
  AzureMonitorBaseExporter: () => AzureMonitorBaseExporter
});
module.exports = __toCommonJS(base_exports);
var import_api = require("@opentelemetry/api");
var import_connectionStringParser = require("../utils/connectionStringParser.js");
var import_Constants = require("../Declarations/Constants.js");
class AzureMonitorBaseExporter {
  /**
   * Instrumentation key to be used for exported envelopes
   */
  instrumentationKey = "";
  /**
   * Ingestion Endpoint URL
   */
  endpointUrl = "";
  /**
   *Flag to determine if exporter will generate Statsbeat data
   */
  trackStatsbeat = false;
  /**
   * Instrumentation key to be used for exported envelopes
   */
  aadAudience;
  /**
   * Flag to determine if the Exporter is a Statsbeat Exporter
   */
  isStatsbeatExporter;
  /**
   * Exporter internal configuration
   */
  options;
  /**
   * Initializes a new instance of the AzureMonitorBaseExporter class.
   * @param AzureMonitorExporterOptions - Exporter configuration.
   */
  constructor(options = {}, isStatsbeatExporter) {
    this.options = options;
    this.instrumentationKey = "";
    this.endpointUrl = import_Constants.DEFAULT_BREEZE_ENDPOINT;
    const connectionString = this.options.connectionString || process.env[import_Constants.ENV_CONNECTION_STRING];
    this.isStatsbeatExporter = isStatsbeatExporter ? isStatsbeatExporter : false;
    if (connectionString) {
      const parsedConnectionString = import_connectionStringParser.ConnectionStringParser.parse(connectionString);
      this.instrumentationKey = parsedConnectionString.instrumentationkey || this.instrumentationKey;
      this.endpointUrl = parsedConnectionString.ingestionendpoint?.trim() || this.endpointUrl;
      this.aadAudience = parsedConnectionString.aadaudience;
    }
    if (!this.instrumentationKey) {
      const message = "No instrumentation key or connection string was provided to the Azure Monitor Exporter";
      import_api.diag.error(message);
      throw new Error(message);
    }
    if (!import_connectionStringParser.ConnectionStringParser.validateInstrumentationKey(this.instrumentationKey)) {
      const message = "Invalid instrumentation key was provided to the Azure Monitor Exporter";
      import_api.diag.error(message);
      throw new Error(message);
    }
    this.trackStatsbeat = !this.isStatsbeatExporter && !process.env[import_Constants.LEGACY_ENV_DISABLE_STATSBEAT];
    import_api.diag.debug("AzureMonitorExporter was successfully setup");
  }
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  AzureMonitorBaseExporter
});
