// @azure/monitor-opentelemetry-exporter/dist/commonjs/export/log.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var log_exports = {};
__export(log_exports, {
  AzureMonitorLogExporter: () => AzureMonitorLogExporter
});
module.exports = __toCommonJS(log_exports);
var import_api = require("@opentelemetry/api");
var import_core = require("@opentelemetry/core");
var import_base = require("./base.js");
var import_logUtils = require("../utils/logUtils.js");
var import_platform = require("../platform/index.js");
class AzureMonitorLogExporter extends import_base.AzureMonitorBaseExporter {
  /**
   * Flag to determine if Exporter is shutdown.
   */
  _isShutdown = false;
  _sender;
  /**
   * Initializes a new instance of the AzureMonitorLogExporter class.
   * @param AzureExporterConfig - Exporter configuration.
   */
  constructor(options = {}) {
    super(options);
    this._sender = new import_platform.HttpSender({
      endpointUrl: this.endpointUrl,
      instrumentationKey: this.instrumentationKey,
      trackStatsbeat: this.trackStatsbeat,
      exporterOptions: options,
      aadAudience: this.aadAudience
    });
    import_api.diag.debug("AzureMonitorLogExporter was successfully setup");
  }
  /**
   * Export OpenTelemetry logs.
   * @param logs - Logs to export.
   * @param resultCallback - Result callback.
   */
  async export(logs, resultCallback) {
    if (this._isShutdown) {
      import_api.diag.info("Exporter shut down. Failed to export spans.");
      setTimeout(() => resultCallback({ code: import_core.ExportResultCode.FAILED }), 0);
      return;
    }
    import_api.diag.info(`Exporting ${logs.length} logs(s). Converting to envelopes...`);
    const envelopes = [];
    logs.forEach((log) => {
      const envelope = (0, import_logUtils.logToEnvelope)(log, this.instrumentationKey);
      if (envelope) {
        envelopes.push(envelope);
      }
    });
    await import_api.context.with((0, import_core.suppressTracing)(import_api.context.active()), async () => {
      resultCallback(await this._sender.exportEnvelopes(envelopes));
    });
  }
  /**
   * Shutdown AzureMonitorLogExporter.
   */
  async shutdown() {
    this._isShutdown = true;
    import_api.diag.info("AzureMonitorLogExporter shutting down");
    return this._sender.shutdown();
  }
  /**
   * Force flush
   */
  async forceFlush() {
    return Promise.resolve();
  }
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  AzureMonitorLogExporter
});
