// @azure/monitor-opentelemetry-exporter/dist/commonjs/export/trace.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var trace_exports = {};
__export(trace_exports, {
  AzureMonitorTraceExporter: () => AzureMonitorTraceExporter
});
module.exports = __toCommonJS(trace_exports);
var import_api = require("@opentelemetry/api");
var import_core = require("@opentelemetry/core");
var import_base = require("./base.js");
var import_spanUtils = require("../utils/spanUtils.js");
var import_common = require("../utils/common.js");
var import_platform = require("../platform/index.js");
var import_connectionStringParser = require("../utils/connectionStringParser.js");
var import_Constants = require("../Declarations/Constants.js");
class AzureMonitorTraceExporter extends import_base.AzureMonitorBaseExporter {
  /**
   * Flag to determine if Exporter is shutdown.
   */
  isShutdown = false;
  sender;
  shouldCreateResourceMetric = (0, import_common.shouldCreateResourceMetric)();
  applicationId;
  /**
   * Initializes a new instance of the AzureMonitorTraceExporter class.
   * @param AzureExporterConfig - Exporter configuration.
   */
  constructor(options = {}) {
    super(options);
    const connectionString = options.connectionString || process.env[import_Constants.ENV_CONNECTION_STRING];
    if (connectionString) {
      const parsedConnectionString = import_connectionStringParser.ConnectionStringParser.parse(connectionString);
      this.applicationId = parsedConnectionString.applicationid;
    }
    this.sender = new import_platform.HttpSender({
      endpointUrl: this.endpointUrl,
      instrumentationKey: this.instrumentationKey,
      trackStatsbeat: this.trackStatsbeat,
      exporterOptions: options,
      aadAudience: this.aadAudience
    });
    import_api.diag.debug("AzureMonitorTraceExporter was successfully setup");
  }
  /**
   * Export OpenTelemetry spans.
   * @param spans - Spans to export.
   * @param resultCallback - Result callback.
   */
  async export(spans, resultCallback) {
    if (this.isShutdown) {
      import_api.diag.info("Exporter shut down. Failed to export spans.");
      setTimeout(() => resultCallback({ code: import_core.ExportResultCode.FAILED }), 0);
      return;
    }
    import_api.diag.info(`Exporting ${spans.length} span(s). Converting to envelopes...`);
    if (spans.length > 0) {
      const envelopes = [];
      const resourceMetricEnvelope = (0, import_common.createResourceMetricEnvelope)(
        spans[0].resource,
        this.instrumentationKey,
        this.applicationId
      );
      if (resourceMetricEnvelope && this.shouldCreateResourceMetric) {
        envelopes.push(resourceMetricEnvelope);
      }
      spans.forEach((span) => {
        envelopes.push((0, import_spanUtils.readableSpanToEnvelope)(span, this.instrumentationKey));
        const spanEventEnvelopes = (0, import_spanUtils.spanEventsToEnvelopes)(span, this.instrumentationKey);
        if (spanEventEnvelopes.length > 0) {
          envelopes.push(...spanEventEnvelopes);
        }
      });
      resultCallback(await this.sender.exportEnvelopes(envelopes));
    }
    resultCallback({ code: import_core.ExportResultCode.SUCCESS });
  }
  /**
   * Shutdown AzureMonitorTraceExporter.
   */
  async shutdown() {
    this.isShutdown = true;
    import_api.diag.info("AzureMonitorTraceExporter shutting down");
    return this.sender.shutdown();
  }
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  AzureMonitorTraceExporter
});
