// @azure/monitor-opentelemetry-exporter/dist/commonjs/platform/nodejs/persist/fileAccessControl.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var fileAccessControl_exports = {};
__export(fileAccessControl_exports, {
  FileAccessControl: () => FileAccessControl
});
module.exports = __toCommonJS(fileAccessControl_exports);
var import_node_fs = require("node:fs");
var import_node_os = require("node:os");
var import_node_child_process = require("node:child_process");
var import_api = require("@opentelemetry/api");
var import_node_process = __toESM(require("node:process"));
class FileAccessControl {
  static ICACLS_PATH = `${import_node_process.default.env.SYSTEMDRIVE}/windows/system32/icacls.exe`;
  static POWERSHELL_PATH = `${import_node_process.default.env.SYSTEMDRIVE}/windows/system32/windowspowershell/v1.0/powershell.exe`;
  static ACLED_DIRECTORIES = {};
  static ACL_IDENTITY = null;
  static OS_FILE_PROTECTION_CHECKED = false;
  static OS_PROVIDES_FILE_PROTECTION = false;
  static USE_ICACLS = (0, import_node_os.type)() === "Windows_NT";
  // Check if file access control could be enabled
  static checkFileProtection() {
    if (!FileAccessControl.OS_PROVIDES_FILE_PROTECTION && !FileAccessControl.OS_FILE_PROTECTION_CHECKED) {
      FileAccessControl.OS_FILE_PROTECTION_CHECKED = true;
      if (FileAccessControl.USE_ICACLS) {
        try {
          FileAccessControl.OS_PROVIDES_FILE_PROTECTION = (0, import_node_fs.existsSync)(FileAccessControl.ICACLS_PATH);
        } catch (e) {
        }
        if (!FileAccessControl.OS_PROVIDES_FILE_PROTECTION) {
          import_api.diag.warn(
            "Could not find ICACLS in expected location! This is necessary to use disk retry mode on Windows."
          );
        }
      } else {
        FileAccessControl.OS_PROVIDES_FILE_PROTECTION = true;
      }
    }
  }
  static async applyACLRules(directory) {
    if (FileAccessControl.USE_ICACLS) {
      if (FileAccessControl.ACLED_DIRECTORIES[directory] === void 0) {
        FileAccessControl.ACLED_DIRECTORIES[directory] = false;
        try {
          const identity = await this._getACLIdentity();
          await this._runICACLS(this._getACLArguments(directory, identity));
          FileAccessControl.ACLED_DIRECTORIES[directory] = true;
        } catch (ex) {
          FileAccessControl.ACLED_DIRECTORIES[directory] = false;
          throw ex;
        }
      } else {
        if (!FileAccessControl.ACLED_DIRECTORIES[directory]) {
          throw new Error("Setting ACL restrictions did not succeed (cached result)");
        }
      }
    }
  }
  static applyACLRulesSync(directory) {
    if (FileAccessControl.USE_ICACLS) {
      if (FileAccessControl.ACLED_DIRECTORIES[directory] === void 0) {
        this._runICACLSSync(this._getACLArguments(directory, this._getACLIdentitySync()));
        FileAccessControl.ACLED_DIRECTORIES[directory] = true;
        return;
      } else if (!FileAccessControl.ACLED_DIRECTORIES[directory]) {
        throw new Error("Setting ACL restrictions did not succeed (cached result)");
      }
    }
  }
  static _runICACLS(args) {
    return new Promise((resolve, reject) => {
      const aclProc = (0, import_node_child_process.spawn)(FileAccessControl.ICACLS_PATH, args, {
        windowsHide: true
      });
      aclProc.on("error", (e) => reject(e));
      aclProc.on("close", (code) => {
        if (code === 0) {
          resolve();
        } else {
          reject(
            new Error(`Setting ACL restrictions did not succeed (ICACLS returned code ${code})`)
          );
        }
      });
    });
  }
  static _runICACLSSync(args) {
    if (import_node_child_process.spawnSync) {
      const aclProc = (0, import_node_child_process.spawnSync)(FileAccessControl.ICACLS_PATH, args, {
        windowsHide: true
      });
      if (aclProc.error) {
        throw aclProc.error;
      } else if (aclProc.status !== 0) {
        throw new Error(
          `Setting ACL restrictions did not succeed (ICACLS returned code ${aclProc.status})`
        );
      }
    } else {
      throw new Error("Could not synchronously call ICACLS under current version of Node.js");
    }
  }
  static _getACLIdentity() {
    return new Promise((resolve, reject) => {
      if (FileAccessControl.ACL_IDENTITY) {
        resolve(FileAccessControl.ACL_IDENTITY);
      }
      const psProc = (0, import_node_child_process.spawn)(
        FileAccessControl.POWERSHELL_PATH,
        ["-Command", "[System.Security.Principal.WindowsIdentity]::GetCurrent().Name"],
        {
          windowsHide: true,
          stdio: ["ignore", "pipe", "pipe"]
          // Needed to prevent hanging on Win 7
        }
      );
      let data = "";
      psProc.stdout.on("data", (d) => data += d);
      psProc.on("error", (e) => reject(e));
      psProc.on("close", (code) => {
        FileAccessControl.ACL_IDENTITY = data && data.trim();
        if (code === 0) {
          resolve(FileAccessControl.ACL_IDENTITY);
        } else {
          reject(new Error(`Getting ACL identity did not succeed (PS returned code ${code})`));
        }
      });
    });
  }
  static _getACLIdentitySync() {
    if (FileAccessControl.ACL_IDENTITY) {
      return FileAccessControl.ACL_IDENTITY;
    }
    if (import_node_child_process.spawnSync) {
      const psProc = (0, import_node_child_process.spawnSync)(
        FileAccessControl.POWERSHELL_PATH,
        ["-Command", "[System.Security.Principal.WindowsIdentity]::GetCurrent().Name"],
        {
          windowsHide: true,
          stdio: ["ignore", "pipe", "pipe"]
          // Needed to prevent hanging on Win 7
        }
      );
      if (psProc.error) {
        throw psProc.error;
      } else if (psProc.status !== 0) {
        throw new Error(`Getting ACL identity did not succeed (PS returned code ${psProc.status})`);
      }
      FileAccessControl.ACL_IDENTITY = psProc.stdout && psProc.stdout.toString().trim();
      return FileAccessControl.ACL_IDENTITY;
    } else {
      throw new Error("Could not synchronously get ACL identity under current version of Node.js");
    }
  }
  static _getACLArguments(directory, identity) {
    return [
      directory,
      "/grant",
      "*S-1-5-32-544:(OI)(CI)F",
      // Full permission for Administrators
      "/grant",
      `${identity}:(OI)(CI)F`,
      // Full permission for current user
      "/inheritance:r"
    ];
  }
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  FileAccessControl
});
